{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1398abd7-d8da-4ba8-a7a8-b1072da18e09",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "f01c7933-6272-4d49-9383-9c1a1f6262f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown, display\n",
    "from rdkit import Chem\n",
    "import rdkit\n",
    "from rdkit.Chem import AllChem, Draw\n",
    "from pymol import cmd\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "import pubchempy \n",
    "import pandas as pd\n",
    "from biopandas.pdb import PandasPdb\n",
    "from io import StringIO\n",
    "\n",
    "def mol_with_atom_index( mol ):\n",
    "    atoms = mol.GetNumAtoms()\n",
    "    for idx in range( atoms ):\n",
    "        mol.GetAtomWithIdx( idx ).SetProp( 'molAtomMapNumber', str( mol.GetAtomWithIdx( idx ).GetIdx() ) )\n",
    "    return mol\n",
    "def printmd(string):\n",
    "    display(Markdown(string))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa357c7f-a465-418c-980a-5c3375bf23fb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Search on PubChem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "70102af5-6417-47a2-9264-e1777862320c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n",
      "/home/marco/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:563: SyntaxWarning: \"is not\" with a literal. Did you mean \"!=\"?\n",
      "  if self.charge is not 0:\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mgaierror\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:1354\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1353\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1354\u001b[0m     \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselector\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1355\u001b[0m \u001b[43m              \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhas_header\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTransfer-encoding\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1356\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:1256\u001b[0m, in \u001b[0;36mHTTPConnection.request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1255\u001b[0m \u001b[38;5;124;03m\"\"\"Send a complete request to the server.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1256\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:1302\u001b[0m, in \u001b[0;36mHTTPConnection._send_request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1301\u001b[0m     body \u001b[38;5;241m=\u001b[39m _encode(body, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbody\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m-> 1302\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mendheaders\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:1251\u001b[0m, in \u001b[0;36mHTTPConnection.endheaders\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1250\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CannotSendHeader()\n\u001b[0;32m-> 1251\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_output\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmessage_body\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:1011\u001b[0m, in \u001b[0;36mHTTPConnection._send_output\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1010\u001b[0m \u001b[38;5;28;01mdel\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer[:]\n\u001b[0;32m-> 1011\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1013\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m message_body \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1014\u001b[0m \n\u001b[1;32m   1015\u001b[0m     \u001b[38;5;66;03m# create a consistent interface to message_body\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:951\u001b[0m, in \u001b[0;36mHTTPConnection.send\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m    950\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mauto_open:\n\u001b[0;32m--> 951\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    952\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:1418\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1416\u001b[0m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnect to a host on a given (SSL) port.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1418\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1420\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tunnel_host:\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/http/client.py:922\u001b[0m, in \u001b[0;36mHTTPConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    921\u001b[0m \u001b[38;5;124;03m\"\"\"Connect to the host and port specified in __init__.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 922\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    923\u001b[0m \u001b[43m    \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_address\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    924\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock\u001b[38;5;241m.\u001b[39msetsockopt(socket\u001b[38;5;241m.\u001b[39mIPPROTO_TCP, socket\u001b[38;5;241m.\u001b[39mTCP_NODELAY, \u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/socket.py:787\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address)\u001b[0m\n\u001b[1;32m    786\u001b[0m err \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 787\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mSOCK_STREAM\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    788\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/socket.py:918\u001b[0m, in \u001b[0;36mgetaddrinfo\u001b[0;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[1;32m    917\u001b[0m addrlist \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m--> 918\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43m_socket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mproto\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    919\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "\u001b[0;31mgaierror\u001b[0m: [Errno -3] Temporary failure in name resolution",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [304]\u001b[0m, in \u001b[0;36m<cell line: 7>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlig_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.pdb\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      5\u001b[0m adfr_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/home/marco/ADFRsuite-1.0/bin\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 7\u001b[0m SearchResults \u001b[38;5;241m=\u001b[39m \u001b[43mpubchempy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_compounds\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyword\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnamespace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msearchtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mas_dataframe\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m n, i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m([x\u001b[38;5;241m.\u001b[39msynonyms[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m SearchResults]):\n\u001b[1;32m      9\u001b[0m     printmd(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSelect **\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mn\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m** for `\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m`\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:321\u001b[0m, in \u001b[0;36mget_compounds\u001b[0;34m(identifier, namespace, searchtype, as_dataframe, **kwargs)\u001b[0m\n\u001b[1;32m    312\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_compounds\u001b[39m(identifier, namespace\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcid\u001b[39m\u001b[38;5;124m'\u001b[39m, searchtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, as_dataframe\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;124;03m\"\"\"Retrieve the specified compound records from PubChem.\u001b[39;00m\n\u001b[1;32m    314\u001b[0m \n\u001b[1;32m    315\u001b[0m \u001b[38;5;124;03m    :param identifier: The compound identifier to use as a search query.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    319\u001b[0m \u001b[38;5;124;03m                         :class:`~pandas.DataFrame` and return that.\u001b[39;00m\n\u001b[1;32m    320\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 321\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mget_json\u001b[49m\u001b[43m(\u001b[49m\u001b[43midentifier\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnamespace\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msearchtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msearchtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    322\u001b[0m     compounds \u001b[38;5;241m=\u001b[39m [Compound(r) \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m results[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPC_Compounds\u001b[39m\u001b[38;5;124m'\u001b[39m]] \u001b[38;5;28;01mif\u001b[39;00m results \u001b[38;5;28;01melse\u001b[39;00m []\n\u001b[1;32m    323\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m as_dataframe:\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:299\u001b[0m, in \u001b[0;36mget_json\u001b[0;34m(identifier, namespace, domain, operation, searchtype, **kwargs)\u001b[0m\n\u001b[1;32m    297\u001b[0m \u001b[38;5;124;03m\"\"\"Request wrapper that automatically parses JSON response and supresses NotFoundError.\"\"\"\u001b[39;00m\n\u001b[1;32m    298\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 299\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m json\u001b[38;5;241m.\u001b[39mloads(\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43midentifier\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnamespace\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdomain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mJSON\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msearchtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mdecode())\n\u001b[1;32m    300\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m NotFoundError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    301\u001b[0m     log\u001b[38;5;241m.\u001b[39minfo(e)\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:292\u001b[0m, in \u001b[0;36mget\u001b[0;34m(identifier, namespace, domain, operation, output, searchtype, **kwargs)\u001b[0m\n\u001b[1;32m    290\u001b[0m             response \u001b[38;5;241m=\u001b[39m request(identifier, namespace, domain, operation, output, searchtype, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m    291\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 292\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43midentifier\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnamespace\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdomain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msearchtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m    293\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/site-packages/pubchempy.py:271\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(identifier, namespace, domain, operation, output, searchtype, **kwargs)\u001b[0m\n\u001b[1;32m    269\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRequest URL: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m, apiurl)\n\u001b[1;32m    270\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRequest data: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m, postdata)\n\u001b[0;32m--> 271\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mapiurl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpostdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    272\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response\n\u001b[1;32m    273\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m HTTPError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:222\u001b[0m, in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    220\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    221\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[0;32m--> 222\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:525\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    522\u001b[0m     req \u001b[38;5;241m=\u001b[39m meth(req)\n\u001b[1;32m    524\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124murllib.Request\u001b[39m\u001b[38;5;124m'\u001b[39m, req\u001b[38;5;241m.\u001b[39mfull_url, req\u001b[38;5;241m.\u001b[39mdata, req\u001b[38;5;241m.\u001b[39mheaders, req\u001b[38;5;241m.\u001b[39mget_method())\n\u001b[0;32m--> 525\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    527\u001b[0m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[1;32m    528\u001b[0m meth_name \u001b[38;5;241m=\u001b[39m protocol\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_response\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:542\u001b[0m, in \u001b[0;36mOpenerDirector._open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[1;32m    541\u001b[0m protocol \u001b[38;5;241m=\u001b[39m req\u001b[38;5;241m.\u001b[39mtype\n\u001b[0;32m--> 542\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_open\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\n\u001b[1;32m    543\u001b[0m \u001b[43m                          \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_open\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    544\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[1;32m    545\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:502\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[1;32m    501\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[0;32m--> 502\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    503\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    504\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:1397\u001b[0m, in \u001b[0;36mHTTPSHandler.https_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1396\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[0;32m-> 1397\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mHTTPSConnection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1398\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_hostname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/revdock/lib/python3.8/urllib/request.py:1354\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1352\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1353\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1354\u001b[0m         \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselector\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1355\u001b[0m \u001b[43m                  \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhas_header\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTransfer-encoding\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1356\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[1;32m   1357\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "keyword = 'kynurenine'\n",
    "lig_res = 'LIG'\n",
    "lig_name = f'{lig_res}_PLP'\n",
    "file = f'{lig_name}.pdb'\n",
    "adfr_path = \"/home/marco/ADFRsuite-1.0/bin\"\n",
    "\n",
    "SearchResults = pubchempy.get_compounds(keyword, namespace='name', searchtype=None, as_dataframe=False)\n",
    "for n, i in enumerate([x.synonyms[0] for x in SearchResults]):\n",
    "    printmd(f'Select **{n}** for `{i}`')\n",
    "aa = SearchResults[int(input())].isomeric_smiles\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "cfbaa4b8-7050-4f71-b956-8450ad299d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABdwAAAEsCAIAAABWiuY5AAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nOzdZ1gU19sG8Ht3QUBBQKrYEHtBQJBYYtckKhpjITasiZioWGLJq7FFEzVWsCRq7CZYgooGK9Zo1ERULCAK2ADpoCCdnffD8EeClEUWdlfv38WHZebMmWcW8Lr29hSJIAggIiIiIiIiIqLKJVV1AURERERERERE7yOGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFSAoQwRERERERERkQowlCEiIiIiIiIiUgGGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFSAoQwRERERERERkQowlCEiIiIiIiIiUgGGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFSAoQwRERERERERkQowlCEiIiIiIiIiUgGGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFSAoQwRERERERERkQowlCEiIiIiIiIiUgGGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFSAoQwRERERERERkQowlCEiIiIiIiIiUgGGMkREREREREREKsBQhoiIiIiIiIhIBRjKEBERERERERGpAEMZIiIiIiIiIiIVYChDRERERERERKQCDGWIiIiIiIiIiFRAS9UFvDtOn0ZyMiQS9O8PraLeVx8fyOVo1w61a5fUz5UrePYMTk6wsSnljuHhCAjIe+3khPr1ASAiAleuAMDgwUVflZYGPz8A6N4dNWqUcgsiIiIiIiIiqiASQRBUXcM7wt4egYEAsGYNpk4tooGuLjIzceAABg0q4qxcDj8/LF+Oy5cB4NNPcfhwSbfLyUHbtq9Dma1bMXYsAOzfj88/z2sgkxVx4ePHefHN33+jXTvFno2IiIiIiIiIlI3Tl5Rv/nxERJShfWoq1q5Fw4bo1y8vkVHEypUICIC1NQwN36JGIiIiIiIiIlIxhjJKpqODlBRMmVKGS54/x/TpiI3FwIHYtavocTSFPHyI778HAC8vVKnylqUSERERERERkQoxlFGyCRNQpQoOHsTRo4pe0qgRLlxAfDz++ANubjAyKqW9IOCrr5CejsGD0bdvOeslIiIiIiIiItVgKKNk9evD3R0APDyQlqboVR07QldX0cZbtuDMGVSvjjVr3qZCIiIiIiIiIlIHDGWUb9EimJri8WMsXlxsm9u3cfx4GVaQyff8Ob79FgB+/BG1ar19kURERERERESkWgxllM/YOC+OWbUKd+8W3eann9C7NyZNKnPnEyciKQlt2mDChHIVSURERERERESqxVCmQowfD2dnZGfjq6+gxD3H//gDhw5BSwubNhW93XVB+vrQ0yviq2lTpdVDRERERERERG+NoUyFkEqxbh2kUly6hG3bimjw008ICsLBg2Xo88ULTJ0KAFOmwMGh9PaZmUV/ZWWV4aZEREREREREVEEYylQUZ2d88QUAzJqFxMTCZ62s0KwZ6tcvQ4fTpyMyEnXrYuFChdpnZ0MuL+IrPLwMNyUiIiIiIiKiCsJQpgItXQozMyQmYsmS8nZ1/jy2bweAdeugr1/+0oiIiIiIiIhIxbRUXcC7rEYNLFuGceOwfn151+WdPj1vbZrRowufSk4GgMmTMWMGFixAzZrlupHqZWTgzh3ExkIuh6kpWrVCtWoltU9KQlJSKX3Wrw+JRIk1EhEREREREZUfQ5mKNWYMdu7ExYt5+1i/tfwJUMXlD2lpSEtDRka57qJiwcFYuBB//om0tNcHdXTQsycWLoSjY9FXrV+P+fNL6Tk7G1r8VSciIiIiIiL1wk+qFUsigZcXnJxw6FC5xmo8flzsKXNzxMVh61aMHQsA+/e//V1Uaf9+jByJzEwAaNUKTZpAJsODB7h1C3/+iePH4eWFr78u4sKaNYvNa2JiEBEBXV1IOU2PiIiIiIiI1A5DmQpnZ4eJE+Hp+Z+9sadPx+HDaNECR4+qrrL/CQ/HTz/h0iXk5KB5c0yfjg8/rNwK/v4bw4cjJwdt2mDrVtjavj714AG+/BIXL2LiRNSqhU8/LXztF1/kraj8pl69EBGBESMYyhAREREREZEa4ofVyrB4MWrV+s+R2Fg8eoSIiLxvd+xAjRp5X7t3A8CxY6+PxMZWYG3BwWjTBhcvYsQIfPEF7t1Dt264erUC71iYIOCLL5CTgxYtcPbsfxIZAI0b4+RJODsDwIQJZZigdfcuTp6ERIJp05RcMBEREREREZEycKRMZTAwwIoVGDbs9ZGePWFoiDp18r7NzCy8WEx29usjcnkZ7pU/Saq42VKFGoSEwMEBvr55y+kOGYIGDbBpE9q2LcNNy+X0aQQHA8CaNUXvLKWriw0b0KYNoqNx4ADc3BTqdvVqCAJ690bz5sqsloiIiIiIiEhJJELBSTVEQLNmMDbG339X1v2mTcPatahZExERJc0zatkS9+5h6FD8/jsSE5GVBV1dGBkV3TgmBtbWyMjAmTPo1q2CCiciIiIiIiIqD05fqhAZGdDQsEsQEB+PGjUq8ZbXrwOAo2MpK7+IQ3fExi4uqFkT48YV29jLCxkZaNUKXbsqsVIiIiIiIiIiJWIoUyHmzYOuLn7+WdV1lJ2fH+Lj4epaibcUl8ypXbuUZuKqPIqsr5OWhk2bAGDmzHJteUVERERERERUkbimTIV4+hRZWTA2VnUdZZSQgIkT4eyMESMq8a5paQCgp1dKM3G5mdRUAFi4EHFxqFu36JbbtiEhAbVqVW62RERERERERFQ2DGUqxNOnAF6v46sREhPx0UeQSHDgQOVuIS2mLWI0UwIxjjE0BICPPiq2mVwOT08A8PBAlSrKqZCIiIiIiIioAnD6UoV49gxAsSM51NCdO/jgAyQl4cyZSi/bzAwAIiNLaSbuH25uXkqzQ4cQGopq1fDFF8oojoiIiIiIiKiiMJRRvuxsREdDJkPNmqouRTG//4527WBlhatX0aBBpd/e0REAAgJKWRv5338BwMmplN5WrQKAceMqd7Hid0FGRikNEhNx5w6ePtXURayJiIiIiIjUDUMZ5YuMRG4urKygpQmTw374AcOHQxDQsCHmzYO7e96XXF5ZFYhzkZ4/x/nzxba5fRt37gDAxx+X1NW//+LKFchk8PBQZoVq7PJlSCSQSGBnh5ycIhps3QqJBBYWJXUSGIjRo2FoiBo1cPly4bNpaVi2DLVqwcQErVqhXj2YmGDiRMTH5zW4eTOvhpCQYm+hrQ2JBLt3l+nhiIiIiIiI3nGaEBtoGs2au6StjcGDASAlRUUVfPwxGjZEaChmzsTly9DRKdwgNxfTpgGApWVercVZvhwAPvtMFQN+VOz2bWzYgClTynbVpUtYvhx+fnmDX7KyCv8aREfDxQUBAZBI0LAhDA3x7BliY7FxI44fx6VLsLJ6PXCmhBE04ikOsSEiIiIiIiqIoYzyadYqv7NmqboCqRS//IKePREQgH79sG1b3u7Xovh4uLvj7FkAWL8+L7KZOxf376NdO8yY8brlo0c4fBgApk+vzPLVx7x5GDiw9L3F802ahA0bAKBuXfTpU/QO7mPGICAA7dtj5040bAgAgoDduzFuHB49wpw52LFDOcUTERERERG9hxjKKJ8YymjKSBm10L07fv4ZEyfi1Ck0bIju3dGkCSQShIXB3x+pqZBIsGwZBg7Ma3/uHK5cKdzJ6tXIzUX79mjXrpLLVwfdu+PMGUybhgMHFL3E1RWmpujXD61bIzu7iFAmKwv29nj8GMePo3r1vIMSCUaOxLlz2LEDx44prX4iIiIiIqL3EEMZ5ROnL2nKSBl14e4OW1t8+y0uXYKfH/z8Xp9q0wZLl6J799dHbG0BoGnT10eSkvLGbHzzTaWUq3aWLcMHH+CPP+Dnhz59FLqkUyd06lRSgypVsHQpfvihiC3SxfE4pa4NTERERERERCVgKKN8DGXeUvv2uHgRUVG4ehXPn0MQYGEBZ2fUq1e45aZNhY9IpXnrBNvbV0KlaqhVK4wbhy1b8PXXCApCtWpK6/nNRAb/2wvLzk5pdyEiIiIiInoPMZRRPk5fKoPISISFoWFDWFnlHbGywoABZe7H0DBva+332NKlOHgQT59iyRIsXVpEg5SUvLd5/XqMGvWWd8nNxc8/49QpyGT47ru3r5aIiIiIiIgYyigfR8qUwbFjGD8e48bh119VXYrGMzHB/PmYMgWrVmHECLRoUbiBICA1FQCyssrc+V9/wdcXT5/i+nU8egQLC6xfX3iD8v/7v9dLzxRSeTusExERERERaQ6GMkqWmoqkJFStClNTVZeiERRJsG7fxsCB6NyZwU2pJk7Ejh24eRMTJuDiRUgk/zmrq4uFCwG8zaCif//FqlWvv+3Zs4i5S+LmV0RERERERKQghjJKpln7YaueIu/Xo0cIDf3Psr5UDJkMGzbgww9x6RJ27So8R6lKFSxY8JY9Dx+Ozp0RG4snT7BnD/bsga8vjh5F586v23h6vp6FVsjnn3OwDBERERERUWEMZZSMc5fKRpEFeLhIT1m0a4cxY7B1K2bPxmefKa1bCwtYWOS9dnfH4MHw8cHYsbh//3Wbjz4qNjorNGaHiIiIiIiIABS1sQqVAwOEslEkxGLQVUbLlsHEBDExRS/3W34SCb76CgDCw3HjRoXcgoiIiIiI6H3AUEbJGCCUgSAgIgJgKKNkpqZ5cczatXkpodLZ2OS9EH84RERERERE9BYYyigZA4QyiI1FRgZMTVG1aknNOPqo7MaNQ9u2yMiAp2e5+klORkhIEccfPMh7kT+niYiIiIiIiMqKa8ooGQOEMhATrFLfLAZdZSeVYvNmtG6NFy9eH8zMzFvod+BAtGlTeicvX+KTTxAejkOH0KHD6+OpqZgzBwAsLNCmDYKC3qbC69dx+jTS0mBjA1dXVKv2Np0QERERERFpNIYySvb8uQBIatdWdR0aQZGtl3Jz8fw5pFLUqlU5Rb0zbG0xYQLWr399JDMTy5cDQIMGeaHM4sV5E8jyt0by8sKhQwBgb48xY1CjBq5dQ+fOcHFBhw7Q18ejR9i9G9HRkEiwciV0dctcmFyOL7/Etm1o1gw1amDlSixYgFOnuL8WERERERG9dxjKKJMgCE+fGlpYNKxb92+g7J9W3zeKjJSJikJODmrVgrZ25RT1LvnhBxw8iKiovG+lUpiaAoCeXt4RHx8EBv7nkuPH81589hm++gqHD+OHH7B8OXx94ev7upmZGdatw+efv01VJ09ixw7s34/BgwHg1i188AHmzoWPz9v0RkREREREpLkYyihTXFzcq1cpurpPq1VjIqMAReYlKTKa5j3WpAn27wcAraL+lKtXh58fHj6Ejg4A6OsjLu4/DVasQHJy0T2LI5OqVMGiRZg2DSdO4MEDvHwJY2M4OKBbt9djZGxs8mooYTDTvn2Qy+HsDAC9eiE8HPXq5Z2yt0fHjrh+XcEnJiIiIiIiencwlFGmZ8+eAajLFWUUpEjgouC6M+8rU9O88SbFsbeHvX2xZ3v2VOguRkYYMqSksyXXAGDgwP98m5/IiGQyZGUpVAkREREREdG7hLsvKdPTp08B1OGwDgUpErhwpExZuLpi/HhkZKi6jrJIScHVq+jUSdV1EBERERERVTqGMsrEkTJlo/hIGYYyCkhLw4ED2L07b7KSppg/H+npWLhQ1XUQERERERFVOk5fUiYxlOFIGYVkZyMmBjIZatYsqRmnLyksfzt2iUTVpShs82Z4emLlSjRrpupSiIiIiIiIKh1HyigTpy+VQUQEcnNRq1bRS9Tm4/QlhWnWWyUIWLYMEyZg/nxMn67qaoiIiIiIiFSBI2WUidOXykDBeUkcKaMwDXqrUlIwdiyOHMGmTfjyS1VXQ0REREREpCIMZZSJI2XKIH+yTQnS0hAfDx0dmJlVTlEaTVOW3wkKwmefISwM8+ejfn34++cdb98eVauqtDIiIiIiIqLKxVBGabKzs6Ojo2UymZWVlapr0QSKRAj5bTRolRTV0ZTpS1u34sEDAFiw4D/H799HkyYqqYiIiIiIiEg1uKaM0kRFReXm5lpZWWmVvEgKiRQPZTRiQo4a0JR3a9UqCEIRX0xkiIiIiIjofcNQRmk4d6lMAqXSaGfnVBubkhppytgP9cB3i4iIiIiISLMwlFEarvJbJiP/+qvmP/+EljjVS4iMBBgzKCoiAuC7RUREREREpDkYyigNR8qUiSJv1/inTxvXqOHboEFlFaXB4uORloYaNaCvr+pSiIiIiIiISDEMZZRGHCnDUEYRqampycnJVatWNTExKaHZkydPHiYm6lhaVlphmotzl4iIiIiIiDQOQxml4fQlxYnDZEp9r/iWKk5TVvklIiIiIiKifAxllIbTlxSn4HslhjK1a9eujJo0HEfKEBERERERaRyGMkqj4OgPgmJDYBISEl69emVkZFS9evXKqkuDKbLDOBEREREREakVhjLKkZqampSUpKenV/IiKSRSZP0dhlxlwulLREREREREGoehjHLkpwwSiUTVtWgARaYvceHkMuH0JSIiIiIiIo2jpeoCNIMgCFeuXDl79uyjR49SUlJMTExatGjRv3///OVOihzWcevWLV9f30ePHmVkZNSpU6dLly6ffPKJTCZTwQOoGUWmL3GkTJlwpAwREREREZHGYShTuqtXr06cOPHGjRuFjk+fPn3GjBnff/+9lpbWxx9/HB8fn5aWJp568eLF6NGjDx8+XLD9ypUr7e3tDx48WL9+/UoqXV1xpIxy5ebi+XPIZLCyUnUpREREREREpDCGMqU4fvz4wIED09PTtbS0BgwY0KVLF1NT0+fPn/v5+Z06dWrp0qXR0dHbtm0DYGJiIi4ok5GR0blz58DAwPr1648bN65ly5aZmZl//fXX5s2bb9261atXr8DAQB0dHVU/mcoIghAZGYnStlXiftiKi4xETg5q1YK2tqpLISIiIiIiIoUxlClJVFSUm5tbenq6lZXV8ePHW7VqlX/Kw8PDx8dnypQpY8aMKXTV77//HhgY6OTk5O/vb2hoKB50dXVt3779sGHDQkJC/vzzz4EDB1beY6iZuLi49PR0ExOTatWqldCMW4wrjnOXiIiIiIiINBFDmZKsWbMmISFBKpX6+PgUTGREAwcOdHFxeXPMy9ixYy0tLW1sbPITGdHgwYNHjx6dlZV1//79iq1bvSm4WAxHyiguJuZihw73nJ1bAJ1UXQsREREREREpirsvFUsQhF27dgHo1atX27Zti2xT3Cyk3r17N23atNDBFy9eZGVlATAzM1NqpRpGkcVicnNzo6KipFKpFVdJUUBo6NXLl7+WyY6quhAiIiIiIiIqA46UKVZYWFhsbCyA/v37l9wyOzt73bp1AHr27Glra1tcsz179gDQ0dHp1auXUivVMIqMlImKisrJybGysqpSpUpl1aXBuCgyERERERGRJmIoU6yQkBDxxZsTlwrJyMj45ptvAGzatKlQKJOcnJyenh4WFrZ79+5t27bJZLKNGze+5x+eFUkQOHepTLj+DhERERERkSZiKFOspKQk8YWpqelbd9KwYcOEhATxdc2aNQ8ePFjcTKj3hyKhDFOGMmGGRUREREREpIkYyhQrNzdXfCGTyUpuaWBgIAhCkac6d+6clJQUExMTFhb2/Pnz4cOHb9++vVOn93o1VkWmLym4GDCJmGERERERERFpIi70W6z8vZNevnz51p34+PicPXv23r170dHRY8aMCQ8P/+STT4KDg5VUo0ZSfPoSUwZFpKWlJSQk6OrqvucLSBMREREREWkchjLFql27tvgif3GZ8jAyMtqyZYuNjU16evrKlSvL36GGys7Ojo6OlslkJW+rxFBGcfnDZCQSiaprISIiIiIiojLg9KVitWrVSldXNyMj4/z584MGDSp/hzKZrE2bNuHh4YGBgeXvTYMkJSWdP38+LCwsISFBX19/0KBBpqamWlol/e6JQUNqauq6devi4uKqVatWr169jz76qEaNGpVVtcZggEVERERERKShGMoUq0qVKn369PHx8fH29l60aJGJiUn5+xSXp8nMzCx/VxohNjZ27ty5O3bsyMnJKXhcX1+/fv36U6dOLW69nhcvXgAYNWpUwYN6enobN24cPXp0hdWrkbj+DhERERERkYZiKFOSadOmHTx4MDExcezYsQcOHKhSpUqhBoIgiHNGxJxFW1tbKpWmpKS4ubktWrTIzs6uYOOkpKSzZ88CKLRt9rsqPDy8Z8+e4eHhAJo1a9azZ09LS8v4+PjTp0/fuXNnxowZQUFBv/7665uTbn799dfHjx8DMDc3/+yzzxo3bpyWlhYcHHz69OkePXpU/oOoOW69REREREREpKEYyhT2+PFjbW3tWrVqAejQocO0adNWr1595MiRDh06LFiwoHv37np6ellZWaGhoX/88cf27dv/+usvQ0PD6tWrA9i0adP48eOXL1/u6+t77ty5H3/8ccSIEeKCwbdu3Zo8eXJ0dLRUKp00aZKKH7LiZWVlubq6hoeHa2tr//zzz2PHji0Yvmzfvv3rr79u3rz5m4nMnTt3vv76a7lc/umnn+7atUt8Y0U5OTklT3p6P3H6EhERERERkYbiR9zXAgICPD09vb29x48fv2HDBvHgypUrtbS0Vq5cef369b59+0okEmNj45cvX+bPx/H19R05cmTBfubPnx8WFrZ3795JkyZ5eHiYmZm9evUqNTUVgEwmW7NmTfv27Sv50Srf77//HhAQAGDNmjXjxo0rdHbMmDG9evWytLR888I5c+ZkZ2c3aNDA29tbT0+v4CkmMkXifthEREREREQaip9ykZGR4e3t7eXldevWLQBVqlSRy+X5ZyUSyfLly4cPH75x40Z/f/+nT58mJiZKpdIGDRq0bdvW1dXVxcUlNzd33bp1ADp06CD24O3tPWzYsC1btly4cCEmJgaAmZlZjx49pk+f7uTkpKIHrVRbt24F0KBBg6+++qrIBkUmMjExMSdOnAAwbdq0QokMFYfTl4iIiIiIiDSURBAEVdegMjExMTt27PDy8oqKigJgbm4+ZsyYSZMm5W+GXaTU1FR9fX3F75KWliaVSnV1dctbrubIzMw0NDTMzMycNWvW8uXLS258/Pjx7OzsJk2aNGnSxMfHR9zo6vHjx/Xq1auUYjVetWrV0tLSXr58aWBgoOpaiIiIiIiIqAze05EyAQEBmzdv3rVrV0ZGBgAHB4cJEya4ubkpMjqjTIkMgKpVqxY6kpSUtHPnTg8PD6lUWqauNMXDhw/FlY9bt25dauMhQ4a8fPly3rx533///Z07dwAYGhrWq1fv/Pnzv//++927d+Vyef369YcNG+bi4vLmGjTvufj4+LS0NGNjYyYyREREREREGuf9CmVyc3OPHTvm5eXl7+8PQCqVuri4TJkypZL39Bk3btyhQ4f8/Pz27NljYWFRmbeuHImJieKLsj5dXFwcAAMDg/79+/v6+uYfv3bt2t69ewcMGODt7f3mHljvM+6HTUREREREpLnezZEab0pOTvb09LSxsenXr5+/v7+hoaGHh0d4ePjRo0crf5flSZMm1axZ09/f39bWVlxC5R2TlZUlvlAkQLl58+aDBw+mTp0KICUlBUBERMSxY8fGjBlz4sSJkJCQs2fPjhgxAsDBgwfnzp1bkYVrHm69REREREREpLne/ZEyISEhGzdu/PXXX9PS0gA0btz466+//uKLL6pVq6aqkrp163br1i03N7dTp0717t178uTJK1eu1NbWVlU9Spe/j7W451TJbGxs8l/LZDIARkZGZ8+edXBwEA82bty4a9euAPbs2bNx48Z58+YV3Cf7PceRMkRERERERJrrnR0pI5fL/f39+/bt26xZMy8vr4yMjB49ehw5cuT+/ftTpkxRYSIjMjc3P3HixLJly6RSqZeXV8eOHR8/fqzakpQof2elsLCwMl0oLoyio6OTn8jkmzBhAoC0tDRxp20ScaQMERERERGR5qqkkTIpKSk3b96Mj4+XSqWWlpaOjo6lDgyJi4sLCQmJjo7W1dW1t7cveUekgl6+fLl3797Vq1eHhIQAMDAwGDp06NSpU5s1a1bex1AqiUQye/bstm3bDh8+/Nq1a23atNmxY0efPn1UXZcS1K1b19zcPDY29vLly8VtiV0kcdRMbGxsenp6oUWX83OHhIQEJZaq/iIjI2/duhUfH6+trV2vXj1nZ+eCfztFhjLR0dHXr1+Pi4szMTFxdnYucvdxIiIiIiIiUj2hgl27du2TTz7R0vpP+mNsbPx///d/r169KvKSsLAwV1fXQqlNjx49QkNDS77Xw4cPZ8+ebWRkJF7SoEGDZcuWJSYmVsBjKVNcXFzv3r0BSCQSDw+PzMxMVVekBKNHjwZQrVq12NhYxa86d+6c+LO7cuVKcacuX76s1ErVl6+v75vbVxkbGy9atCgjI0Nsc+nSJS8vr7CwMPHb+Pj4oUOHFvxzk0qlrq6uMTExqnsOIiIiIiIiKppEEAQlRjyFbNy40cPDIzc3F0Dz5s0bN24sCMK9e/dCQ0MBODo6njlzxtDQsOAl169f79WrV3x8vI6OTufOnWvWrBkeHn7p0iVBEMzNzf/99983l88QBOHMmTObN28+ePCgeK8OHTpMmTJlwIAB4hol6k8QBC8vr1mzZmVlZTk5Oe3du7dBgwaqLqpcAgICnJ2d5XJ5//79Dxw4UEXffz0AACAASURBVCiVK05OTo6VlVVcXNzo0aO3b99e8NSQIUP27dtXo0aNiIgIRXYu12iCIMyYMWP16tUAZDJZ27Zt69evn5WVFRAQIM4I69q1q5+fX6H3ISEhwdnZOTw8HEC7du0aNWr0+PHjS5cuyeVyW1vbS5cucS0eIiIiIiIi9VJxec+xY8ckEgmAli1b/vvvvwVP+fn5WVpa1q1b9969ewWPZ2dn29raAujQoUNUVFTBrsR4xc3NrWD7lJSUTZs2tWjRQnwWHR0dNze327dvV9xDVah//vlHnL9TvXr1vXv3qrqcMpPL5ampqfnfTp8+Xfy5dO/ePSAgIP94dna2r69vnz59goODBUFo0aJFnTp1Vq1aJZ796aefAEgkkh9//DE9PV0QhJSUlBkzZohdLViwoFIfSUW8vLzy37pHjx4VPOXj42NoaOjo6Pj8+fNCVw0fPhxAlSpVjh49mn/w7NmzVatWBfDNN99UQuVERERERESkuIoKZXJycqytrQHUqVMnLi7uzQZBQUFvTqnIyMjw9PRs27ZtfHx8oVOjRo0S04qcnBxBEMLDw2fPnl2jRg3xs2vNmjUXLFjw5lVvR8wCVCI5OXnw4MHiQ7m5uaWlpamqkjJJT0/fuXNny5Yt3d3d8w/m5OSIq/OKLCwsnJ2dGzdunD8xbc6cOYIgiMM35s2bJ16VlZXl4uIiNqhWrZqNjU3+eJC+fftmZ2er5gkrUWJiorgQtbOzc/40pYJu37795i9GTEyM+MZ+++23hU599913AAwMDDTl14mIiIiIiOg9UVGhjJ+fn/hBevfu3Urp8JdffhE7FEfQdOrUSfz2ww8/3L9/v1I+q+fm5h45cqRHjx6DBw8uf2/lsWnTJh0dHQAODg4PHjxQbTEle/z48cyZM42NjcUfR5MmTcTULN/Zs2f79OkjDtYQyWSyVq1aLVq0SAzRZs+e7eHhcezYsfxLsrOzV61a1bBhw/xLmjRpsm7dutzc3Mp+PFXIHyZTptVz/P39xatu3rxZ6NS9e/fEU35+fkqtlIiIiIiIiMqlonZfOnPmDICqVasOHDiw5JbJyckvXryQyWQl76+kq6srvhDnMU2bNs3a2trDw8PR0bH81cbHx2/ZsmXjxo0REREAzMzMXr16pcJts8ePH+/k5PT555/fvHnT0dHx559/FmemqJWAgABPT09vb++cnBwAjo6OHh4eQ4cOLbSOT9euXbt27ZqVlRUTExMXF1e9evVatWoVXAxl2bJlhXrW0tKaPn369OnT4+LikpOTzczM8hdvfh+cP38egI2NTfv27UtuGRUV9eLFCz09PWtr68TERPFgfkCWr1GjRjKZLDc39/bt2+KS0kRERERERKQOpBXU761btwDY2tqWuibrsmXLrK2tHRwcSm4WHBwMwNjY2NTUFED//v137txZ/kQmJCRkypQp9erVmzNnTkRERKNGjdauXfvo0SMVJjKi1q1b37hxY+jQoSkpKSNGjBg5cuSrV69UW5IoKyvrwIED7dq1c3Jy2r17t1QqHTx48OXLl69fvz5y5MjidjqvUqVKnTp1Wrdu3bBhQ8WX6TUzM2vUqFGhRGb37t1ZWVnlfQw1FhgYCMDZ2bnUljNmzGjevLmYe1pZWYkHxZ3gC9LW1hbPipkjERERERERqYmKCmXi4+MBlDz4RXHZ2dm//fYbgD59+kilSqhZLpf7+/v37du3WbNmXl5eGRkZPXr0OHLkiJjRqDyRERkYGPz+++87d+6sWrXq7t27nZyc7ty5o8J6YmJili9fbmNj4+rqevXqVXNz89mzZ4eFhe3fv7/UMR3KsmXLlpEjR7Zv317cwOudJP7t1KpVq0xXOTs7m5ubA/j+++8zMzPzj8vl8k2bNsXExAB4+fKlUislIiIiIiKicqmo6UtpaWkAFBkTMWLEiDZt2ohLqBRnw4YNERERUqk0fxeet5aSkuLt7b1mzZr79+8D0NfXHzZs2JQpU5o3b17OnivIyJEj7e3tP//88/v377dt23b9+vVjxoyp5BoCAgI2b968a9eujIwMAA4ODhMmTHBzc6v8rakdHBxsbGwCAgIcHR03bdo0ZMiQSi6gEqSnp6PAfL0SzJ49e8SIEeKm8tra2kuWLBk/fvzly5cdHBzc3Nxq1qz5+PFjHx+fu3fviu01ZYd4IiIiIiKi90RFhTLiYBMxmilZy5YtW7ZsWUKDf//999tvvwUwffp0Ozu7ty4pLCxsy5YtmzZtSk5OBmBjYzN+/Pjx48e/uQaHumnVqtWNGzemTp26efPmoKCgSrtvbm7usWPHvLy8xEVkpVKpi4vLlClTevToUWk1FOLk5HTz5s3x48fv27dv6NChx44d++WXXwquIvwO0NfXT0xMVORvx87OruBfxJdffpmcnDx37tzg4OA5c+bk9/bTTz9t3Ljx8ePHYnxDREREREREakIiCEJF9NuzZ09/f/82bdr8888/5eknODi4W7du0dHRn3zyyeHDh0seUFOcS5cueXl5HTx4MDc3F0CHDh2mTJkyYMAAjRs44Ovr27t37+LWbVGi5OTknTt3rl69+unTpwAMDQ1HjRo1ffr0evXqVfStFbRr164JEyakp6c3b958//79LVq0UHVFStO0adOQkJABAwb4+Pi8xeVRUVE+Pj6hoaFSqdTW1rZfv37Gxsb6+voZGRkrVqwo/1gzIiIiIiIiUpaKCmVmzZq1YsUKPT29+Pj4tx7IcP369d69e8fFxfXs2fPo0aNlTWQyMjL279+/YsUKcfqGjo6Oq6vrjBkzWrVq9Xb1qJucnJzIyMiEhAQ9Pb26desqZSmc0NDQZcuW/fbbb+JMpebNm3t4eLi5uanhaJSbN29+/vnnDx8+1NfX//nnn0eMGKHqipRj2LBh3t7etWvXfvr0qUQiKX+HgYGB9vb2APz9/bt3717+DjVeSgqePkVCAgwNYWUFM7OyXR4TA1NTFBfppqfjyRPExcHYGM2a/adZTg5SUwHA0BBF/mQFAS9eAIC+PrQqahgjERERERGpj4pa6Ff87Jeenv7HH3+8XQ/79u3r1KlTXFzcW4yRiYqKWrhwYe3atUeNGnX37t2aNWsuWLAgIiJi165d70Yic/PmTVdXV1NTU2tra0dHx+bNmxsZGfXs2fPChQsK9nDt2jUtLS2JRCKRSD744IP84y9evNi6dWtWVpa48vHdu3fd3d3VMJEB4ODgEBAQMHz48NTUVDc3N/XZoKqcxNlhERERZ8+eVUqH4t+goaFhpa3HrL78/NC1K0xM0LIlOneGvT3MzWFnhw0bUOqWXnI5Dh9G586wtISBAZ4+LdzgwQO4ueVlMZ06wdYWpqaYMwf56y6fOgVjYxgb5yUvb0pIyGugpB89ERERERGpO6Fi5ObmNmjQAEDt2rXj4uLKdG12dvbs2bPFvGDWrFk5OTlluvzvv//On5fUvn37ffv2ZWdnl6kHNbds2bL8ARSmpqZ2dnaNGzcWH1kikSxevLjUHjIyMsSFjcX1dJydnQueXbFixcOHDyusfOXbuXOnOEqoadOmgYGBqi6nvFJTU2vUqAGgZcuWKSkp5ewtMjJSXErm66+/Vkp5miorSxg1SgAEQJDJBHt7oUcPoV07oWrVvINOTkJ0dNHXZmQIO3cKzZrltRS/QkP/08bfXzA0FABBS0twchK6dBEsLfNa9ughiP8E+fnlHUlKKvpGcXF5DU6eVOrDExERERGRmqqoUEYQhFOnTonZQYsWLf7555+Cp9LT03fv3r1z505BEHbs2NGnT59hw4aJp549e9axY0cAurq6YoOyysnJadq06eDBgy9fvlz+p1A3v/zyixjHODg4XLhwQS6Xi8ejo6MnT54skUgaNmyYmJhYcifz588H0LBhQ3GFkUKhjCYKCgqytbUVf23Wrl2r6nLKa/fu3eJP2cnJ6datWwVPxcfHr169+uDBg4Ig/PDDD926dZswYUL+2Xv37uX/SgiCEBAQ0LRpUzG8K2s2+q6ZODEv7/jqKyE29vXxtDRh1SpBR0cAhNathczMIq7t1k0ABKlUaNdOmDSpiFDmxQvBwkIAhJ49hYiIvIPZ2cK0aXmNN20SBIYyRERERERUWAWGMoIg/Pzzz/mDVmxtbYcPHz569OjOnTuL/3Vfv359uVw+e/Zs8UOjIAgPHz40MTER21tYWDgWxdvbu9T7ZmVlVehzqUpsbKw4JKR169YvX758s8Hvv/+ekJBQcifBwcHiXLBTp07Nmzfv3QhlBEFIS0vz8PAQf3kGDhyYVNznXvUTGho6e/bsDz/8sGCe8uOPP+aPh2rZsuWgQYNcXV0dHBzEn12HDh0EQRg6dKj4yyBecu3aNalUWq9ePRcXlwEDBogplTge6tq1a6p5NjVx6ZIgkQiA8N13RTc4fDivwdKlRZw9elTYvFl4/lwQBOHevSJCmchIYfRowd5eSE//z4XZ2UKTJgIgdOkiCAxliIiIiIiosIpdS3LChAkODg7ffffduXPn7ty5c+fOnfxTrVq1cnd3l8vlDRo06NKli5GREQBx2VqxQUxMTExMzJt9xsbGlnrfStifSCW2bdsmLpuyadMmAwODNxuIn9JLIJfLx40bl5mZOWLEiJ49e/71118VUqgq6OnpeXp6duzY8YsvvvDx8QkICNi7d2/B5XLUjSAI/v7+np6ex48fl8vlAG7cuOHo6Cie/b//+78PP/xw4cKFFy9evHv3rrhYNQCZTNatW7fJkycDcHBwePHiRaNGjcRTCQkJtra2gYGBT548EY/o6+sPGjRo8eLFtWvXruzHUyuenhAENGiA+fOLbvDppxg0CAcOYP16zJhReJFdF5dS+reywvbtkMsh/e8qXVpa6NQJISEICSlH9URERERE9M6qqN2XCklKSrp582ZMTIxEIjExMbG1tbW0tHyzWUZGRlRUVMldmZiYiANt3kPdunU7d+6cvb39zZs3366HtWvXTps2zdjY+P79++bm5vPnz1+8eLGzs/O1a9eUW6oKPXjw4PPPP79165aOjs7y5cs9PDyUsoeREmVmZu7bt0/BfcGSk5Pv3r0bHx8vk8ksLCyaN2+ur69fQudPnjx58uRJamqqpaVly5Ytq1SpUlGPoSlyc2FkhNRUzJ+PRYuKbXbqFD7+GAD++Qdt2iAnBwBkssLbJAUFQdx/PTQUDRqUfvcZM7BqFSwsEB2NY8fQpw8AJCXByKiIxvHxeVtBnTyJjz5S6OmIiIiIiEiTVdKuq8bGxt26dSu1ma6uro2NTSXUo6ECAgIAKLKHTvfu3S9evNivXz8fH5/8g0+ePBHnKy1fvtzc3Lzi6lStxo0bX716ddasWV5eXlOnTj137ty2bdvEpXNVLioqavPmzevXrxdHhFlaWrq7u0+aNMnU1LS4S4yMjD788EPFb1GvXr169eoVeer27ds2NjYlZzrvoODgvI2o27YtqVnbtpBIIAj491/IZBCHLJ07hy5dynX3hw8BgP+sERERERFRUSoplHkLo0ePvnPnzoYNG9qW/FHqvZGRkfHy5UsA1tbWpTbO+Z+CB93d3VNTUz/44INx48ZVUJFqQkdHx9PTs2vXrmPHjvX19XVwcPj99987dOigwpICAgI8PT29vb3FH4qjo6OHh8fQoUMrbapdbGzsJ598Ur169X379tnZ2VXOTdXC8+d5L+rUKalZ9eowMkJSEqKjlXbruDicPg0AvXr95/iwYYVnSIlK3ZabiIiIiIjeLeobyty6dSswMJCTL/KlpKSILxQZ6eDu7t6rV6/GjRvnH9m1a9fJkye1tLQ2bdokLbTyxTuqf//+rVu3HjJkyJUrV7p06TJ37tz58+dX8rNnZWX5+vquXr366tWrAKpUqTJ48OCpU6cqMtxJuV6+fGliYnL37t127dqtWbPG3d29kgtQGXGYDIBq1UppaWCApCSkpKBOHaxdCwD/W6/nLS1bhvR0VKuGQu/28ePl6paIiIiIiN4V6hvKPHv2DECdkv9z+31StWpV8UVGRkapjYcNG1bw2/j4eHH361mzZr1XoyTq1q17/vz5b7/9du3atTt27Jg2bVqlLUgUExOzY8eOdevWRUZGAjA3Nx8zZszEiRNV9SvdsGHDf//9d/bs2V5eXhMmTDh58uTWrVuNjY1VUkylys9i0tNLafnqFQAYGMDMDFOmlPe+x45hzRoA+PFHFJotePhw0QnRy5cYOLC89yUiIiIiIs2hpqHMq1evEhMTdXV1S1hr431TrVo1HR2dzMzM5/nTMRQ2efLkuLg4a2vrOXPmVERt6qxKlSqrV6/u2rWrmZlZ5SQyN27c2LRp065du8T4zN7e/quvvnJzc9PT06uEu5dAV1fX09OzU6dOX3zxxaFDh27cuLF37953f3pg/r8hz5+jefNim6WnIzkZQN5Su+X0998YOhSCgNGjMXly4bOdOxe70C8REREREb1P1DSUefr0KYC6deuq2745qmVra3v9+vUbN26U6aoTJ07s3bsXgL6+/pgxYwqeunfvHoDQ0FBXV1dzc/P169crsVq10rdv3yKPJycnV69eXfE5TdnZ2ampqQYGBlpvrAkil8v9/Py8vLz8/f0BSKVSFxeXKVOm9OjRozyVK93AgQMdHR2HDBly7dq1jh07LlmyZNasWe/mH1puLmQytGgBHR1kZiIgAN27F9v45k3k5gJA69blve+FC+jbFykpcHXFr78W3r+JiIiIiIjof9R0bRHOXSpSp06dAFy4cCG6LGuR5u+ffffu3QP/FRQUBCAxMfHAgQPHjh2riJrV0MOHDydPnty0aVMtLS1jY2Ntbe0mTZp8++238cWPU0hLS1uyZEmLFi10dXVr1Kiho6Nja2u7fPnyglPJPDw8+vXr5+/vb2xsPHPmzPDw8KNHj6pbIiOytra+cOGCh4dHbm7ut99+279//8TERFUXpTxyOY4eRc+e+P57ANDRgbh91YEDJV21dy8AGBvn7bv01rZtw0cfISUFAwdizx7IZOXqjYiIiIiI3mkSQRBUXUMRfv311y+//HLMmDHbtm1TdS1qJDg4uEWLFoIgjBw5cufOnQpedf/+/Tt37hR5at++fT4+Pg0bNvzxxx+rVavWu3dv5RWrpn766ae5c+fm70tVvXp1cU8rACYmJgcPHhSTr4ISExO7desWGBgIwMrKqlatWhEREeIkMicnpzNnzlSvXh3AjRs3hgwZMnHixC+++KJaqWvKqgdfX9+xY8cmJibWqVPH29tbtRtUKUFiIrZswcaNePoUABo3RkgIAPj4YNAgAPD2xpAhRVwYEgJ7e2RkYMYMrFhR0i2CgtCiBQCEhqJBg/+cysrCrFnw9IRUigULMG9e4TEyx46hTx8ASEoqdvqSOHnq5El89JEiT0xERERERJpNUEvz588HsGDBAlUXonYmTJgg/uAmT5786tWrgqeys7N9fX3v3bsnCML69eunTp26devWknubN28eAGdn5wqsWJ2sXLlSfPdatGjh4+MjvoFZWVknTpxwdHQEYGdnl5OTU+iqSZMmAdDW1j5w4IB4RC6X79mzRyaTAZg+fXplP4ZSPXnyRNwKSktLa8GCBbm5uaqu6K2EhAgeHkK1agIgAELDhsLatUJKSt5ZuVzo2VMABD09Yd++wtcGBAj16gmAULu2kJQkCILw9Kng7i64uwshIYUb37uXd4vQ0P8cf/JEaN9eAAQDA+HgwaKL9PPLu1a8y5vi4vIanDxZlocnIiIiIiJNpaYjZcaOHbt9+/Zff/113Lhxqq5FvaSlpfXv3//06dMAzM3N+/TpY2Njk5OT8/DhwwsXLkRGRk6YMOHnn3/u3LnzxYsX+/Xr5+vrW0Jv8+fPX7x4sbOz87Vr1yrrCVTmwYMHtra2WVlZ7dq1O336dKHBLJmZmQsXLpwxY4aJiUmhC2vVqhUVFTV69Ojt27cXPD506NC9e/daW1s/evSowquvSDk5OUuWLFm8eLFcLu/evfuePXssLS1VXZRi5HKcPQtPT/j5QRAgkaB7d3h4wMWl8CiV2Fh88gnEqXwODujZE+bmSEnBlSs4cwa5uTAzw7FjcHICgBs38iYxnTuHLl0AwM0NmZkA8PIlTp4EgF69IG5O368f6teHi0veOsGtWxceQSNatAiPHnGkDBERERERFaTWC/1yTZk3Va1a1c/Pb+XKlT/99FNsbGyhmKBu3br29vYAzM3Na9Wqxb2rCtqwYUNWVpa2tvauXbvenF6ko6OzdOnSIi9MSkoCUK9evULHGzVqBCBZ/CiuybS0tBYuXPjhhx+OGDHizJkzdnZ2u3fv/ki9Q4H0ly/1duzAunUIDQUAfX2MHAkPDzRpUvQF5ua4cAGLF2P9ety8if8ttAQAUilcXbF8Oayt847o6qJRIwD43z70OHQob8PsfMeP572oXx8JCcj/NbhxA0UuxT1pUpkfkoiIiIiI3nVqOlKmSZMmDx48CA4Obtq0qaprUVNZWVmXLl0KCgqKi4vT1dW1tLR0cHCws7Mr0zY69+7dCwoKqlGjRvcSdqV5V9jY2Dx69MjFxeXo0aNlurBp06YhISFt27a9fPlywX2aPvvss8OHD3/44Yd//fWXsotVjYiIiOHDh1+8eFEqlc6cOfOHH36Qqd86tWFhYVu2bNm8eXOUqanuw4eoXx/u7vjyS9SoodD1r17h/Hk8eID4eBgaol49dO0Kc/NSrjp3Lm9vpjfVqwdd3bzFa0rg5ITc3Ly8pmtXvLF7FwBkZ+P8eQBwdFT0cYiIiIiISJOpaShTrVq1tLS0lJQUfXGCAFH5xMfHm5mZAVixYsWMGTNKbtypU6f4+Phx48Z98803AJYsWSIuvuPi4uLp6WljYwPgr7/+6tatmyAIfn5+H3/8ccU/QSXJzc1dvHjxkiVLcnNzu3Tp8ttvv1lZWam6KAAQBOHMmTOenp7Hjh2Ty+UADo0b179PH/Trxx2OiIiIiIhIQ6njlthxcXFpaWk1atRgIkPKEhUVJb4Q5xyVLCQkJDg4OCYmRvx29uzZffv2BfDnn382atSoW7duP/zwQ58+ffT09Hbt2vUuJTIAZDLZwoULT58+XbNmzfPnz9vb2x/Pn6ejIpmZmbt27bKzs+vZs+eff/6pra3t5uZ269at/r/+is8+YyJDRERERESaSx3XlHn27BmAunXrqroQenfk73ttYGBQauPhw4e/ePHC2dlZ/FZbW3vNmjXXrl2LjY2Vy+Xnzp07d+4cgIEDB/bs2bPialahrl27BgYGjhw58sSJE3369Jk8efLKlSu1tbUruYznz59v2rRp/fr1CQkJACwtLd3d3SdOnCgOeiIiIiIiItJ06jhShqv8ktLp6OiIL7KyskptvHr16q1btw4aNEj89ty5c3Z2dnFxcbNnz7527do333xTs2ZNAD4+Ps2bN//7778rrmwVMjMz8/PzW7JkiUwm8/Ly8vf3r8y7BwUFubq61q1bd9GiRQkJCW3btvX29n769OnChQuZyBARERER0TtDHUMZjpQpp88++6xFixbXr19XdSFqxOh/OxDHx8eX6cKEhITPPvvs1atXM2fOXLZsmbOz88qVK589e+bj42NlZRUfH9+3b9/ExMQKKFn1pFLp3Llzz549O3fu3F69elXmrTMzMw8cOCCRSAYPHnzp0qUrV64MGTKk8ofqEBERERERVSj1DWU4UuatBQUFBQUFvbnr8/vM2tpaT08PwK1bt8p04e7du1+8eKGrqzt37tz8gzKZbMCAASdOnJBKpYmJib/99puSy1UnHTt2XLJkSSXf1MHB4Zdffnn8+PH+/fs7dOhQyXcnIiIiIiKqHOoYynD6UjlFRESAb+B/aWtri2vEHD58uEw7jj148ABA7dq1q1evXuiUra1t7dq1AYSGhiqvUjUiCIKvr+/YsWPt7Oxq1apVp06d9u3bL1iwQIxNSxAcHDx+/PguXbqsWLGiyAbXr18fN25cixYtLCwsmjdvPmrUqDdngbm7u6vJxk9EREREREQVhAv9vmvi4+O5d1WRRo8efeHChbCwsB07dowZM0bBq6pWrQogMjIyJydHS+s/fy/Z2dniDk3GxsZKr1blnjx5MmjQoEKT4CIiIq5cubJ69er169ePGjXqzasuXbrk5eV18ODB3NxcAIIgzJw5s1CbVatWzZ49W2wAIDY2Njg4eM+ePQsXLhS3HiciIiIiInpPcKSMJlmxYoWjo+PevXtLaMN3rzhDhw5t1aoVgEmTJh05cuTNBn///XdKSgqAHTt2bNy48cqVKwC6du0KID09feXKlYXar127NjMzE8BHH31U0cVXsqioqPbt21+/fl0qlY4bN+7MmTPPnj17/PjxoUOHunbtmpqaOnbs2MDAwIKXPHv2zN7evmPHjgcOHNDT0xPHEL3p+PHjM2fOzM3N7dq167Vr1168eHHjxo3evXvL5fL58+cfOHCgUp6PiIiIiIhILUjKNJWjEuTk5Ojq6gJIT0/nup6FjBo1ateuXdu2bSthoIevr2///v379u1bZO7wngsKCurevXt0dDSAbt269enTx9ra+sWLF6GhoUeOHLl79+7WrVvHjh1rYWERGxs7c+bMn376SRCE7t27nzt3TiqVfv7558OHD69bt+7z58/37du3Y8cOuVzu6uq6b98+VT+ZkvXp0+fYsWPa2to+Pj59+/YteEoQhG+//dbW1nbEiBEFj+fk5Dg7Ozs5OX366afdu3efP3/+ihUrOnXqdOHChYLX2tra3rt3z97e/urVq/lbYuXk5HTu3Pnvv/+2sbF58OCBTCarhGckIiIiIiJSObWbvhQZGZmbm1u7dm0mMm9SZGIXR8qUoHnz5leuXJk0aZKfn9/Zs2fPnj1b8KyhoeGbl0gkksOHD3/55Zf79+/39vb29vbOPyWVSidMmLB27doKr7ty3blz59ixYwCmMRucDwAAHxhJREFUTZtWKJEBIJFIli9f/uZVWlpaN27cKLnn69ev37t3D8CiRYvyExnx2u+//75Hjx7h4eF//fVXly5dyvkIREREREREGkHtQhluvVQCRQIXvoEls7a2/vPPP4ODg0+ePBkWFhYXF2dgYFCnTh1HR8fu3buLo7SePn0qCEL+CjLVq1fft2/fwoUL//zzz6CgoFevXhkYGDRr1uzTTz9t1KiRSp+mQohziGQy2dSpU5Xb8+nTpwHo6em9OeGrS5cuRkZGycnJp06dYihDRERERETvCbULZcTcgav8vkkQhMjISADFrdYh4jLJimjWrFmzZs2KO1twEIeCl7xL/vnnHwAtW7asWbNmyS03b97s5+dnYWGxefNmRXoWh8nUr19fDL8KkslkTZo0uXbtmtiGiIiIiIjofaB2oQwHehQnNjY2IyPDzMxM3A+oOJy+ROUk/gopkkDdvn37yJEj1tbWCvYcFRUFwMLCosizZmZm+W2IiIiIiIjeBwxlNIaCaQtHylA5vXjxAsWssFNIw4YNO3bsaGVlpWDPr169AmBqalrkWXNzcwDiBlhERERERETvA7ULZTh9qTiKpC25ubnPnz+XyWSKf04mKkTc/Egul5facurUqWVad0YqlQLIyckp8mx2dnb+3YmIiIiIiN4HUlUXUBhHyhRHkZEykZGROTk5lpaW3LuK3po4RiYxMVHpPevr6wOIi4sr8qx43MDAQOn3JSIiIiIiUk9qF8pwpExxFImr+O5R+TVu3BjAnTt3lN6zuPpMbGxskWfFUEbxFWqIiIiIiIg0nXqFMmlpaYmJibq6usWtOvE+UyRw4TgjKr8PPvgAwIMHDx48eKDcnlu1agUgPDz85cuXhU5lZGQEBwfntyEiIiIiInofqFcok587SCQSVdeidjhShirH0KFDxYVdli5dqtyee/fuDSAnJ+fQoUOFTvn5+aWlpQFwcXFR7k2JiIiIiIjUljqGMhzoUSSOlKHKUadOnTFjxgDYuXOnp6fnmw0yMzOfP38OICQk5PTp05cuXVKw54YNG3br1g3AwoULC65Zk5KS8t133wFo164dR8oQEREREdH7Q712X+J2zsXJysqKiYnR0tKytLQsoRlDGVKK1atX//PPP7dv3546derhw4dHjhzZokULHR2dhw8fXrlyZdeuXd26ddu3b9+6des2bNhgbW396NEjAGFhYWfOnBF7CAwMBPD8+fPNmzeLR1xdXY2MjFauXNmhQ4fHjx+3a9fuu+++a9as2cOHD5cuXXr//n0dHZ01a9ao6pGJiIiIiIgqnzqGMswU3hQZGSmXy2vXrq2lVdKPjNOXSCkMDAwuXrw4fvz4AwcOnD9//vz584UaPH78WBCEQgevX7/u7u5e8MjDhw/zj3Ts2NHIyMjBwWH//v0jR4588ODByJEj81tWr159+/bt4nI2RERERERE7wn1CmU4fak4CqYtTLVIWQwNDfft2zdnzpzDhw8HBATEx8cDMDc3t7W1/fjjj9u3by+RSObMmePu7l6lShXxkgYNGowfP764Do2MjMQXLi4uwcHBu3fvvnTpUmJiorGxcbt27UaOHGllZVUJz0VERERERKQ+1CuU4fSl4iiStqSlpSUkJOjq6pqZmVVWXfSOs7Ozs7OzK/JUWFiYhYVFwSTFycnJyclJkW4tLCxmzJgxY8YM5VRJRERERESkmbjQr2ZQZKRM/rvHvauoEjg7OxsYGMTFxam6ECIiIiIiIk2lXqFMREQEGMoURZGRMhxnRJUmLS0t8f/bu/e4KMu8j+O/ARSQo3JW0zybuBLIuia74QF9mY8asgloSrqukYcgO2hlm7jFqqtto+uuh9wIEXMTn9Q8FQimLGakYHgohRRLRAQRUVAOM88fY6wP4jDmMDcDn/dfwz3Xdc/vNvL1mq/X9buuXrWxsXF1ddUz7OWXXw4NDT19+rTJCgMAAAAAM9KMQpni4uKKiooOHTrY29srXUuzY0gowzojmEx+fr6IdOnSRf+yrH379m3durW2ttZUdQEAAACAOWlGPWWsrKyWLl3K97cGGbJ9iZUyMBkDW0qz9g0AAAAA9GhGoYyzs/OCBQuUrqKZMnz7El+AYQKGpISlpaXl5eWOjo5OTk6mqgsAAAAAzInxQ5nLly/v378/Ly/v+vXrDg4OvXv3fuqppxr9VlZZWbl161YbG5vAwEAPDw+jV2XWysvLr1271q5dOxcXFz3D2L4EkzF8Px1LtwAAAADgfowZyhQWFi5YsGDTpk0ajebu6zY2Nm+++eYbb7xhZdXAxxUVFcXFxa1ataqgoEBE1qxZ88ILLxixqhbAwC+3bF+CydB5GgAAAAAentFCmdzc3KCgIF37zwEDBowaNcrV1fX69evJycmZmZlvv/32qVOnEhMTLSz+21r4u+++W7FixaZNm27fvm2sMlokA/cl6YZ17tzZFDWhdXugM9pNVBMAAAAAmBvjhDJVVVWhoaH5+flt27bdsGHD1KlT696KjY3dsmXLjBkzBg8efHciIyJpaWn/+te/evXqFRwcPH78+IkTJxYWFhqlnhbGyckpJCTEz89Pzxjd2VXt27d3cHAwWWFotWhyBAAAAAAPzzihTGJiYlZWloio1eq7Exmd8PDwoKAgV1fXetenTJkSGBjYr18/3Y/1IhvUeeKJJ7Zt26Z/DFtFYEqGHKvE7yQAAAAA6GecHOTDDz8UkR49ekRGRjY44N5ERkQcHBzqEhk8JLaKwGSuXLlSUVHRoUMHe3t7PcP4nQQAAAAA/YywUqaqqiozM1NEQkJCGl3tsmPHjqqqqscee6x///4P/9Gow0k3MBkDl8CwUgYAAAAA9DPCSpnc3Fxdp15fX99GB0+fPj00NPTjjz9++M+tM2/evOjo6OrqaiPe0+zQvwMmY8gSGI1Gc/HiRZVK1alTJ1PVBQAAAABmxggrZUpLS3UvPDw8Hv5uD+rMmTOrV6+uqanJzs5OTExstWcPEcrAZAxZAnPp0qXq6mpPT09ra2tT1QUAAAAAZsYIK2Wqqqp0L9q0adPo4JycnPz8/Pnz5z/85+r07t07JSWlU6dOBw8e9PX13b17t7HubF7YvgSTMfzoJX4hAQAAAEAPI4Qyjo6Ouhfl5eWNDu7UqVOXLl2cnJwe/nPrBAYGZmdnjxkzpri4eNy4cdHR0XU5UesxefLkF154oWfPnkoXgpbPkO1LdPkFAAAAgEYZIZTx8vLSvcjLy3v4u/0yrq6uu3btUqvVbdq0WbVqVUBAgILFKGLOnDlr1qyp+28BNB1DVsGwUgYAAAAAGmWEnjIdO3bs2LFjQUHBoUOHXnzxxYe/4S+jUqmio6OHDBkSHh7+zTff+Pn5rV+/PiwsTKl6TKaoqOjIkSNXrlwREQ8Pj4CAAGdn5/sNzs/PP3v2bINv9erVq2vXrk1VJVoQQ1bB0OQIAAAAABplhFBGRMaNG7du3brPPvvs0qVLRl+sUVZWZvh2p1//+tfHjh2bOXPm1q1bw8PDd+/evW7dOltbW+OW1Eykp6cvXLjw0KFDWq227mLbtm0nTZr017/+1d3d/d4pH3zwQWxsbIN3i4uLmzZtWhOVihajpqamsLDQ0tKyY8eOeoaxfQkAAAAAGmWE7UsiMmfOHEtLy1u3bs2YMcO4/VxOnjzZtWvXmJgYjUZj4BQnJ6dPPvlk3bp11tbWCQkJAQEB91sbYtb++c9/Dh069ODBgyIyaNCg5557btq0aYMHD66uro6Pj//Nb36jW6pQz6VLl0SkS5cuQfcYMGCAqZ8BZqi2sPDAb3+b+NRT+ht7s30JAAAAABplnFDmV7/61euvvy4ie/fuHT58eEZGRl2GcuvWraSkpFGjRuXk5IhI3759PTw8lixZYuCd9+zZU1ZWtnjx4gkTJly9etXwkp5//vmMjIyePXtmZWUNHDgwMTHxAZ+pWUtOTp47d25tbe2AAQOys7OPHDny0UcfxcXFHT58+OjRo3369BGRsrKyeycWFBSIyMyZM5Pv4efnZ+rHgBmyPn8+4MCBsJIS/cO61NY6Wlu32iPqAQAAAMAQxgllROTPf/7zK6+8olKp/vOf/wQEBLi7u/v4+HTr1s3Ozm7ixInJyckJCQkiUlRUVFRUdOPGDd2s4OBg1c90ecGsWbN0P+oygtdee2379u0dOnTYuXPn448/np6ebnhJfn5+x44dmzRpUnl5+ZQpUyIiIm7evGms51WQVqt98cUXtVqtl5fX/v37661w8fX13b9//9dff92/f/975+r+kPVvPAH00a3A0r8E5vbtbdnZZVqtl6enaYoCAAAAAHNktFDGwsJixYoVGRkZzzzzjLOzc0lJybfffnv+/HkrKyt/f/8lS5a8+eabIhIdHT1//vwnn3xSN8vOzq79fdSdtP30009nZWUFBAT8+OOPw4YNe6CtTA4ODps3b46Pj2/Xrl1CQoK/v79uwY5Z+/LLL7///nsRWbx4saur670DOnXq5Obm1uBc3fYlDmnCL3fhgoiI/mYxFy6IViudO6ssjPY3DAAAAAC0PKq7e8Qai0ajuXLlyvXr121sbDw9PfX3njBQTU3Nu++++84772g0muHDh2/atOmBkoXTp0+HhoaeOHHCxsZm6dKl0dHRD1+SUmJiYhYvXmxlZVVSUlIXXTWoqKiooKBApVL5+PiISHV1tY2NjUajyc7O1l0BHtjcufKPf8jKlRIVdd8xqakyYoQEBsqBA6YrDAAAAADMTZP8O7aFhYWHh0evXr0eeeQRoyQyImJlZRUTE/P55597enqmpqY+/vjjX3zxheHTH3vssa+//joqKurWrVsvvfTSxIkTG2y5Yha+/fZbEendu7f+REZE4uLifH19Bw4cqPuxsLBQt8jo5MmTr776anBw8Lhx4+bNm3f48OGmrhkthyHbl3SraejyCwAAAAB6mdnmgqCgoOzs7JEjRxYVFY0ePfr111+vra01cK6tre3KlSvj4+Pt7e2TkpIGDRqUnZ3dpNU2kZKSEvlFfWF0e5dE5Nlnn33vvfd27Nixa9cutVo9ZMiQZ5999tatW0YuFC2SIduXdMEN52EDAAAAgF5mFsqIiIeHx969exctWmRhYbFs2bIRI0ZcvHjR8OkRERGZmZkDBgw4c+bMwoULm67OplNZWSkiNjY2jY4MDw9PSUlJTk7W/divX7/Y2Njp06dv3br1xIkTeXl5ycnJwcHBIrJ58+a5c+c2adloIQxZBUMoAwAAAAAGaJKeMqaRlpY2ZcqUgoICV1fX+Pj4MWPGGD63srLy7bfffvnll82x5e3w4cPT0tKGDRuWmppqlBu+8cYbS5cutbCwOHv2bPfu3Y1yT7RMFRViZyc2NlJRISrVfYeNHi2ffy67d8uD/F8JAAAAAK2N+a2UqTNs2LDs7OzRo0cXFxePHTs2Ojq6urrawLm2trbLly83x0RGRNzd3UUkPz/fWDecP3++hYWFRqP5/PPPjXVPtEx1e5f0JDLCShkAAAAAMIgZhzIi4ubmtmfPHrVabWVltWrVqoCAgHPnzildlPFptdqUlJRp06bpUic/Pz8ROXfu3OXLl41y//bt2zs7O8tdTWeAhhnYwdeQZsAAAAAA0OqZdygjIiqVKjo6Oj09vVu3bpmZmb6+vlu3blW6KKO5ffv2xo0bfXx8Ro4cGR8fv23bNhEZOXKkiGi12o0bNxrlU8rLy0tLS+XnNTjAfRmyBKa0VMrLxcFBnJxMUxQAAAAAmCmzD2V0Bg0alJmZOXbs2LKystDQ0MjIyNu3bytd1EO5dOlSTExM586dn3vuuZycHE9Pz0WLFo0YMUJEfH19n3jiCRGJjY3Ny8sz/J5VVVUpKSn3Xt+9e7eutdDgwYONVD5aKEOWwOjGdO1qinoAAAAAwJy1kFBGRFxcXHbu3KlWq9u2bbt+/fohQ4bk5uYqXdQvcfTo0YiIiK5duy5evLi4uHjgwIHx8fEXLlyIiYlxc3PTjVmzZo21tXVZWdmTTz65c+fOu7s1l5eXr1+//r333hORxMTEgICAwMBAEamurg4LCxs1atS8efOuXLlSNz4pKWnWrFkiMmLECH9/f5M+KsyOIedhGzIGAAAAACBipXQBxqTbyhQQEBAeHn7s2DE/P79169ZNmjRJ6boMUlVVtWPHjvfff//w4cMi0qZNm4kTJ+oe597BPj4+SUlJYWFhBQUFTz/9tKenp5+fn7W19fnz57/77rvKysr27dvPmTPnp59+ysjIsLS0FBELCwsPDw+tVqtWq//+97/7+vo6OzufOXPmwoULItKnT5/Nmzeb+JFhfgzZvkRDGQAAAAAwTIsKZXT8/f2PHTsWGRm5ZcuWyZMn7927d+3ate3atVO6rvsqKiqKi4tbvXr1Tz/9JCJubm5/+MMf5syZ84jer75jx47NycmJiYnZvn17YWHhnj17dNctLCyGDBnyxz/+0crKqk+fPhMmTNCFMpaWlmvXrg0ODl66dGl6evo333yjG+/m5jZ9+vS3337bzs6uiR8U5s+QRr8cvQQAAAAAhlHdvfOlhdm4ceOsWbMqKir69ev373//u3///kpXVF9WVtbatWsTEhIqKytFxMfHZ/bs2VOmTHmgCKm6uvrUqVOFhYUajcbFxaVfv3729vb6p1y7du3777+/efOmh4dH3759dakN0Dg7O6mokPJy0fM7NmWKJCZKfLxERJiwMgAAAAAwPy05lBGRkydPhoWFnTx50tbWVq1WP//880pXJCKi0Wh27969atUqXdtdCwuLMWPGREdHjxgxQqVSKV0dcB/FxeLmJh06SEmJvmGBgXLwoKSlydChJioMAAAAAMxTy2n02yBvb++vvvpqypQplZWVkZGRERERN27cULCesrKylStXdu/effz48SkpKY6OjlFRUXl5eZ999llQUBCJDJo1Azv40ugXAAAAAAzTwkMZEbG3t09ISIiPj7ezs0tISPD39z9+/Ljpyzhz5kx0dHSnTp1eeuml/Pz8nj17qtXqixcvrly58tFHHzV9PcADM6SDr0YjBQWiUknnzqYpCgAAAADMVwts9NugiIgIf3//0NDQkydP7tu3z8fHxzSfq9FoUlNTV65cuXv3bq1Wq1KpgoKCoqKixo4dy7oYmJkxY+T8eamp0TemsFCqqsTTU6ytTVUWAAAAAJirFt5Tpp6Kior169dHRUVZWDT5EqHy8vKPP/5YrVafPn1aROzt7SdPnhwVFeXt7d3UHw0o5soV+cc/RKWSRYuULgUAAAAAmrvWFcrcT0lJSdu2bR0cHAyfUl5ertFonJyc7n0rLy/vgw8+WL9+fWlpqYh069YtMjJy5syZHTp0MFrFgOJKS6W0VFxdxdHxgedqtaJ/pVijAwAAAADA/LX8njL3U1ZW9u677w4aNMjKysrV1dXR0dHBweGZZ55JT0/XM6ugoEAXrzg6Ojo7Ozs5OUVERJw7d+7uMb///e+XLVtWWlo6bNiwTz/9NDc3d8GCBSQyaCFSUiQkRJydpUMH6dFDnJzE1VXCw+WrrwydPmaM2NiIn1/DA4qKJCZGPD3Fw0MSE41YOAAAAAA0N610pcyhQ4dCQkKKi4t1Pzo6OtbW1t68eVNEVCrVggULlixZcu+snJycoUOHXr16VUTc3NwsLS0LCwtFxN7ePjk5efDgwbphCQkJX375ZVRU1IABA0z0PIAJVFXJ9OmyefOdH728xNlZSkqkqOjOlblzRa0WS8uG5+7YIcuXS2bmnSvW1nLr1v8bc/asrF4tGzZIRcWdK8uXy6uvGv9BAAAAAKB5aI0rZY4fPz5q1Kji4mJnZ+f333+/qKiorKzsxo0bZ8+enT17tkqlUqvVJ06cqDerqqoqJCTk6tWrXl5eKSkpRUVFly5dOnz4cLdu3W7cuBESElJ32PbUqVM3bNhAIoOW5rnnZPNmUakkMlJ++EEKCuTUKbl8WU6ckIkTRURWr5aXXmpgokYj3t4SGiqZmfLYYzJ6dANjNmyQPn1k1SpRqSQkRFxcmvZZAAAAAKAZaHUrZbRarZ+fX3Z2tpOT08GDB++NTuLi4ry9vQcNGlTv+vr16yMjI0Xk4MGDv/vd7+qu5+Tk+Pr61tbWLlu2bP78+U1dP6CMpKQ7ycuKFfLKKw0MiI6+E6mkpsrQofXfXbpULCwkOFh6975zq3orZXJzZdkyefppCQoSGxvp21e+/56VMgAAAABatlYXyqSlpQ0fPlxE1Gp1dHS04ROHDRt24MCBwMDAAwcO1HtrwoQJ27dv9/Hxyc7ONmKpQDMyeLAcOSJDhkh6esMteG/fFm9vycuT//kf2bVL360aDGXqIZQBAAAA0Aq0uu1Lu3btEhFbW9vp06cbPqu6ujojI0NExowZc++7uovHjx+/du2akcoEmpOCAjlyRERk5sz7HopkbS0zZoiIfPHFf5vCAAAAAADur9WFMpmZmSLi6+vr2Ng5vosWLerRo8eIESNE5OzZs1VVVSLSvXv3e0fWXTx16pSRywWaA10iIyIBAfqG6bb1VVfLsWOybZt4eYmXl/zcThsAAAAAUI+V0gWYmu68pJ49ezY68sqVKz/88INKpRKRop/Pl3F3d793pJubW90UoxUKNB8FBXdedOumb1hdZHnxolRXS2GhiEhtbVNWBgAAAABmrNWFMtevXxcRBweHRkcGBgaqVCpd4FLx83YMl4YOhXF1ddW9KC8vN1qhQPNx/bqIiI2NWOn9G6Nu9VlZmfj73+kHbGfXxMUBAAAAgLlqdaFM27ZtRUS3F0m/sLCwsLAw3es2bdroXlQ01Cyj7qK1tbVxqgSaFRsbEZGqKtFq79tTRuS/jXvt7MTPT/z8TFEbAAAAAJitVtdTpn379nLXdiQD1TWgaXCDUt3FRvvUAGapQwcREY2mkQYxly/fedHQgjIAAAAAQD2tLpTx9vYWkQc9u7quB02Dac7ln7+LGtKqBjA/Pj53Xhw/rm9YVlb98QAAAACA+2t1oUxAQICI5OfnHz161PBZLi4uXbp0EZEjdcfQ3EV30dnZuZv+NqiAmerf/87il//9X33DkpJERPr2FS8vU1QFAAAAAGau1YUy4eHhtra2IvLWW29ptVrDJ44fP15EkpKS6rWVqamp2bJli26AhUWr+/NEq2BlJTNmiIh89JHk5TU85sgR+ewzEZEXXjBdYQAAAABgzlpdiODi4vLaa6+JyL59+2bNmlVZWVlvwA8//HD8+HERSU1N/dvf/hYXF6e7Pnv2bEtLy+Li4nnz5t2d5rz11lvnz59XqVRRUVGmegjA5ObPFy8vqayUMWMkN7f+u1lZEhIiGo306yeRkSIiJ05IbKzExsrNm6YvFgAAAADMguqBVou0DNXV1SEhIbt27RKRRx55JCwszNvb29LSMjc3NyMjIzU1ddSoUXv37p09e/aaNWt69OiR+/NX0IULF/7lL38RkcDAwNDQUEtLy+3bt+/bt09EoqKiVq5cqeBDAU0uPV3GjpWyMrGxkdBQ+e1vxdlZSkokNVU+/VRqasTLS5KTxdtbRGTTJpk6VUSksFA8PKSkRN588859zp2T5GSxtLyz+kZEpk2TJ56QDz+Uuu2Bn3wi167J4MEyYICIiLu7vPOOSR8WAAAAAJpeawxlRKSmpiY2NnbFihU3btyo95aVldXkyZM/+uijOXPm1AtltFrtn/70p2XLltXU1NSNt7CwiI6OXr58uaWlpekeAFDEiRMSFSVpaQ28FRwsarV07Xrnx48/lunTRUQuXBB3d8nPl0cfve9t4+Jk2jSZOlU2bWp4QK9ecubMw5UOAAAAAM1OKw1ldK5du5aSkpKVlVVYWGhlZeXu7t6/f/+goCCXnw/01Wq1KpWq3qwff/xxx44dubm5Wq22W7du48eP7969u8lrB5STkyNffCEXLsi1a9K+vfToIU89JfqPHrtxQzZvvu+7Q4dK796SliZnzzY8wMlJwsIeqmYAAAAAaH5adSgDAAAAAACglFbX6BcAAAAAAKA5IJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQAKEMAAAAAACAAghlAAAAAAAAFEAoAwAAAAAAoABCGQAAAAAAAAUQygAAAAAAACiAUAYAAAAAAEABhDIAAAAAAAAKIJQBAAAAAABQwP8B2Rd1GWv6hAAAAAFZelRYdHJka2l0UEtMIHJka2l0IDIwMjIuMDMuMgAAeJx7v2/tPQYg4GWAAEYg5ofiBkY2jwwgzczMKMGAYCKJMrF5aICYLBLMcBYLQpqVjQMqyMYBYTFJsLNxJABZTEwSHBAWUB0nXB0XXB03BwfYGEZGCR52sEJGJgledqggkwQfNwNjBhMjUwITcwYTM0sCC2sGEwubAhu7BhMbBwMHJwMnFwMXtwYTF48CJy8DM58CK0OCCCMbAysLMxOj+D6oV8GAX3O5zYHnIhwHQByv+y0Hluybux/EvrGv7sD+b812IPaG5WoHtsyNsgexuxQ27//RqbwXxGZROLl/yY3DYPVn+/bs2ZrJB1bzIVPCbvoPJgcQe7bNdHvbm2tAljKUblV2yMluAZv54EyBw9wFSmC9ymdTHQLZIG5gKVri0PdEHmy+0VY9B/3r98FmflkufWA6qxjYTDEATjNN/lrOtJ4AAAG0elRYdE1PTCByZGtpdCAyMDIyLjAzLjIAAHicfZNNbhwhEIX3fQouMKh+KWrpmbEiK3KPlExyh+x9f6VqxjR4Y2haNHwU8F71VrL8uv7891GOQtdtKwW+edy9/GUA2N5Ldsr59cfbXi73l/MYudz+7PffBTUfyPqVfbnf3scIlks5UXUxA84eUiNrBSo8ylxLSUpFJVIrJ6zExB0miYPkJLl6B/Ye88Gb0ARpgPLcvCUXJ61d2ZtPkAeoCWJld2EspwCiS8spZZDtSaohdX/0QitZYuogLckYJga0mFdtDMt12gB7ucUoSu9xCaogTYgnZ4PzCIgVOypbRmY2dZ1gPzQPiyKQqrnCQ5/eeNHHDzDd4WqNFDRDNlZqy954uIMUp+SqDk3a00flviiEhz3IgWpF7o79U0zgFT0MQil7nNQovPS4nEk3lYU8HIpM23NbYYoMijUsAo4Lelj0ul+/pOAzKc+3/TqTUqLRzDyJxjO9JNtMoqw6UyU+pnktAun0yCKQTSd6zNrUO66YDRaj8PHCxRHK1yIRJYOrFPy5ShYZH1fSRS3Jo62SrALk9/ijo7/9B+cWxvt/S5hCAAABXHpUWHRTTUlMRVMgcmRraXQgMjAyMi4wMy4yAAB4nG2Su25CMQyGX6UjSMHyNRejDhULS6E7OgOtuhWBULvx8HVOu+UscfzZ/m1LPn3sHSc6haGpvxyvy7Q6HfbspNM6XA1353laPZ+OXgLtIlanMN7mEOFfjCiCx5DisLOAdIG920RPj9WGoWkpKSxx5pq2GwUy5pw2BCwsHQm0ipoQ2EpRDsKQUVoNVE1abjNSYZLEIKrYCYG0ppI2CJHcOFvA+LMgFYlasyxpi0Baa8gyoOYuT0CVTHuhSLEWdRzJJT59hpolkgRKZus5WYy7kIA1jCbzMjFW2hqQ1Mb/E6D0LIbCrVIiKFqLzXNaIW59YxNES+t0/r5e3u7XmyNcrl8v4b2eb4efy/vnHdBphOQ8QnYZobgulKvbSNXzCLOXERavI6zeRtjiNBb6x1JLW1EczgLuZ7SAJc5zxPb4BdbjyK2m5CwXAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pubchempy.Compound\n",
    "plp = 'CC1=NC=C(C(=C1O)C)COP(=O)([O-])[O-]'\n",
    "m = Chem.MolFromSmiles(aa)\n",
    "m2 = Chem.MolFromSmiles(plp)\n",
    "Draw.MolsToGridImage([mol_with_atom_index(m)], subImgSize=(500,300))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec96a2a-861e-4f9e-a93f-19fca5d1b25b",
   "metadata": {},
   "source": [
    "## Choose the nuber of the Nitrogen to bond on PLP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "ef925eaa-5405-4218-a345-f4de01251e7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " 13\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABdwAAAH0CAIAAACo53h7AAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nOzdd3yN5//H8U/2IIQvMYMgk9hqxKpRo9XG1lr1rVJ7JQhVW4zQGD8NXWiqNWq0RlKxqVEkZoQKIkKEhEREcnLO/fvj+EZyRK3EHcnr+cgfznVd98n7pPF9PLy/133dRoqiCAAAAAAAAN4sY7UDAAAAAAAAFESUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMgAAAAAAACqglAEAAAAAAFABpQwAAAAAAIAKKGUAAAAAAABUQCkDAAAAAACgAkoZAAAAAAAAFVDKAAAAAAAAqIBSBgAAAAAAQAWUMijo0tPTU1JS1E4BAAAAAChwKGVQQGm1Wn9/fzc3NzMzM2tra3t7+xkzZqSlpWVeEx8fP2fOnAoVKnzwwQcZgydOnCielZeX1xuPDwAAAAB465mqHQBQR+/evX/99dcePXrMmDHD2tr6wIEDvr6+6enp06ZN0y+YP3/+tGnTbG1ttVptTExMxoXR0dEJCQkBAQHFixfXj7i5uanwAQAAAAAAbzkjRVHUzgC8aZs3b+7UqdO4cePmzp2bMRgWFubq6mphYaF/eejQIXNz83r16nl6el6/fv3kyZP68WXLlg0bNiwlJSVjJQAAAAAAr4CdMiiIfvrpJysrq8mTJ2cerFWrVuaXHh4e2V578+bN4sWL08gAAAAAAF4TZ8qgIAoLC3N0dCxcuPDTU0OHDjUxMfmXHWS3bt2ysLCYOnVqz549P/vss6CgoNxMCgAAAADItyhlUBAlJCSUK1cu26lGjRr179//X66tUqVKyZIl79y5Y29vf+7cufbt23/55Ze5ExMAAAAAkJ9xpgwKorJlyzo4OBw6dOhFFn/00UeZz5Qx0KtXr7Vr10ZGRlaoUCFHMwIAAAAA8jl2yqAgcnBwuHr1ao68VY8ePbRabWhoaI68GwAAAACg4KCUQUHUtm3bmJiYPXv2vP5b3bt3T0SKFCny+m8FAAAAAChQKGVQEA0ZMsTOzq5fv35///23fkSn0wUHB8fFxW3fvn369OnPuvDBgwcdOnRYv369TqcTkWvXrs2aNcve3v5Zj2oCAAAAAOBZOFMGBdSJEye6det25cqVqlWr2traRkZG3r9/f926dTt37gwICNBqtREREfqq5cGDBzqdTr8X5p9//hk6dOivv/5asmTJUqVKRURElC5detOmTXXr1lX7AwEAAAAA3jKUMii40tLSQkJCzpw5o9Vq7e3tW7RoYW9vf//+/QcPHpQrVy4+Pn7Dhg0Gl3z22WcmJibnz5/fu3dvUlKSo6Nju3btrK2tVckPAAAAAHirUcoAOSA2NjYwMHDs2LFqBwEAAAAAvDVM1Q4AvPW0Wm3jxo0jIyMrV67cqVMnteMAAAAAAN4OHPQLvC4TExMvLy8RGTVq1MOHD9WOAwAAAAB4O1DKADlg0KBB9erVi4qKmj9/vtpZAAAAAABvB86UAXLGX3/91aRJE0tLy3Pnzjk4OKgdBwAAAACQ17FTBsgZjRs37tWrV0pKiv5WJgAAAAAA/h07ZYAcExsb6+zsfP/+/R07drRr107tOAAAAACAPI2dMkCOKVWq1KRJk0RkzJgxGo1G7TgAAAAAgDyNUgbISSNHjnRxcQkPD1+8eLHaWQAAAAAAeRq3LwE5LCQkpE2bNjY2NhcuXChbtqzacQAAAAAAeRQ7ZYAc1rp1a09Pz6SkJB8fH7WzAAAAAADyLnbKADkvKirK1dU1JSVl3759TZs2VTsOAAAAACAvYqcMkPMqVKjg5eWlKMrIkSO1Wq3acQAAAAAAeRE7ZYBckZKS4ubmdvXq1YCAgEGDBqkdBwAAAACQ51DKALnlt99+69q1a/HixSMiIkqUKKF2HAAAAABA3sLtS0Bu6dKlS7t27eLj46dMmaJ2FgAAAABAnsNOGSAXhYeH16xZU6fTHT9+vFatWmrHAQAAAADkIeyUAXKRq6vr8OHDtVrtsGHDKEABAAAAAJmxUwbIXUlJSc7Ozjdv3gwMDOzVq5facQAAAAAAeQU7ZYDcZWNj4+vrKyJeXl6JiYlqxwEAAAAA5BWUMkCu69u3b6NGjRISEv766y+1swAAAAAA8gpuXwLehPDwcDMzs6pVq6od5PkURYyM1A4BAAAAAAUAO2WAnHf79m1/f//+/fv36tVr8uTJly5dcnV1NWhkbt68OW3atGXLlhlce/DgQS8vr549e44cOXLHjh06ne41w3z5pRgZyejRWQa//lqMjOT+/SyDMTHi4yP/+Y906PB4ZMMGMTLK5uv06dcMBQAAAACglAFy2vbt26tWrTpp0qTY2Ni0tLTAwMDq1auvXr06Y8HZs2f79+/v4OAwderUxYsXZ4wrivLFF180bdo0KCjIxMTk0qVLU6ZMyam9bEuWyMmT/7ZgyhRxcJBNm8TSUhISHg82bCjr1mX56tBBChWSChVyJBQAAAAAFGimagcA8pWoqKiePXtWrFgxODi4bNmyIpKenj5lypRKlSplrNm2bZuZmdlvv/323XffhYeHZ4x/9913y5cvX7hw4ej/bWtRFMUoJ24lsrGRGjVk0CA5ckRMTLJf8/778skn4uws7dvLvXuPB8uXl27dnqxJTZURI+S//xVb29cPBQAAAAAFHaUMkJMCAgKSkpJ++OEHfSMjIqamprNmzcq8Zvz48fo/BAYGZh7/+uuvq1WrNnLkyIyRHGlk9Pz8pHFj+eYbGTYsy3h6umi1YmEh77zz/DdZvVri4iRTQAAAAADAq6OUAXLSvn37SpcuXb9+/aenNmzY8Oeffy5ZssTCwuLp2du3b4eHh3t7e+t0unPnzt29e7dmzZrFihUTkfT09KioqMyLFUW597/dLCYm5bTa0iKSlibJyYZv26jR4z80bCh9+siXX0qXLlKmzJMFgwbJqlWSnv78j6Yo4u8vnp5SpcrzFwMAAAAAnotSBshJ0dHRFStWzHbqyJEj33777ddff51tKXP16lURuXbtWtWqVa9duyYilpaWCxYsGDJkSGxsbJVnFyEtWgTv3Vv6WbPnzz/587x58vvvMnq0/Prrk8E2baRo0ed9KhER2bZNzp+Xb799ocUAAAAAgOeilAFykqIopqbZ/7Xy8/Pz8/N71oWpqakicuzYsYULF7Zs2fLRo0dffPHF8OHDGzVqVL58+cqVK2debGRkZPu/Y13KlTOqW1dExNxcChUyfFsrqyd/LlVKpk9/fChMhp49pWfPF/poCxZI/frSuPELLQYAAAAAPBelDJCT7Ozsrl+//goXlipVSkTGjRvXuXNn/Yi/v/+WLVu2bNkyderUy5cv50i8IUPkxx/F21v69n25C0+ckL17Zf36HEkBAAAAABDhkdhAzmrQoEF0dLT+/qOXUq5cOSMjo3/++SdjpGjRoiLy6NGjHIxnYiLLlsnZs5LpCd0vZMECqVRJPD1zMAsAAAAAFHSUMkBO+vzzz42MjIYPH67RaDIGdTpdenp6YmLijRs3FEXJ9sJChQo1bNhw7dq1yf87rfeXX34RkQYNGuRswoYN5bPP5PTpxy83bJCJE59zSXS0bNggo0fLM27MAgAAAAC8CqNn/RMRwKuZN2/e+PHjq1Wr9vHHHxcvXvzKlSvr16+fM2eO/ryYpKSktLS0efPmicjvv/8eGxv7+eefi8iUKVOOHj363nvvubq6du3aNTY2dsWKFU2aNNm5c6eJicnr5PnyS1m8WBITn4zEx4uLi8TFyb17MmbM46cvnT4tLVqIiDx4IIoiNjaPV4qIl5d8/71cvy6FC79OEAAAAABAFpQyQM7bvXv38uXLT506lZ6eXr58+ebNmw8aNOjcuXOHDx/28fG5c+dO36fOdNm0aVPhwoUPHjw4f/788PBwGxsbT09Pb29vS0vL1wxz9qxcuiSdOmUZPHlSLl8WT09JT5e0NClaVBISJCTE8Npu3UREgoKkSBGO+AUAAACAHEYpA7w1NBrN9evXNRrNgwcPMgaNje10OvuUFMl8+Iy7u5QqpUJCAAAAAMCL44gI4K0RFRVVtWpVg8GmTVcfONDHYPCXX170QdcAAAAAALVQygC57vr161FRUR4eHq/5PhYWFpUrVzYzMyuc6XAXe3vzunXF0lKsrJ6sLFnyhd7w+HEZNUqWLZMaNV4zGgAAAADgpXH7EpDrevXq9csvvyxatGj48OFqZ8lixAhZskSaNJH9+8XISO00AAAAAFDAUMoAuevo0aONGjWysLC4cOFCxYoV1Y6TRWKiODvLrVuydq107652GgAAAAAoYIzVDgDkZ4qieHl5KYoyduzYvNbIiEiRIjJjhojI2LGSnKx2GgAAAAAoYChlgFy0du3agwcP2tnZjRs3Tu0s2fvvf6VBA4mOFl9ftaMAAAAAQAHD7UtAbklNTa1Wrdrly5e//fbbAQMGqB3nmU6ckHfeEVNTOXNGnJzUTgMAAAAABQY7ZYDcsnDhwsuXL9esWbN///5qZ/k3detKv36SliZjx6odBQAAAAAKEnbKALni9u3bTk5O9+/f//PPP9u0aaN2nOe4fVucneXePdm6Vd5/X+00AAAAAFAwsFMGyBWTJ0++f/9+x44d834jIyJ2djJliojIyJGSmqp2GgAAAAAoGNgpA+S88+fP16xZU0ROnTrl5uamdpwXkp4utWuLpWVQr14xo0b9V+04AAAAAJD/UcoAOa9du3bBwcEjRoxYtGiR2llewv79p5o3r1W4cOHw8PDy5curHQcAAAAA8jluXwJy2LZt24KDg4sVK/bVV1+pneXlNGtWs1u3bg8ePPD29lY7CwAAAADkf+yUAXJSenp6rVq1zp07t3DhwtGjR6sd56VFR0e7uLgkJyfv2bOnRYsWascBAAAAgPyMnTJATgoICDh37lyVKlWGDBmidpZXUb58eR8fHxEZNmxYenq62nEAAAAAID9jpwyQY+7du+fo6Hjnzp1NmzZ5enqqHecVpaWlubu7X7x4cfHixcOHD1c7DgAAAADkW+yUAXLMzJkz79y506JFi7e3kRERc3NzPz8/EZkyZUpcXJzacQAAAAAg32KnDJAzIiMj3dzcNBrNsWPH6tatq3ac1/XBBx9s27ZtwIAB3377rdpZAAAAACB/YqcMkDO8vb1TU1P79euXDxoZEVm0aJGFhcUPP/xw9OhRtbMAAAAAQP5EKQPkgL/++mvTpk3W1tbTpk1TO0vOqFKlypgxY3Q6XXBwsNpZAAAAACB/MlU7AJAf3L59u2TJkkOHDrW3t1c7S46ZOHFiu3btmjVrpnYQAAAAAMifOFMGyBmJiYlmZmZWVlZqB8kVf/3115YtW65du2ZhYdGwYcO+ffsWKlTIYM2+ffs2b97crVu3xo0bZwxu3bo1ODj47t279vb2n376qaur65sNDgAAAAB5F7cvIXfcuCGrV8vt24bjgYFy/PiTl6mpsmOHLFwo8+fLhg1y717275aSIuvXi5+frFwpUVFZpi5dkoAAWbhQ/vhD0tNz9BPc8Pb2rlevXpUqVerVqzdlypTExMTMCx4+fDhr1qyGDRs6OTm99957e/bsyWhkdDrdxo0b27ZtO27cuByMpAqdTjdgwAAPD48NGzYYGxvHx8d7e3vXqFHj2rVr+gVarXbNmjX16tVr0aKFv7//3r17M64dMWLERx99FBMTU7Zs2T179ri7u2/btk2djwEAAAAAeQ+3L+W88HB5+FBq1xbjTJXXxYuiKOLsbLg4PV1OnhRnZylaNJu3unhRkpKkSBFxdJTUVDl9WipVkpIls6w5d07MzcXRMac/xmsKDZV+/WT/frGzyzLev78MHy716omIHDwovXpJdLQ4O4uZmUREiIWFLFokn36a5ZKjR8XTUxITxdFRrl2T1FRZtuzxGi8vWbhQypWTIkUkIkKqVZOgIClT5vXjh4WFtW7dWqPR9OvXz8HBITo6esmSJZs2bTp+/Li5ubmIpKWltWjR4vLlywMHDixXrtzmzZs9PT1Xr17dp0+fVatWTZ8+PSoqytzcPDk5+fXDqGvRokXff//9uHHjZs+ebWJiIiLR0dGzZ88umulXds2aNZ07d16xYkXmQ47v3r27ZMmS4cOHL168WETS09Nr1KgxderU999//81/CgAAAADIixTktFq1FBFl2bIsg61aKU2aZBlJTFT8/ZWKFRURZcaMbN4nPFyxtFRMTJT33lMURbl8WRFRli83XFajhtKuXQ7GzyF//KGIKPv3G46bmiqjRyuKoly+rNjYKO+8o0RGPp66e1fp2lUxMlK2bXuyXqtVXF2Vli2V+HhFUZTERKVRI8XaWrl3T/nzT8XISPn+e0WnUxRF2bVLMTZWBg58/exardbNza1EiRL//PNPxuCVK1e2b9+e8fLy5cvNmzc/c+aM/mV6erqrq2u1atUURfnpp59Wr1599+7dNm3aeHh4vH4eFel0unLlyrm7u2u12ucuTk1NFZFZs2bpX54+fVpE/Pz8Mha8//775cuXz62sAAAAAPC24falXGFqKhMnyq1bz1yg0YiTk6xYIT17Zr9Ap5OBA8XdXfLF45Wz4+srGo1s3CgODo9HiheXwECpXFkmTXqyzNhY9u+XtWulWDERERsbGTxYHj6UM2ekTRs5f17++18xMhIRadlS6tWTY8deP9rBgwfPnz8/bty4KlWqZAxWqlSpffv2GS8rV668d+/e6tWr61+amJg0bNjw4sWLiqL07t27T58+xYsXf/0kqrt8+fKNGzc6dOhgbJzN/1aMGjVq2bJlz7rW2dm5TJky8+bN27Fjh4hERkbu3bu3d+/euRgXAAAAAN4qlDK5on17sbKSMWOeucDMTE6dknPnZObM7BcEBMjhw7J8uZiY5FJGtW3bJs2aSblyWQYtLKRbNwkLk5gY+fFH2b5dRKRECSlR4skaMzMRkdRUEREXF8PLHz16/WhHjx4VkSZNmjw9df78+Ro1amzcuPHpqdjY2BIlShjpG6L8Ijo6WkQqVqyY7ezatWv37NnzrGvNzc1///13jUbToUMHZ2fnJk2a9OnTZ+azfuMBAAAAoODhTJlcUaiQzJ4t/ftL797SoUOWqYMHpUgRqVHD8KyVzGJiZNIkGTVKatfO7aS5bOlSMegvdDoRkeRkuXlTPvoom0sqVxYR+ecf8fGRBg0Mf3wismuXWFlls4MoPl5OnpRPPhERmTlTfvzRcEFamiQni4iPm9vy8+f1Y8nJyWlpaZlXXb9+PS4uTkQqVKjwdDpjY+NChQqZmhr+xbl+/XpISMiwYcOy+URvM0VRRMTkGdXgzZs3/+VanU4XEBBgamq6bNmyQ4cObdq0KTAwsGnTpp/o/xsBAAAAQIFHKZNb+vWTgAAZNUpathRLyyfjXbpIo0ayefO/XTtkiBQpIlOmZDN14IDhSHy8lC372nFzydWrcv9+lhH9I9j1PUjhwtlcoj8+9tEj2btXnn689MmT8tNPMn682NoaTk2aJDrd41ufbtyQyMhnhUp/+DAhIeFZs1qtVt+56E9IMeDi4nL48GGDQY1G07dv32LFik2ePPlZb/uWsrOzE5Hr16+/wrWBgYHff//9n3/+2aZNm8GDB8fFxXXv3r1///4eHh7P2noDAAAAAAUKpUxuMTKSZcvknXfE11emTXsy/u238p///NuFa9fKli2yZUv2lcX27bJ/f5aR2Fj538EmeY+fnzRtmmVEf/NRkSJiZiY3bmRzif7f/yVKGN6aJCJXr0qnTlKnjowfbzi1ZIksXy7Ll4v+X/sTJ4q3t+Eac3MpVEhEpohM/N+YtbW1hYWFwcJy5cqJyLVr1yrrt+38q7S0tI8//vjkyZMhISG2T1dFbzkXF5dixYodeLoLfAGHDx82NjZu2bKl/mXJkiXHjRvXoUOHkydPUsoAAAAAgHCmTK6qU0e++ELmzpV//nky+OGH4uHxzEvi42XkSOnSRT78MPsFvr5y7VqWL1fXHI79JpiYSK1aEhaWzVRoqFhZSbVqhuMnTkjz5lKihGzfLtbWT8Z1Opk6VUaOFD8/+fzzx4P29lK5suFX+fJSrJgUK1a42BNPNzIi0qJFCxH5/fffn/s5bty40aJFi3379m3fvr1+/fov9NnfKiYmJgMGDNi3b19gYGDmcf02ohs3bty5c+dZ1zo4OOh0uhMnTmSMhIWFiUj58uVzLS8AAAAAvE3YKZO7Zs6UDRtk4sTnr9SbM0diYyUsTOrVezxy4YKYmEi9evLtt7mUUSWffipDh8q6ddK9+5PBs2dlwwbp21cMupLVq2XwYGnZUgIDH9/fpJeYKP36SVCQ/Pij9OuXU9GqVavWoUOHZcuWtWzZsmPHjvrBR48excTEFC9efNeuXXXr1q1UqdL+/ft79OhhZ2f3999/O2Q8QyrfmTp16sGDB/v27btx48bmzZunp6efOHFi9+7dV65cqVWrVtOmTTdu3BgSEhISEqLVakUkODg4MTHR1dX1k08+8fPz8/T0HDt2bKVKlY4fP+7v7//BBx/ky/YKAAAAAF4BpUzusrWV+fOlXz+xsxNHx+evb9HC8OamZcseP5LI3DyXMqpk4EDZuFH69pXTp6VDB7GwkEOHZMYMKVdOZs8WEXFzk7p15YcfZOhQ+fZbqV9f+vSRP/98fLmbm2i10qmTXL0q48eLtbWsX/94qmPHLKf4vJIffvihQ4cOH374YYMGDdzc3O7cuXPo0KFatWrNnTu3a9euK1assLe379ixY3p6euXKlQcOHJhx4dKlS0VEf+LvyZMndTpdmzZtjIyM/sxInvdotdpnHeUrItbW1nv27FmxYsWWLVuWLl1aqFChypUr+/v7m5qaTpo0yd7eXkRiYmL0O2Jat24tIidOnDAzM+vXr19YWNj8+fPXrVsXGxtbunTp6dOnjxgx4o19LgAAAADI44z0T1dBDqpdW1xc5JdfHr9UFGnVSvbskSZN5MAB+fBDqV5dZs+WmzclJka0WmnQQAYPls8+k6JFpWpVw3dr3FhsbCQ4WCIjpUoVWb5cMjUAIiI1a0rZsrJjh8TFydSpcuaMODrK9OmGD5t+04KDpXt3CQqSRo2yjJcoIYMGyaxZIiKPHsns2fLDD48Pl7G1lW7dZObMxw+m+vxzcXSU3r2zPzLH21s0GvH3z2YqIkJKlnz9T5Cenr5x48a9e/feuXOnTJkybm5uPXr0MDY2PnLkSLVq1a5fv745u+Oahw4damRkpK9mMly/fn3x4sX/+ffDhFSSlpZWt27dDz/8cPLkyZZPlVm3bt0qXbq0KsEAAAAAIN+jlMl5BqWMiJw7J7VrS4MGcuCANG8uNWvK4sUyfbrh85Xef1+2bjV8txcvZbp3l3r15JNP5Lvv5OBBCQnJnY+XG+7dk7Q0KVlSjIzUjvLqjhw5kpKS8u677xqMz507d8KECV988cU333yjSrB/5+fn5+3t7erqevr0aYNHfQcFBXXq1Omrr77y8fFRKx4AAAAA5GOUMjlMp5OYGDE3f7zbI8ONG2JklOXZ1QkJYvBcZmtreXpTQkyMGBlJmTKi1UpsrNjaZjnlVkTi4sTERIoXF51OjI1FRK5fl5o1JT4+xz4UnmvXrl1t2jQwfIYAACAASURBVLSpVKnSuXPnrLI+yfvSpUvu7u4ajebIkSN57TiV+Ph4R0fH+Pj4oKCgtm3bZp7SarW1atU6e/asn5/f2LFj1UoIAAAAAPkYT1/KYZ98IjNnPi5HMitXLksjIyLFihk+HSjb20TKlpUyZURETEykbFnDRkZESpaU4sVF5Mk3DQsTd/fX/iR4GS1atKhZs+aVK1fmz59vMOXo6Dh8+HCdTjdq1Ki81oFOnjw5Pj6+ffv2Bo2MiCxfvvzs2bOVK1fWn48DAAAAAMhx7JTJSX/9JU2aiKWlRESIvb06GWJipEkTWbVKmjZVJ0CBdejQoaZNm1paWp4/f75SpUqZp5KSklxcXGJiYn766afevXurFNDQhQsXatSoodPpwsLCqmc9uCcpKcnR0TE2Nva3337r3LmzWgkBAAAAIH9jp0yOURQZNUoURSZMUK2RiYyU99+XmTNpZFTg4eHx8ccfp6SkeHl5GUzZ2NjMmTNHRLy8vO7fv69GumyMGTNGo9EMGjSo+lNHKc+YMSM2NrZx48adOnVSJRsAAAAAFATslMkxq1dLv35SrpxEREihQioE2L5dhgyRr7+WFi1ERIoWzeYuKuSqW7duOTs7JyYmPn1Ei6IoLVq02L9/v5eX19O3OL15u3fvbtWqlY2NzaVLl0qVKpV56sqVK66urhqN5ujRo/Xq1VMrIQAAAADke5QyOSMlRVxcJCpKVq+WPn3UydCliyQmPnn566+SJx/BnM/Nmzdv/Pjxrq6up06dMjMzyzwVGhpav359Y2Pj06dPu7i4qJVQRLRabe3atc+cOTNv3jxvb2+D2W7dum3YsKFfv34rV65UIx0AAAAAFBSUMjlj2jSZOlVq15bjx9mfUqClpaXVqFEjIiJiwYIFY8aMMZgdPHhwQEBA69atd+7cqUo8vYCAgMGDBzs4OISHh1tYWGSeOnz4sIeHh6WlZUREhL1at+EBAAAAQMFAKZMDYmLEyUmSk2XfPmnWTO00UNvOnTvfe+89GxubiIiIMvpHZ/1PfHy8s7PznTt3Nm7cqNZxLUlJSU5OTrdu3Vq/fn3Xrl0zTymK0rBhw2PHjk2dOnXKlCmqxAMAAACAgoNNHTlg0iRJTpauXWlkICLSpk2bjz76KCkpycfHx2CqePHi06dPF5FRo0Y9fPhQjXQya9asW7duNWrUqEuXLgZTgYGBx44dK1eu3NNnFQMAAAAAchw7ZV5XWJjUrSumpnL2rDg6qp0GeUNUVJSrq2tKSsq+ffuaZn0Ulk6na9iw4d9//63KbpQrV664ubmlpqYePXq0fv36madSUlJcXFyioqJWrlzZr1+/NxwMAAAAAAogdsq8Li8v0elkxAgaGTxRoUKFsWPHKooycuRIrVabecrY2Njf39/IyGju3LlXrlx5w8EmTJjw6NGj3r17GzQyIjJ//vyoqKjatWv3UeuoagAAAAAoYNgp81r++GPPrFmuly+XvnRJbG3VToO8JCUlxc3N7erVq8uXLx84cKDBbJ8+fQIDAzt37vzbb7+9sUgZh/heuHChQoUKmadiY2MdHR2TkpL27dvXjNvwAAAAAOCNYKfMq0tLSxs7dtDRo2V8fbfSyMCAlZWVn5+fiPj4+Ny5c8dg1s/Pr2jRohs3bgwKCnpjkaKjo4sVK+bl5WXQyIiIj49PUlJS586daWQAAAAA4I2hlHl1S5cuvXTpkouLS79+bdXOgryoS5cubdu2jY+Pf/rsmFKlSn355ZclS5Z8k8f9duvW7dKlS+PHjzcYP3Xq1KpVq8zNzefMmfPGwgAAAAAAuH3pFSUkJDg6Ot69e3f79u3t27dXOw7yqPDw8Jo1a+p0uuPHj9eqVSvzVFpaWnJycrFixdTKlqFNmzYhISFjx47Vb+0BAAAAALwZ7JR5RVOmTLl7927Lli1pZPAvXF1dhw8frtVqhw0bZlCAmpub54VGZvPmzSEhIcWLF584caLaWQAAAACgYGGnzKuIiIhwd3fX6XShoaHu7u5qx0GelpSU5OzsfPPmzcDAwF69eqkdJwuNRlO9evWLFy8uXbp06NChascBAAAAgIKFnTKvwsvLS6PRfP755zQyeC4bGxtfX18R8fLySkxMVDtOFkuXLr148aKLi8vTz4cCAAAAAOQ2dsq8tN27d7dq1crGxubixYulS5dWOw7eAoqiNG/e/MCBA+PGjZs7d67acR7LOBdp27ZtHTp0UDsOAAAAABQ47JR5OTqdztvbW0QmTZpEI4MXZGRktGjRIhMTE39//4iICLXjPHblypUiRYq0adOGRgYAAAAAVEEp83K+//77kydPVqpUaeTIkWpnwdukdu3aAwYM0Gg0QUFBamd5rE6dOufPn1+1apXaQQAAAACggOL2pZegKErNmjXPnDnz66+/9ujRQ+04yLtSU1N37dp17tw5jUbj6urarl07KyurO3fu/PPPPw0bNhQRRVF2794dFhZmbm5er169Ro0aZVyrKMrOnTtPnjxpYWHx7rvvGjxI+9+lpKR88803mzZtun79upWVVYMGDSZMmODi4pJ5TUJCwooVK7Zs2eLl5dW5c2f9YFJS0syZM7dv3/7gwYMaNWp8+eWX9evXz4mfBAAAAADgmShlXs79+/dXrVo1fPhwIyMjtbPkAq1WTp+WmBgpVEhq1pSnH9h886bExGRzYZ06ki9/IK/k6NGjPXr0uHbtWqVKlQoVKhQREVG2bNlNmzbVqVNHv+DOnTsfffTRkSNH3NzcHjx4cPXq1b59+/7444/Gxsapqamenp5BQUH6qevXr0+YMGH27Nkv8n2TkpJatmx58uTJbt26NWrUKCkp6ddff71y5cqxY8eqVasmIqmpqV5eXitXrjQ1Nb13756vr++ECRNE5NGjRx4eHhcuXBg6dKitrW1gYODVq1cPHjyYERgAAAAAkBtM1Q7wmE6nu3v3rk6nK1mypLFx9jdV3blzx9ra2traOvdiPHz48M8//zx//rxWq61WrVqHDh0sLS0zLyhSpIiTk9PcuXMtLS3r16/v4eGRe2HetLVrZexYuXFDjI1FpxMTE+ndW5YsERubJ2u+/VamTMnmWo1GTPPK75K6YmJiOnToULRo0RMnTuhLjRs3bgwYMODatWsZHcfMmTMjIyPDwsL0T+/y8fGZM2dOp06dPD0958+fHxQU9O233w4YMEBRlDFjxvj6+rZs2bJ169bP/dZffvnl8ePH169f37VrV/3IuHHjfv31Vzc3N/1LCwsLOzu7zZs3N2jQwCbTf9aff/755MmTGfu/hg4dWqVKlcmTJ2/bti1HfzYAAAAAgKwUtSUlJY0ePbpEiRL6PCVLlpw2bZpGo8lYoNPp/P39y5Ytq19Qq1atgwcP5kaS3bt3lylTxsjIyMnJqXr16iYmJg4ODmfOnMlYEBcX17hxY0tLy3r16lWtWlVEunfvrtVqcyPMm7Z2rWJkpHTqpEREKIqiJCUpAQGKlZXSrJmSnv5kWUqKEh+f5athQ8XDQ63UedC4ceNE5PDhw/+yRqPRREZGZryMi4sTER8fH0VR7O3t69atmzGVmJhoZWXVo0eP537ftLS0woULt2nT5kVCpqSkiIivr6/+5aBBg0xMTDL/pRswYICFhUU++d0GAAAAgLxK5YN+09LSWrduvXTp0k8//XTv3r2HDh364osvfH19v//++4w1c+fOHTt27GeffXbixImgoKC0tLSOHTveunUrZ5NERkZ++OGHtra2586di4iIOHPmzJUrVxwdHW/cuJGxZv78+YmJiVeuXPn7778vXbo0b968devWbd26NWeTqECjkdGj5Z13ZP16cXISESlcWAYNkmXLZP9+Wbv2yUpLSylW7MlXZKQcOSJeXmoFz4N27txZuXJl/cExBkJCQpYuXSoipqamDg4OGePm5uYikpqampCQcP369fbt22dM2djYNGnS5PTp0//+TbVa7fnz5x88ePCsrVseHh6jR49+1uWWlpZarfb+/fsZI1WrVk1NTdW3RQAAAACAXKJyKfN///d/R48eXbZs2fz585s3b964cePp06eHhoYOHDgwY03NmjVXrFgxffr0OnXqtG3bdtmyZQkJCTlehfj5+SUnJ69fv97V1VU/Ym9vHxwc3LZt24w1vr6+u3btyngSdseOHUXkn3/+ydkkKvjrL4mJkUGDxMQky3jv3mJrKxs2SEKC7N8vd+8aXujnJ5UrS8eObyxp3hcZGVm5cuVsp37++ecxY8Y8Pb5z504Radasmb4BzNgUple6dOno6GgR+eyzz4oXL25lZWX0lJUrV+oLFHt7+2y/ta2tbeHChZ+VuUuXLkZGRkOHDr1582ZcXNzPP/+8aNEiEYmNjX3BTw0AAAAAeAUqlzK//PJL+fLl//vf/2YedHFxMTIyevToUXJysoi0b98+8wInJycRuX79es4m2blzp7u7u/48VANbt25dtmyZiBgbG9vZ2WWMb9682cjIKD8cK3P+vIiIo6PhuKmpVKki58/L4cPSvLkcPJhlNjpafvtNxo41rHIKNo1Gk/m4lsxmzZp16tQpg8EHDx58+eWXDRs2/PDDD/W/8FZWVpkXWFtbP3z4UKfTPXr0KCEh4dGjR9m+uampqYg8a3bbtm0zZsx4VuamTZv6+flt3ry5bNmydnZ2M2bM0O/0MThQCQAAAACQs1Q+nPX06dMtW7bM9mTfnj17Hj58+On/r/7ChQsioj/SZcGCBenp6frx5OTktLQ0g8X37t1Tsj5eKj09PSkpKfNIkyZNRowYcfXq1Y7P2PGxatWqXbt2DRkyRP8yNDR09erVYWFhp0+f/v777xs0aPCCHzbvevBARKRkyWym7Ozk5k2pVUsCA6Vu3SxTCxeKjY306/cmEr49ihUrpt/Y8rSyZcsa7IJJTU3t2bNnXFzchg0bjIyMypQpIyJ37tzJvCYuLq5MmTLGxsYrVqxYunSppaWlQWujd/nyZRGJiop6tdhjxoz59NNPz549W6pUqapVqw4ePNjU1LRChQqv9m4AAAAAgBehZimTkpKSmppq8M/UDJ6enrVr1356fMGCBeXKlevUqZOITJkyRb+54HWYmJgMHTpUq9U+6/6OxYsXP3z4MONlamrqvXv3ROThw4ehoaF9+vQxfdsfPFS0qIjIrVvi7Gw4dfOm2NpK2bLSq1eW8cRE+eEHGTZMChV6QyHfErVr1963b19KSkq21Ulmd+7c+fjjj//++++QkBD9Fq0yZcqYmpqe129c+p8LFy7ob0oqVKhQoWf/tCtXrlyhQoWtW7fOnj37Wc8v+3fFixdv1qyZiKSlpQUHBzds2JCdMgAAAACQq9RsEywsLIyNjRMSErKd/fTTT58eXLhw4fbt29euXasvULy8vDLu17C2trawsDBYX7RoUYN/oJqYmBQpUiTziL29vYmJSdGiRZ+1wUG/fyFDw4YN9Td3BAUFtW/fvkqVKiNHjnz2p3wbuLuLiERESPPmWcZTU+XyZXn//WwuWb5cHj2SoUPfRLy3St++fbdu3Tpnzpxp06b9y7LQ0NAuXbqYmpoeOnQo46Y5MzOzVq1abd68eeHChba2tiJy4sSJ8+fPT58+/bnf18jIyNvbe/jw4d7e3nPnztUXhYqinDt3rnr16lu2bLGzs2vUqNGzLtfpdPq/KYqizJgxIyoqyt/f/6U+OAAAAADgZRkZ3N3zhjk5OVlbW4eFhb3I4kWLFo0ePXrBggX/8hyZV9ayZcszZ87ExMSYmZm9+FUlSpRo1qzZxo0bczzPG6XVStWqUrSo/P23ZP74AQEyeLBs3iwffZRlvUYjVatKq1byww9vOGnepyhKly5dNm3a1KdPn65du/7nP/8JDw//9ddf58yZ8+OPP65fv/727ds//vjjkCFDChcu7OvrW1S/TUmkWLFirVu3PnLkSLNmzerUqTN16tR79+5NnDgxNTX17NmzxYoVe+631ul0Q4cODQgIqFixYuPGjTUazcmTJ2NiYm7fvl2pUqV33313w4YNq1atCgwM1Ol0u3fvrlq1aqVKlRo3buzj41OnTp0KFSrY2NiEhoZGRkb6+PjMmjUrl39UAAAAAFDQqVzKjBo1avHixSdOnMj2TqUMKSkpQ4YMCQwM/Prrr4cNG5YbSVauXNm/f39fX98JEyY8a01kZOTdu3fr16+vf5mQkFC6dOmPP/545cqVuRHpjdq6VTw9pU0bmTNHqleXhAT5+WcZP15atZKtW+XYMZkwQWbNksaNRUQCA6VPHwkLk5o11c6dF2k0moULF37//feXLl0SkdKlS7ds2fKrr77as2fPgQMHfv755yZNmhjcoyQitWrV2r17t4gEBwdPmDAhLCzM1NS0ZcuWixcvdn76trJnO3r06ObNm6OioooUKeLg4NC1a9fKlSvv37+/WLFi7u7uO3fu3LVrV+b1bm5uvXv33r59e0hISHx8fIUKFbp06fLvfx8BAAAAADlC5VImOjra3d29WLFia9as0d8T9PDhw7Vr13bv3n3dunVXr16dNm1aVFRU165dIyMj165d26pVq1xKotPpOnbsuGPHjr59+3bu3Llw4cIXLlz4+eefV6xYMXfu3L1790ZFRQ0cOPCnn36aNm1a48aN7969O2/evGPHju3du/etfABTaqoMHixjxkj16o9Htm6VESPkypXHL83NZcAA8fMTKyvZt08+/1y++Ub0P/86dcTOToKC1En+9khNTVUU5dVOZklLSzM1NX2102FeXFRU1J49e/pxWjMAAAAAqEHlUkZEjh492qtXr8uXL5cuXbpQoUJRUVGWlpZBQUHz5s3TP33Jycnp0qVLBg+deffdd3/77becTZKWlrZgwYKVK1devHjRxMSkTJkybdu2nTRp0h9//HHu3Lnly5enpaXNmTNnxYoVN27cMDIyqlu37syZM9u2bZuzMd4QPz/x9paaNSU0VIyMHg8qikRESHS0WFtLjRqS7cnHOp1cvSrFi4ut7ZvMixwXHx/v4OCgP6+6ekY3BwAAAAB4U9QvZUREo9EcPHgwPDzcxMSkQoUKTZo0sbGxSU1N1Wq11tbWW7duTUlJMbikVKlS+ifF5Ib09HRFUZ51uEx6erp+F4O5uXkuBch1cXHi6Cj370tQkLylpdLb5urVq/ojpdUOksWIESOWLFnStGnTffv2GWV0cwAAAACANyJPlDKv7/Tp02vWrDEYfPq5ThqN5sGDB/o/z58/v2LFii/7jfbv3z9s2LCpU6d27tz51aLmCUOHyrJl0qGDbNumdpQCQaPRuLu7m5mZ/fHHH5UqVVI7zhMJCQnOzs5xcXHr1q3r1q2b2nEAAAAAoGDJJ6XM+vXru3fv/lKXnDx58hVOMw0ICBg8eLC9vX14eHihQoVe9vI84cIFqVFDdDoJCxODm1bu35feveXLL6VBA5XC5U+LFi0aNWqUs7PzmTNnXurxXm/Ad9999/nnn5cvX/7ChQtv6680AAAAALyd8m4ps2fPnqVLl86dO7dq1arPXXzx4sVNmzYZDD79FGEzM7PC/zsnpXXr1i/ymGEDOp2uYcOGf//995QpU6ZOnfqyl+cJHTrIjh0yZIj83/8ZTo0bJ/PnS/PmsnevCsHyqYSEBEdHx7t37/7xxx8ffPCB2nEM6XS6Ro0aHTt2bNKkSTNnzlQ7DgAAAAAUIHm0lNHpdO+8886JEydmz57t4+OjdpwsDh8+7OHhYWlpee7cOQcHB7XjvKTdu6VVK7GxkUuXpFSpLFNXroirq2g0cvSo1KunUr58aPTo0f7+/u+++67+idd50PHjxxs0aGBqanrmzBknJye14wAAAABAQZG7D9x9ZT/++OOJEyfKly8/cuRItbMYatSoUe/evVNSUsaMGaN2lpek1cqoUSIikycbNjIiMm6cpKZK3740Mjno8uXL33zzjbGxsZ+fn9pZnqlevXr9+vVLS0vz8vJSOwsAAAAAFCB5cafMgwcPnJycbt68+csvv/Ts2VPtONmIjY11dna+f//+9u3b27dvr3acF7Z8uXzxhTg4SHi4WFhkmTp8WDw8xNJSIiLE3l6lfPmQp6fnli1bPv/88xUrVqid5d/cvn3b2dn53r17W7duff/999WOAwAAAAAFQl7cKTN37tybN282aNCgR48eamfJXqlSpSZPniwiI0eOTE1NVTvOi0lKEv0hOPPmGTYyiiIjR4qiyPjxNDI5aO/evVu2bClcuPC0adPUzvIcdnZ2U6ZMkbfrVxoAAAAA3nJ5rpSJjo5euHChkZGRn5+fkZGR2nGeaeTIkdWrV7906dLixYvVzvJCDgUEKGlp4uEhXboYzgUGyt9/S7lywt0rOUen0+nvBvLx8SlTpozacZ5v2LBhNWrUuHz58tdff612FgAAAAAoEPLc7Uu9e/f++eefP/744zVr1qid5Tl27drVunVrGxubCxculC1bVu04/+bq1auurq5lraxOhITY1qmTZS4lRVxcJCpKVq2Svn1VCpgP/fDDD5999ln58uUjIiKsra3VjvNCDhw40Lx5cysrq/Dw8AoVKqgdBwAAAADyuby1U+bYsWNr1qyxtLScPXu22lmer1WrVp07d05KSho/frzaWZ5jwoQJjx498vjgA8NGRkTmz5eoKKldW3r3ViNa/vTgwQP9DW7z5s17WxoZEWnatGnXrl0fPnw4YcIEtbMAAAAAQP6Xt3bKNG3a9ODBgxMnTpw1a5baWV7I9evXXV1dHz58uHfv3mbNmqkdJ3tHjhxp3LixpaXlhQsXDLc/xMSIk5MkJ8u+fZJX87+NvvrqqxkzZjRo0ODw4cN5+S68p0VHR7u4uCQnJ+/Zs6dFixZqxwEAAACA/CwP7ZRZt27dwYMH7ezs8v7Gkwz29vbe3t6KoowaNUqr1aodJxuKonh5eSmK4u3t/fQNKduWLNHa2EjnzjQyOejGjRsLFy4UkTx+LlK2ypcvr98mM2zYsPT0dLXjAAAAAEB+lldKmbS0tIkTJ4rIzJkzixQponaclzBu3DgHB4fQ0NC8+czjX3755dChQ6VKlfJ66hDf0NDQD+fNq/jo0YN581TJll9NmDAhOTm5R48eTZo0UTvLq/D29nZycjp37lxAQIDaWQAAAAAgP8srpczXX399+fJlNze3/v37q53l5VhZWS1YsEBEJk2adOfOHbXjZPHo0SN91eXr62tjY2MwO27cOJ1O9/GAAYWrVFEjXf4UGhq6Zs0ac3Pzt+UWvKdZWFj4+fmJyFdffRUXF6d2HAAAAADIt/JEKRMXF+fr6ysiX3/9tampqdpxXlqnTp3at2+fkJCgP9s171iwYMG1a9dq1qzZr18/g6nNmzeHhIQUL17cx8dHlWz51ciRI3U63ZgxY6q8zVVXx44dO3TokJCQoC/1AAAAAAC5IU8c9Dt48OCAgIAPPvjgjz/+UDvLK7p06ZK7u7tGozly5Ej9+vXVjiMicvv2bUdHx8TExJ07d7Zu3TrzlEajqV69+sWLF5cuXTp06FC1EuY/69ev7969u52d3cWLF4sWLap2nNfyzz//VK9eXaPRHD58+J133lE7DgAAAADkQ+rvlAkPD//uu+9MTU3nzJmjdpZX5+joOHz4cJ1ON3LkyLzQc4nIxIkTExMTPT09DRoZEVm6dOnFixddXFwGDhyoSrZ8KeNcpOnTp7/tjYyIVK1adfTo0TqdbujQoTqdTu04AAAAAJAPqb9Tpl27dsHBwcOHD1+8eLG6SV5TUlKSi4tLTEzM6tWr+/Tpo26YW7duOTg4aLXas2fPOjk5ZZ5KSEhwdHS8e/fu1q1b33//fbUS5j/z588fN26cm5vbqVOn3sa78J724MED/a/0nj17mjdvrnYcAAAAAMhvVN4ps2PHjuDgYFtb26+++krdJK/PxsZGv9nH29v7/v376oYpXbp0aGhoQECAQSMjIlOnTr17927Lli1pZHLWX3/9JSILFy7MH42MiBQuXPi7777bv38/jQwAAAAA5AY1d8potdpatWqdPXt2wYIFY8aMUStGDlIUpUWLFvv37/fy8po/f77acbKhf8SVVqsNDQ11d3dXO05+c/Dgwbz8GOyYmJgdO3ZER0cXKlSoUaNGHh4eBgsSExM3b94cFRVVunTpNm3aVKxY8ek3+eOPP27evFm6dOkPP/ww1xPfvSv+/rJjh9y8KYULS506Mny4NG5suCwuTubOleBgefBAHB3F21vatHk81b27REY+WWlhIYcO5XpsAAAAAHgxapYyy5cv/+KLL6pUqXL+/Hlzc3O1YuSs0NDQTz/91N/f/91331U7SzY++uij33//feDAgcuXL1c7i4iIosjJk2JlJW5uWcbDwqRECSlf3nB9crJEREj16pL59+XmTTlzRszNpVo1KVky1zPrabXaDRs2BAcHx8bG2tratmrVqk+fPmZmZpnXhISErFu3LjY2tkKFCn379s04ATo5Ofm77747fPiwTqdzd3cfPHhwiRIlcjuwr6/v1KlTjY2NnZyckpKSrly50r59+/Xr1xcqVEi/YNeuXV27dlUUxcnJ6fLly48ePVq9enWXLl0yv8nRo0c9PDwsLS1r16594MCB3E187Zo0by7378uAAeLkJImJsmaNhIXJkiUyZMiTZffuSd26otHI4MFStKisWSOHDklQkLRtKyJSooTUrSstWz5ebGOT5VoAAAAAUJWapUxCQsL06dObN2/u6empVoaclZaWdvv2bTMzs1KlSr2xb6rRaP78889Tp06lpqZWrlz5o48+srW1NVij1WpDQkJ27drl7+9vamp64cKFChUqvLGE/yI1VSwtxdhYDh2Shg2fjBcpIgMHip/fk5FbtyQgQJYulbt3JThY3ntPRP6fvfuMiurqwgD80qRIRxTEAgqIiiIo9oQgdo29d4mFWLEFxY69d6PGlthjIXZFjCVYI4KCCgJioSu9M+37MYgwL9AyywAAIABJREFUYD4j4AzxfRYra+aUe/fF0bVm55x9IBRi1ixs25b/ulIlzJmDhQvLPezk5ORu3brduXOnVatWDRs2jI+P9/HxsbOz8/X11dHRkY6ZPn36xo0bO3bsaGFh4efn9/Tp00OHDg0aNCghIcHZ2TktLa1z585CofD06dPq6up///13jeIpqLKzf//+0aNHDx48ePv27dKPx7Vr1xYvXnzw4EHpfQUCgZWVlb29/aFDh7S0tNLS0pydnV+9ehUVFaWhoSG9iFAodHR0rFatmlAozM3NLfekTPv2CAjA/fsoOFxcJMLAgThzBo8eoX79/MZHjzBzJn77DaamAJCVhbp10bAhfH2RmwtNTWzfDje38g2ViIiIiIjo80jKjkgkevjw4R9//HHhwoW3b9+WOEZ6aHRKSkoZ3lcRRERE9OvXT1NTU/pbrVOnzm+//faxwV5eXtJhN2/eLOV9AwMD69SpA8DS0rJJkyaamppVqlTx9fUtPCYqKsre3h6AhYVFkyZNNDQ0fv7551Let6zk5EgAiaqqpGlTiVD4oV1HRzJjxoe3T59KNDQkrVpJXF0lgOTy5fz2JUskRkaSy5clYrEkI0MyYYIEkFy5Uu5hDxkyRFlZ+dixYwUtQUFBCxYsEL5/hoSEhNq1a+/Zs0f6Ni8vr3HjxmZmZhKJ5NixY82aNUtNTS2YCGDhwoXlF61YLLawsLCyssrNzf2HYW/fvi2ISiKRbN++HUBAQEBBy+LFizU1NcPDw11cXNq2bVt+AUskEsmzZxJAsmyZbHtUlERFReLu/k9zO3eW1KolkUgkL15IAMnp0+UVJBERERERUemUWaFfHx8fS0tLBweHXr16de3a1dTU1N3dXSQSFQxIT0/ftGmTpaVly5Ytt2zZUlb3VQSvXr1q1arV9evXly9ffvfu3evXrzs7O48ePTogIKD44NDQ0OXLl5dJPY6kpKQuXbrk5OTcu3cvLCwsICAgOjraxcUlNja2YIxIJOrTp09UVNTDhw9fvHgREBDw7t27MWPGlP7uZWjkSAQG5i94KZGNDV6+xO3bcHUt0j5hAi5eRMeOUFJC5cr46ScAuHevfKONiYk5duzY0KFDBwwYUNBoa2u7ePFiFRUV6VtjY+OIiAjX9+Gqqal169YtOjo6MTFxwIAB9+7d09XVlXbVrVtXWVn57du35RdwWFhYZGRk3759S9wkuGXLlosXLwKoUqVKQVQApOWK8/LypG9DQ0NXrFixYMGCugXrVkojJQXJyUV+3r3Dixf5P3l5uHMHAFq0kJ1oZobatXH3Lq5dg40Nbtwo4eIxMahaFQDi4gAgIgJLl2LBApw5A57tTUREREREiqRskjJ37tzp3r27np6en5+fSCTKyMjYsmXLsWPHnj59Kh0g3Ryxe/fuwYMHl8kdFcqsWbNSUlJ8fX3d3d1btGjh5OS0e/fup0+fStenFCYWi8eMGdOwYcMJZVHY4pdffomNjd29e3fz5s2lLQYGBkePHh02bFjBmKtXr96/f3/BggUFwVSuXFnRjgdq2BBjx2LePERHF2nPycG1a3j9GkpKKHFDmKEh3tdpAYAHDwDAxqY8YwVu374tEom6dOlSvCslJWXw4MEnT54EUJCgkcrMzFRWVlZXVwegrJz/904sFq9evRpA//79yy/gyMhIANIVVcXNnj370KFDxdt9fHz09fUbN26M95/bunXrzpgxQ2bYkiVL6r5nbm5uWIyGhoaSkpKSklJcixZQUsr/MTCAoWGRH2Nj1K2b//P0KRISAJRQVQhArVqIj4e6OqpVg7q6bO/ff+PxYwwdCgBiMbS0cOgQHj3CxYvo2RNduuB9momIiIiIiEjuyubLuYeHh7a29uXLl6tWrQqgcuXKbm5uo0aNKqhGoaamFhQUZGxsLBQKpedG/2dkZ2efPn26R48ednZ2hdulZ1GnpaWpqKgU1FL9+eef79y5c+/eveTk5NLf+sqVK/r6+p07dy7edf369fv377u7u1+5cgXAoEGDAERHR+vo6BReDaE4lizB77/D3R3Hj39ojItDu3ZYvhxz5vzT3KgoXL+Ox4+xaxfGjUPv3gDwww9IT88fkJoqu0JCJEJaWv5rK6tLYWHzAOTk5GRnZ8tcPD09XSgUFrz18fGJiYkBUOLJRLm5uefPn5cmMgoTCATe3t7t2rXT1taWtjx79uzHH38MCwtLTU319vb+7rvv/ukJS0e62qXg1jICAwML6uAUuHHjhre395o1a6R/hbdv337r1q2//vpLppIxgOTk5BeFjzf6OLFyoRSwnh6Ui2aElZWhp5f/WkUF0hvl5JRwoexsqKujdesSlsmkp2PUKDRogIkTAaBNG6SmoiAFuWsXxo/HL7/k9xIREREREclbGSRlkpOT/fz8XF1dpRmZAtKvczdv3tTT07OzszP+YufifFkhISF5eXkfO17a2tq6RYsWp0+fBhATEzN37tzJkyc3bdrU19e3YEx6evqyZcukr6XVdgpfQSgUphdkF96zsLBYtWrVy5cvLSwslJSUit/3/Pnza9eudXNzCw8P19XV3bdv36pVqxITE1VUVAYMGLBjxw5FS81UqYJly/Djj7hwAV275jcaG+PIERRNdpUgNBRr1iAqCmpqsLfP/7J/9iw+cUuQunqKv7//J8YpEAikf0DKyiWsMqtWrVpaQbKnkPnz58fExJw5c6agxcjIaODAgTExMYcPH545c6atre3HVrKUnpGREYCoqKgSe62srGRanjx5MmDAABcXl0mTJgGIiYmZN2/e2LFjix+hDWDu3LnSYQBUVFSKf660tLTUi69n+b+ePweA169L+ON/9arkz0RaGrp2RWoqbt5EQfKo8KKwsWMxYwZu3GBShoiIiIiIFEQZJGWeP38ukUikC0OK69+/f6tWrf7444/S30gxpaSkAKhevXqJvZMmTSo452jChAlaWlqLFi2SGZOdnb1q1ap/ddMWLVoAEAgEBWtwZMyePdvNzU1HRyczMzMtLe3ixYt79uyxtLS8evXq9OnTK1WqtH///n91xy9g3Dj8+iumTIGLS35L5coYNOj/T3RxwaNHEIuxbRt+/BGamhg5Env2fFhmUXxZhooKCrIHysqdxOIHADQ0NApKNRfQ1tYuvDxER0fn1atXAF6/ft2y8HlRH7dz587Vq1evXbu2SZMmBY1Vq1b98ccfAbi7u1tZWc2ZM+fYsWOfcrXP0KhRI1VV1RIrHBV348aN/v37N2jQwNvbW/rg69atS01NPXz48PH3q5ikWUJDQ8NHjx7VrFlTmvQpY99+i0qVcO4cvv++SPv9+4iLQ7FdVHj6FH36IC8PV6/iY+ktiQQSSQk7noiIiIiIiOSkDJIymZmZAKpUqVJi7549e8rlO5vC0NLSApCUlFRi77x586Qvjhw5cvr0aW9vb72CPRrv6ejoFGzp0tPTk1mCoaqqWnx3ifTIbUNDw2iZKizvGRkZSX/tVapUMTIyunr1qnRBTcOGDe/cuXP06NFffvml+FYU+VJWxpYtaNkSW7d+5vTJk7F0KU6exMiRst/l/5EB0PTTR7ds2VJJSenGjRuFC/2WSCgUzpkzZ/369WvXrp0+fXqJY4yMjBo2bPj48eNPD+Df0tHR6d2794kTJ6ZPn96sWbN/GLlr165JkyYNGDDgl19+KchPDRs2TCb95OXlJRQKvby8yvGvtrExxo7Fzp3o1Al9+uQ3xsVh7FiYmGDMGERH4+ZNODmhenX88QdGjYKdHY4fR+H1eqdOITIS06bl5+TWr0dmJkoqBkRERERERCQXZZCUkWYZCp/4U1j37t1LfwtFZmlpqaSkFBER8Q9jMjMz3d3dtbW1L168KD3pRppMWbt27evXr4cOHerh4fEZt3ZwcPjtt98SEhJkNo4VZmZmlpSUlJ2dLU0eAahVq1Zubm5mZqa+vv5n3LRcNWuG8eOxbNmnFmPNzERKCszM8t+KRBCJUHT7V9kzNzfv2rXr7t27hwwZUnhHT1ZWlrq6elhYWJUqVapUqfL27dshQ4bcu3fv+PHjfQrSCsCzZ89Onz7t4eEhTZMlJyc/e/bsn3Mlpbdu3bpbt261b99++vTpbdu2zc3NvX///o0bN65evWpqavr9999v3rx55MiRx48fd3Z27tmz57lz56QTGzdubG9vL1OyeufOnbm5ueVanBgAVq/G69fo2xctW6JhQyQlwdcXmpo4cwb6+rh5E0OG4Nw5XLqEH36AsjKUlfPr+0odOYJHj7B0KbZsgYMDXr5EQABGj/6kxVdERERERERfRBkkZerVq6eiovLs2bPSX6oiMjIyatGihbe39+rVq4uvgpHKyclxcnICUFDfV7oBJD09PSsr67NvPWLEiL17986bN2/Xrl0fG9OmTZu1a9ceP3585MiRAEQi0ZUrV2rXrq2AGRmpFSvg7Q3p7ykuDgMGYPx4DB2K8HCkpiI0FADCwmBkhOrVsXQpvL2xezecnJCSAi8vJCaivHMFAHbs2OHs7Ozs7NyvXz8HB4eUlJRbt26lpqaeP3++fv36CxcuHDp0qLOzc3R0dI8ePe7fv3///n3pxOHDh/v7+8+dO/fKlStdu3bNzc09cOBAenr63LlzyzXgmjVr3rt3b+HChZs3b164cKG6urq5ufngwYPz8vK6du1qb2//5s0bX19fAwODwMDA8ePHF0xcvHhxvXr1ZK6mra1d4unaZUxLC2fO4PJlXLqE6Gjo6WHlSgwdml8PuE0bXLuGxo3x4AFKzGmqqWHxYgwciJMnERmJdu2wfj3Ks6AyERERERHRv6UkKYt1Bd9///2ff/4ZFBT0f4uVCoVCNTW1JUuWFOzr+Q/w9fXt2LHjt99+u2fPnrp16wKIjY29cOHCDz/8MH/+/Nq1a48ZM6b4lA4dOty8efObb74pza0nTpy4ffv2bt26DR8+3NDQMDIy8vjx49OmTfPz89uwYUNCQoKWllbz5s2fP38+b948Kyur/fv3nz17ds+ePa6urqW5b1nJzYWGBtavx7RpHxp//RWjRmHGDEyfDhcXTJ+OsWPx/fd4v3oj3+LFmDgRbm44dSr/cCVNTXh64st8stLT07dv33758uWEhAQjI6N69eqNGDHC3t5+165drVq1kkgkCxYsKD5r0aJFbdq0uXfv3qZNm4KCggDY2dlNnz7dwcHhSwQNAMjKyipYNkVERERERERyVDZJmWfPnrVo0cLQ0HDVqlVOTk4ikcjf3//ixYvbt2///vvvbW1tV65cGRsbGxMTIxKJWrRo8eOPP/7www96enqWlpalv7siOHz48OTJk5OSkmrWrCkSiWJiYmrWrBkUFFSvXr2C05cKK6ukjEQi2b17988///z48WOJRGJqaurk5OTh4fHs2bPz58/v2rVLQ0MjISFhxowZZ86cyczMtLGxmT179rBhw0pz0zIkkSAyEkZGKLzGSNqop4fCFUvi4iCzqMjAAAYGAJCQgLAwqKigcWMw20BEREREREQVRdkkZQAEBARMnjz51q1b0rcqKirdunU7cOBAjx497OzsNm3a5OXltXDhwsJTunfvfvbs2Y9dUCQSKSsrl3jesyK4cOGCo6Nj4XO+MzMzr1+//urVKx0dHQsLi1atWqmoqOTm5iorKxcvqRsfH3/z5k1nZ+ePFUj+DAKB4B9q9+bm5n7OycSKyt8fGzdi+XLUrPn/Bx89ij17PrwVCJCR8eGtufndly8nZWVl5ebmykxMTU3V19cPDw8vk5ipzKSkYMUK/PQTFKGIeGQktm3D3bvIyICREdq3x48/ovj2wNBQbNyIgACoqsLREZ6eMDZGcHCRRWIFVq5E039RfJqIiIiIiCqoMkvKSCUkJLx8+VJbW9vc3Fxmi0ROTk52dnbhFjU1NW1tbbFYLHPeEIA7d+5MnDhx1qxZgwcPLsPwykpMTIy1tbWqqmpISIiJiUnhrry8vCVLlkydOrUMsy2lFBoa6ubmZmtru2XLFnnHUma++w43bsDDA+/Prfonq1eXXHVEyt7+akBA+4/1GhgYfOxoLZKbwYNx9Cjc3PDzz3KOxMcHvXvDyAiDB8PQEC9e4PBhGBjg6lVYWX0YdvcuOnRA3bro0weZmdi5E9raePQI6enYsaPIBQMDcfky/P3xBXe0ERERERGRvJRxUuZfeffunYeHR3Z29uHDh2W69u3b5+rqamZmFhISoq2tLZfw/sGoUaN+/fXXfv36HT9+XKZr3bp1M2fObNOmjZ+fn1xiK+7p06dNmjQRi8X+/v52dnbyDqcMnDyJfv1gbIywMHyktnIRr14hLOzDW1VVFD5kXFk5XSx+rqmpqaGhITNRT09PRUVFYYsif73CwtCoEQQC3L0LR0e5hZGYCGtrWFnB1xcF/0yFh6N1a5ib4949FCz027YNly7h5ElICyRfuoQuXbBtGyZMkL1mhw5ITcX7ytBERERERPTfJs+kTFRUVP369TMyMv78809nZ+fCXRKJpHXr1nfv3p0zZ87y5cvlFWGJAgMDmzZtqqqqGhwcbFX4f4YDSUlJVlZWSUlJFy9e7Ny5s7wiLM7d3X3Tpk1t27a9efOmwu4I+0R5ebC1RVgYfv4Zbm7yjobkZdYsrF2L1q3h5wd5faQ3bsS0abh/XzYxtGULpkyBnx8KnZheRE4ONDUxeTI2by7SHhQEOzscO/YljhAjIiIiIiIFILtv6EuqUaPGnDlzAEyePFkgEBTuUlJS2rp1q7Ky8rp160KlxyArjJkzZ4rF4qlTp8pkZAAsWLAgKSmpS5cuCpWRAeDl5WVqaurn53fs2DF5x1JamzcjLAz166PYkVb0NVmwANWr4/ZtHDwotxj8/KCpiWbNZNudnPJ7V67Et9+WMDExEUB+nerC1qxB7dro3bvMIyUiIiIiIsUkz6QMgJkzZ1pbWz958uTnYrUhmjZt6urqmpeXN2XKFLnEVqJTp05dvXrV2NjY09NTpiskJGTXrl0qKiqrV6+WS2z/QFdXd9myZQBmzpyZUbjIbUWTlIQVKwBg/Xqoqso7GpIjHZ38ekKzZiE19Qvd9M0bXL364W18PGrWLGGdjrk5AMTGIju75NgOHICKCvr2LdIYHY1jx+Duzk82EREREdHXQ85JmUqVKq1btw7A/Pnz4+LiZHpXrlxpZGTk4+Nz7tw5eUQnKy8vb/bs2QAWLVpUvM7IjBkzBALB+PHjbW1t5RHd/zFq1KiWLVtGR0cr2nawf2XBAiQlwcUFCrYUieRh2DB8+y3i47F0abnfSyDApk1o2BADB+avcwGgrIy8vJIHA1BVxeLFePRItjc4GEuW4Icf0LhxkfbNm6GhgdGjyzx2IiIiIiJSWHJOygDo3r17t27d0tLS5s2bJ9NlZGQkPUXb3d09JydHHtEVsXXr1rCwsPr1648dO1am688//7xw4YKOjs6CBQvkEtv/pcg7wj5RSAh27YKKCjZulHcopAiUlLBxI1RUsGkTQkLK8UZnz6J+fbi7Iz0dHTpAJMpvr1EDMTEQi2XHv3wJoOTT2kND0akTHB2xYUOR9vR07NoFNzfo6pZ19EREREREpLjkn5QBsGnTJg0NjX379t27d0+ma8KECXZ2dhEREdIFNXKUnJws3QG0bt06NTW1wl0ikcjd3R3A/Pnzq1WrJp/4PoFi7gj7dDNmQCDAuHFQyKVIJA/29hg7FgIBJk8ul+uHh+P779GjByIiYGODy5dx5AiqVs3vdXZGXh7+/FN21qVL+b0yLl5E69awssLZs9DSKtK1Zw8yMzFpUnk8BBERERERKSx5nr5UmKen54oVK5o2bXr//n1l5SKpIj8/v2+//VZDQ+PZs2e1a9eWV4STJ0/eunWri4uLr6+vTNfOnTvd3NwsLCyePXumrq4ul/A+UWJiYr169RITE8+ePdu9e3d5h/Mv/PknXFygo4OwMChw4ou+uMRE1KsHU00c/Bl2ZfeRzsrC6tVYuRK5udDXx6JFmDhRttpLZibq14e2Nq5d+/Ch9PdHu3Zo0wYXLuDpU0RHo0MHSCRYvRqenhgzBlu25B+MXUAkgrU1WrfGgQNlFj8REREREVUEipKUycrKatCgwatXr/bs2ePq6irTO3jw4KNHjw4YMEBehweFhoY2atRILBYHBAQ0atSocFd6erq1tXVcXNzx48f79esnl/D+lS1btkyZMqVu3brBwcEaGhryDueTiESiYcP8Tp92WrQIP/0k72hI0fj+hntu0K6GiU+hqlnaq0kkeHIUvx7B2rNQVoarK5Yvh7FxyYP//hvdukEgQM+eMDHB8+c4dw4NGuDSJZiYYMwY7NsHkQgjRuDAAVSpUmT5jIUFVq0CgGPHMGgQHj6EvX1pgyciIiIiogpFIbYvAdDS0lq5ciWAOXPmpKSkyPSuXbtWW1v7999/v3btmjyiyy/iO3bsWJmMDIDly5fHxcW1atWqr8xZKopKcXaEfbo9e/YcPfqdnd3AqVPlHQopIJdhqNoQKS9xq9QHn8U/xq/f4eQQVL6Awb1x9y5++eWjGRkAjo4ICcGcOUhNxYMHUFfHzz/j/n2YmADAhAk4fhwArK0xbhz69IGBwYefgvIxiYlYuJAZGSIiIiKir5CirJSRateu3bVr16ZOnbqxWCnXFStWeHp6NmjQIDAwUKakS3n7888/XVxcdHR0nj9/biL9rvXemzdvbGxssrOz796927x58y8ZVWkoyI6wT1Th1iKRHLy5jb1toaqBiU+gb/E5V8hJwfWF+Hs7xEJoGsFpPppPhtLH09ZBQXjwgIclERERERFRaSjKShmpLVu2qKmpbdu27fHjxzJdM2bMsLa2fvr06fbt279kSCKRaNq0aQDmzp0rk5EBMGvWrKysrGHDhlWgjAyAtm3bDhw4MDs7+6eKsBeowq1FIjmo2RqNh0KYDZ+Z/3quRIJHv2FrPdzbDAAtpmBqBFpM/WhGJiUFs2ejWTO4uaFiHmRGREREREQKQrFWygCYMmXKli1b2rZte/PmTSUlpcJdPj4+nTp10tXVDQkJMTU1/TLxHDx4cPjw4ebm5s+ePZOpwHL37t3WrVtraGiEhITUqlXry8RTVqKjo21sbDIyMi5fvtyxY0d5h/NRFXQtEslBZjy21ENuKoZehGXnT50VF4Cz4xDzAAAs2qHLZhg3/OhgiQQHDuCnnxAfD1VVuLpi2TJUqVIGwRMRERER0VdJsVbKAFi6dKmJiYmfn99xaS2GQjp27Pj999/37NlTVeYMlPI0cODADRs2bNiwQSYjI5FIZs6cKf1vhcvIADAzM/P09AQwbdo0gUAg73A+SroWaejQoczI0P9RuRq+nQsAl6dD/MkfaUE2YvyhY4Zev2K47z9lZB4+RJs2GDkS8fFwcoK/P3buZEaGiIiIiIhKQ+FWygDYvXv32LFja9SoERISUrly5cJdAoHgCxeU+ZjDhw8PHTq0WrVqYWFhOjo68g7nc+Tl5TVu3Dg0NHTjxo1TFbKCboVei0RyIMrDDju8C0HHtWg141NnPT0Bq65Q0/rogKQkLF6MbdsgEqF6daxYgeHDUXQdHxERERER0WdQuJUyAFxdXVu0aBEVFbVixQqZLgXJyOTk5EiXmaxYsaKCZmQAVKpUafPmzQAWLFgQGxsr73BkVfS1SCQHKpXQZQsAXF+M9JhPndWg30czMmIh/HdhY19s3gxlZUyZgpAQjBjBjAwREREREZUJRUzKKCsrb9u2TVlZec2aNc+fP5d3OCVYt27dq1ev7OzsRo4cKe9YSqVjx47du3dPS0ubO3euvGORdeTIkVu3blWrVm3WrFnyjoUqjjrtYdMLeenw31naS728hp32ODceKtfhMR5Pn2LTJlTYJCwRERERESkgRdy+JOXq6rpv377u3bufPXtW3rEUkZCQYGVllZaWduXKlfbt28s7nNKKiIiwtbXNzc29fft2y5Yt5R1OvpycHBsbm1evXu3du3c0Tx2mfyUlEi+uwt71nw60/mfpMfCdjccHAQkM6qLzRlh3L9MQiYiIiIiIAEVOyiQkJNSrVy8lJeXcuXPdunWTdzgfjB07dvfu3b169fL29pZ3LGVj3rx5y5Yta9q06f3795WVFWLx1IkTJ/r379+kSRN/f38FCYkUzptbSHiS/1rbBFUbwqCu7BiJGBE+iAuAshqqN4P5d0V606IQdh5p0dCqgrodUcUGYgH+3o4/5yMvHWpaaD0LbWdDVUP2skRERERERGVBcZMyADZt2uTu7l63bt0nT56oq6vLOxwAeP78eYMGDZSVlYODg62treUdTtnIzs5u06bNyJEjJ02apKKiIu9w8t28eVNdXb1FixbyDoQU1cXJ+Hs79M0hkSAjFsIc2Lvi+11Qev8ZznqLw98j5gFMmkCQiXchaDgAfQ/nD7i/BT6zoKwK4/rISYUwG/2OwHsUkiMAoOFAdFwL3RpyezoiIiIiIvoKfLmzpT/DxIkTd+/ebWFhkZaWZmxsLO9wAMDa2vr3338PDw+vuBmZd+/epaenm5iYaGpqSls0NTUfPnxYMCAnJychIcHU1LRcyyrfvn371q1bSUlJpqam3bt3r1OnjsyAGjVqnD59+uzZs1ZWVv369ZM5h4sIANS0MCUCAMQCXPXE7bWo0wG2g/J7b61G6mtMCEKV+gBwYzGuL0KDfmjQH+EXcXEKmrmh80aoqANAbipyUpERiyo26LwJdTvK6ZGIiIiIiOgrotArZQCkpKTo6+vLN4a4uLiHDx9mZ2fXrl3bwcFBZjdNVFRUfHx84RYrKytdXd0vG+Mn2blz55o1ayIiIgCoqal1795906ZNNWvWLBgQHh4+YcKEK1euAFBRURk6dOiWLVvK/FkSExP79+9/7do1MzMzExOTN2/eJCcnL1269KeffioYs3///vHjx+vp6VWrVi00NLRGjRrXr1/nGUxUxMXJCNyPOen5b4XIADSiAAAgAElEQVTZWK6N5pPReWN+i0SM9Gjovv+E56RglQFaz0KH1dj3LZLCMTUCqppFrhl9D6ZNoazQ2WoiIiIiIvrPUIhqHf7+/kOGDDE3Nzc0NGzQoMH8+fMzMjKkXQUZmb179zZr1szQ0LBevXqLFi3Kzc39AoGlpqYOGjSoevXq3bp1GzRokKOjo62tbUBAQOExixcvblaUNOuhaObNm+fm5tawYcPLly/7+/vv378/KCjIzc2t8JhevXqFhYX99ddfCQkJx48fP378+IQJE8o8kkGDBt26devAgQNv3rx58OBBTEzMokWLYmJiCvKDr1+/Hj9+vLOz86tXr4KCgvz9/ZOTk6dMmVLmkdB/ijAHEglUNRBxGXfWAYCS8oeMDABlFQAQCyDIxJvbsOkJVU0Is5EWBYkof4xZC2ZkiIiIiIjoi5H/1w9vb++BAwfWqFFj7Nixpqam4eHhW7duvXXr1p9//lkwZuvWrZMnT+7Zs+f06dP9/f2XLl364sWL3377rVwDE4lEPXr0uHPnzvr1611dXXV1dYOCgn788ccDBw7Y29sXDIuLi3Nycipc9FdPT69cA/sMoaGhK1as6NOnz4kTJ5SUlAA4ODi0b99eQ+NDBdOoqKgnT54sXbq0bdu2AHr37t29e/dLly6VbSR37tzx9fWdM2fOsGHDpC0qKiqenp6Fx5w/fz4vL8/Ly0u6wapRo0YjRozYtm2bIiybIkUkESMlEr6zAQnqdkDQEQTuQ6sZssMifADA/Dskv4BEBLEIv7XHy+uQiKBbA122wKbXl4+diIiIiIi+ZnJOyqSmprq6utarV8/Pz68glzFixAiBQFAwJisra968eU5OTidPnlRRURkyZIiamtqqVatmzJhhZ2dXfrGdPXv25s2bq1atcnd3l7Y0atTo6tWrMiWHo6OjGzVqZGBgUH6RlN7vv/8uFovnzp0rzchIVa1aFUBubm5GRoaenp6WlpaKioq/v79YLJZu0YqNjTUxMSnbSKR7owYOHFi8KyIi4vfff+/fv396ejoKLZIC4ODgIBKJIiMjC6fDiJCXgcXvP9KVdNBpAyxcUK0xvplTwsirc1GjJax7IOY+AAQdxjeeaL8Cojxc8cCJgXALzK8+Q0RERERE9EXIOSlz4sSJlJSU/fv3F15dYmNjU3jMX3/9lZqaOm7cuIKDgVxdXVetWnXu3LnyTsooKSmNGzeucKM0IxMZGRkaGurk5KSpqRkbG9upU6fyC6NMBAUFKSsr29raFu86fPiwq6vrnTt3WrZsOXbs2B07djg5OU2dOvXly5eBgYFnzpwBEBERceLEiby8vMzMTJnpGRkZhTNoAMRicWpqqsywnJyc7OzsS5cuvXz5EoCFhUXxSJ49e+bp6Wlra9usWTNpYIsWLQKQm5t79+5dANHR0UzKUBGqmhh8BkpK0DGDgUV+yV4tY2gVrQsuysXxAchOxMBTUFKCVhUAaLcULfPzrei1H1usEHwM3y36ovETEREREdHXTc5JmQcPHgAo8djjCxcuuLm5HT58+NWrVwAK51+sra01NTWl7devX09MTBSLxYXnCoVC6WqLwrKysopXoklJSZEpdSwQCCpXruzl5fX8+XNjY+MS98scP37cw8MjPDzc3Nz87du3586dO3z48Nu3by0tLWfOnDlixIhP/w18GampqUZGRpUqVSreZWdnN2fOHDMzMwCzZs3y9vYODg7u378/gGHDhjk6OgIICwubPXt26cMQiUQCgUBZWVlLS6t4r4uLS2RkZLVq1TQ0NPr377948WIfHx9DQ0N/f39tbW3p9NLHQP8pyiqo0/7/jMl6i+MDEReAEb4wbgAAOtUBIC3qwxi9mgCQnVROYRIREREREZVIzkmZxMTESpUqVatWrXiXrq5uw4YNK1euHBcXB8DIyKhwb5UqVWJiYgD07NkzLS2tbKOytLT08vLKysqSuWmB3r17W1lZmZiYSCQSLy8vbW3tJk2apKen79ixY+TIkSKRaPTo0WUbUilpaWmlpKQU7EsqzMHBwcHBAUBcXFzbtm1btWp19OjRiIiI3bt3b9myJTQ09NatW3Xr1vXw8KhUqVLxc6m1tbVlTs5WUlIqnsnS0NDQ1NRUVVU1MjISi8XR0dG1a9eWGaOpqWlubi59ffTo0T59+ty4cUNXV3fSpEmRkZETJkwo8XNC9E9i/fF7P6hUwg+3P2xNUtWEqQNCvNF+ZX5Z3/DLAFCtsdziJCIiIiKir5KckzJqamoCgSAvL0+mUAuAtm3bXrx4Ee+rkMgscsnJyZEmCFxcXJSUlAp2NkmpqKgUP8hZU1OzcF1bKT09PZk8hZqamqmpKQAdHZ3IyMgSw7aysrKyspK+LlyktlOnTo0bN167dq2iJWWsrKwEAsHr168Lsh7F7du3LzY2dt26derq6g0aNFi/fn39+vXHjRvn4+PTrVu3lStXlkkk0gSQv79/8aRMYcrKyoMGDRo0aJD07ejRo9XV1Rs1alQmMdB/3IWJePQb5qTj4W5cmARVDXw7D6/+wqu/AECrCur3gdMCHO2Nw93QZDQyE3BjMYys0WiwvEMnIiIiIqKvi5yTMrVq1ZJIJK9fvy7IcRQn3VkTHx9f8DVeKBQmJSVVr14dwKlTp8opNltb2xs3bsTGxkpzNJ9CVVW1SZMm0josCqVr165r1qzZu3evl5fXx8ZI68UUTmbVqVMHQPECMaXRq1cvfX19Ly+vTp06FV93U6KXL18eOXKkX79+nzievhaqmtAo6TQuUwfkZQDA4wNQ0wKAv5Z/6K3WGPX7oF5P9DuCG0twcggqacO6Ozqtgxo/YERERERE9EXJOSnTrl27lStXnjp1ysPD42NjGjRoAMDX17d58+bSlmvXrolEohLL1pah/v37b9u2be3atevWrfuHYW/fvjU2zq8qKhaLg4ODpbkMheLk5NSpU6eVK1dWrVp17Nix6urqYrHYx8fHwMBAIpEcPHhw5syZ0pOwvby81q9fr6qqmpWVtX79+kqVKn377bdlGImuru7OnTuHDBni6Og4adIka2vr2NjYS5cuWVhYtG7dulevXt7e3t26dduxY4eqqqqpqWlwcPDWrVurVKlSVkt16L+j7WwYWZfQbv8D7H8AgFE3SujNTYUwB6oaaDgQDQdCIkGhI8mIiIiIiIi+KIlciUQiR0dHbW1tHx+fgsa4uLiwsLDIyMjdu3dHR0dLJBJ7e3szM7OXL19KJJKcnBwXFxd9ff2kpKTyDm/w4MEAJk2aFBgY+Pbt2wcPHnh6evr4+Bw4cKB58+ZRUVH79+/X19dfv359YGBgcHDwqFGjAOzfv7+8A/sMKSkpvXv3BqChoWFhYaGurq6iorJ8+fK9e/cCuHPnjkQi8fT0VFdXNzExadmypaGhoY6OztGjR8sjmBs3bnTu3FlavtfExKRjx47nzp0LDAwcNWpUYGCgWCyeOHGidIFS9erVXV1do6KiyiMMqqjEYkngr5I1VSWLIAm/9C8mhp6RrKkmub643CIjIiIiIiL6F5QkRc8e+vJev37dvXv3oKAgOzs7S0vL6Ohof3//kSNHdurUqX///j4+Ph06dAgMDHRxcZFIJG3atAkODo6JiTlw4MCAAQPKOzaBQLBw4cJt27YV1BK2tLRcs2ZNamrqhg0bzp07p6en5+Hh8euvv2ZlZQGoVq2al5eXzCnaCiUsLOzevXvZ2dmmpqbSVJdEIhEKhaqqqkpKSgCioqLu378fGxtbtWpVFxcXQ0PDco1HIpEofXydQk5OTvEyQPS1i7qDC5MR6w8AFi7ouhVVbD517ptb2PsNVDUw8Sn0zcstRCIiIiIiok8i/6QMAKFQePbs2du3b2dkZJiYmDg4OHTp0iU9PT0sLMzGxkZa5SQ+Pv7gwYMREREmJiYDBgywsfnkr2FlEd7z58/z8vKqV69etWrV4gNycnIiIiJUVFSsrKxkSg7T50lJSZkxY0ZAQMDff//NXynly07EDS/c3wqJGDpmcFkOu39//PypYQg6hPp9MeBEOYRIRERERET0LyhEUoZIRm5ubqNGjcLCwrZt2zZhwgR5h0PyJhBg/y4kzkNuClQ10HoW2s7OL+L7b2XEYWs95KZh6EVYdi7rQImIiIiIiP4F5f8/hOiLU1dXX7NmDYB58+a9fftW3uGQXF27BgcHjJuE6Eaw7o4JT+Ds9ZkZGQDaJvhmLgBcng6xoAzDJCIiIiIi+reYlCEF1bNnzy5duiQnJ8+bN0/esZCcvHqF/v3Rrh2Cg2FtjZ5zMfgsDEp9ullLdxjVw7tnuLelLKIkIiIiIiL6TNy+RIorPDzc1tZWIBDcuXOn4EB0+irk5eHnnzFvHjIyoKWFWbMwZw7U1cvs+i+u4EBHVNLB5FBom5bZZYmIiIiIiP4NlUWLFsk7BqKSGRoapqen+/n5PXnyxNXV9R/OaaIvTShEaCjevIGyMrS1PzosMxPPnyMrC7q6kPnjS0xEWBhEIujoyHaJRHBwwG+/QSDAkCE4cwY9ekBVtSzjN6iLuEAkPEbWO9j0KssrExERERERfTJuXyKFNn/+fDMzszt37hw4cEDesRAAIDcXM2fC0BANGqBpU5iYoGlTXL8uOywvD+7uMDZGo0awsECdOvDxye9KSkLPnjA2RpMmqFED9vYICCgyV0UFAwbAzg43buDQIZiZlcuDdNkMNS08+g2v/yqX6xMREREREf0/TMqQQtPW1l61ahWAn376KTU1Vd7hfPUkEvTrh82b4eGBsDAkJODqVairo2NH+PoWGTltGvbswdatiImBvz8MDNCrF6KjAWDyZFy9Cl9fCIUICYFQiJ49IRLhxQv89lv+T5cu8PfHN9+U47Po1ULrmYAEF6dDLC7HGxEREREREX0Ea8qQopNIJM7Ozjdu3Jg+ffq6devkHc7X7eRJ9OuHHTswfvyHxqwsODhALEZICJTf53n9/JCeji5d8t/6+qJDBxw8iKFDYWaGpk1x5kx+18aNmDYNISHIy8PZs/mN33xTvhkZKUEW1vbAkWhMdC/yRERERERERF8EkzJUAQQHB9vb2wN4+PBho0aN5B3OV6xfP1y9ioQEqKkVad+yBVOm4N49mJlBKETt2rITQ0NhY4O1azFjBho3Rmoq/P1RpQoATJmC3bsREwN9/S/0FIVJ00yGhnj+HEZGcgiAiIiIiIi+Yty+RBWAra3tuHHjhELhtGnT5B3L1y0oCJaWshkZAA0a5PeOGAEXlxImPnoEAPXrA8DChYiJQaNGWLoU+/Zh715s3y6fjAyAvn3RqROSkrBwoXwCICIiIiKirxhXylDFkJycbG1t/e7du5MnT/bp00fe4XytTE3RtCnOnZNtf/oUDRti3TrUqoW0NLi6FukVCNCqFQQCPHwIFRW8fYsePfDyJVJTkZ0NBwccPJifr5GLZ89gZwexGA8eoEkTuYVBRERERERfH66UoYrBwMBg6dKlANzd3TMzM+UdztdKWxvv3pXQnpAAADo66NdPNiMDYNo0BAVh+3aoqEAoRIcOyMlBSAji47F3L+Li0KoVwsLKPfiPqV8fkyZBJMKkSWCSmoiIiIiIviAmZajCGDt2rKOj45s3b1avXi3vWL5W9esjIqKEs4qePweAhg1l24VCTJqEnTtx9CjatAGAa9fw6BE8PaGnBx0djB6NmzeRmop9+8o/+o9bvBimprh1C0eOyDMMIiIiIiL6yjApQxWGsrLypk2blJSUVq9eHRkZKe9wvkp9+uDdO3h7F2kUi7F7N2rWRPPmRdoTEtCxIw4cgLc3evfObxQKARSpSmNsDCUlZGWVa+D/h44Oli8HgBkzkJYmz0iIiIiIiOhrwqQMVSStWrUaPnx4Tk7O9OnT5R3LV2noUDg6YswYnDiRn16JjcXQoXjwAGvXQlUVy5ZBWoz57l00bYrwcFy6hDZtkJyM5GRkZqJ5c1SujOXLERsLAHl58PSERIJu3eT5XABGjsQ33yAuLj87Q0REREREVP5Y6JcqmPj4+Hr16qWmpl64cKFLly7yDufrk5iIH37A6dPQ0oKODt6+hYEBNmzA8OEA0KMH3rxBQACqVEFiouzcvn1x4gTOn8eECYiLQ82aePcOOTlYtgwzZnz5R5H18CGaN4eKCh4/Rr168o6GiIiIiIj++5iUoYpn3bp1M2fOtLS0DA4OVldXl3c4X6WXL+Hvj4wM1KyJ1q2hoSE7IDAQIpFso4EB6tQBgLw8+Pvj9Wtoa6N1axgYfImYP8W4cfjlF3ToAB8feYdCRERERET/fUzKUMUjFArt7e2Dg4NXr149a9YseYdD/yFJSXB2xowZGD4cSkryjoaIiIiIiP7jmJShCunq1avt27fX0dEJCQmpXr26vMOh/6h375CRgapVoaX10TESCeLioKUFPT3ZrpwcxMWhWjVoapZrmEREREREVEGx0C9VSC4uLkOGDJk8ebJe8W/CRKUkkWDjRlhYwNgYFhbQ1cV33+HOnRKGrV0LExNUrw59fTRsiOvX87vS0zF2LPT1YWEBPT0MH46MjC/7DEREREREVAFwpQxVGDk5OTExMRoaGv+wNCYnJ0ejWH2T9PT0t2/fVqtWrXLlymUZ0MOHuHYNb99CTw8tWsDZueQNL0Ih/vgDGRkYNapI+4sXOH8e7dujfv2yjIpKb8IE7NiBH3/E0KEwMsLTp1i6FE+e4Px5uLh8GLZsGRYtwoIF6NULiYmYOhVhYXj2DLVro29fnDuHlSvRpg38/ODpiV69cPQoYmIQFJQ/3coqv8IOERERERF9rZiUoQrg6dOns2bNunLlikAgAGBubr5kyZJhw4YVDJBIJJcvX16/fr2vr+/mzZsnTZokbU9JSXFzcztx4oRIJFJTUxs0aND27du1tbWTk5Nnz55dMD0vLy8zM1P6urKq6j7pYc9ZWcjNlQ3l/HlUqoTMTIwYgVOnYG4OS0tER+PZM7RogVOnUDhhlJ6OvXuxYQNevUKlSh+u5u+PTZtw5AiEQqxYgUKRkPz5+eGbbzB/Pry8PjRmZsLRETk5CAuDikp+45UriI3FiBFFJu7cCRcXWFpi5kysWZPfNWcOVq1CeDjevcOhQ/mN3bqhY8cv80xERERERKSYVOUdANH/ERIS0qZNG01Nza1bt7Zs2TItLW3//v2urq5NmzatX78+gOzsbEdHx8jISCcnJ4lEklsokzJy5MgLFy5s3brVxcXF19d36tSpIpHo0KFDWVlZu3btKvF2+hoa+3JyPhqNUIhKleDmhnPncPgwBg/Ob795E717o2dP3L374Uv7nj04eRKTJiEoCEeP5jfm5mL4cLRti6NH0a9fqX89VNYOHEClSpg5s0hj5cqYNg3jxsHPD02bQiiEvj46dCgyxtISAKKiEB4OAK1afejq3h0rV+LvvzFwIJo3L+8nICIiIiKiioJJGVJ006dPz8vLu3//vpWVlbSlbdu2np6edd5v/dDU1Ny2bZujo6NIJNLV1S2Y+OjRozNnzsyZM8fNzQ2AlZVVaGjo5s2bvby8TE1Nd+7cWTBSTU1NW1tb+lpFSQnS5WNaWih+3ra6OkJDcegQPDw+ZGQAfPstNmzAyJE4cwa9e+c3urvD3R0APD2LXOHpUwDIzi7dL4bKR2AgatVCoQ9SvsaN83s3bsStW0hIkB0g/WO1soKJCQA8e/ahS18fAKKjyydiIiIiIiKqqJiUIYWWkpJy+fLlESNGFGRkpKQZmXfv3lWqVElXV9fJyQlAenp64THXr18HMGTIkIKWwYMHb9q06dq1a2PGjBk3btxnxnT5MiSSEha59O+PUaNw6RJq1sSlSxg/HsbGn3kLkqOUFJialtBuZgYAycno2xfNmsn2Sov+1qiBPn1QuTJat8bKldDURP36ePQIGzYAQFJSOYdOREREREQVDE9fIoUWHBwsFottbW1L7K1Tp8748eM/NvfNmzcALKWbSoCC19GlXLAQGQkAFhay7ZqaqFYNL17g7l3Mn4+4uFLdheRFS6vk7Mm7dwCgrY1hwzB3rmzvmjW4dAmbNkFaTNrbG927Y/589OuHixexYgWAEs7MJiIiIiKirxtXypBCS0tLA/Cx45Y8PDysra0/Nvft27daWlqFD2MyMDBQUlKKj48vVUwCAQBoapbQpa2NvDyMGoWePfP3sFCFU68eLlyAUAjVov88SivF2NiUMGXNGsyejU2b0KdPfkvVqjhyBADEYigrw9v7o3OJiIiIiOgrxpUypNA0NTUBJCYmltg7d+7c/v37f2yuvr5+dna29MAmqfT0dIlEYmRkVKqYpJuSoqJk2yUSREWhalVoa6NmTaipleouJC89eiA9HSdPyrbv3w89PXz3XZHGrKz8hTM//4zJk0u4mrJy/tzKlWXnEhERERHRV49JGVJo0oUw4dJFCv+SmZmZRCKJiYkpaJFuXPrYuptP5egIAA8fyrYHByMnh2frVHgDBsDRERMm4OLF/JasLPz0E86fx8KF0NbGL79gzhwACA9Hq1a4fBmXL0NmG118PFJTASA7G+vX4+xZzJwJHZ0v+yRERERERKTouH2JFJqZmVmTJk1+//33JUuW6PzL77SOjo4Azp49O2nSJGnL2bNnATRv3jwvL2/y+3UNubm5WVlZhSceUVVVSU6GUFjkcgIBMjJw4QI6dECtWli2DN275xcQASASYd48aGlh2LDPeExSIKqqOH8eQ4eia1dUrQpjY0RGQizG0qX5Z2lduIBbt7BiBbp2RVgYALRr92F6hw7w8YGHBw4fRtWq+aWFpk/HggXyeBgiIiIiIlJoShLp6b9EiurixYvdunX77rvvdu3aJa3UGxkZeeXKlXHjxk2fPt3W1tbV1VU6Mj09XVdXd+3atTNmzAAgFoubNGkSHx/v6+vbqFGjoKCg9u3bW1pa3rp1Kzc3t3CtGRniypWVMjNL7nv1CrVq4fp1dO2KunUxbRrq1EF8PHbswM2b2L0bo0dj+3ZMnoxHj2Bri127AMDbG76+2LYNAPr0wcuXePgQAgEmTULv3ujcGebm6NixbH9vVFqPHuH+faSno3p1uLh8OEtLJIJEAlVV+PkhJ0d2lqEhHBwgEOD6dbx+DR0dtG2LUi7OIiIiIiKi/ygmZagCOHjwoLu7e2JioqmpaV5eXmJiorW1dUBAgImJSbdu3Y4cOeLp6blCesDNe/369Tt+/HhISEi3bt1evnxZtWrVhIQEa2vrixcvmpubi8Xi3bt3S0eqq6traWkVnttHWVlFS0u2KIyaGrS1YWsLdXUAePwYixbhyhVkZEBdHW3awNMTLi4AcO0aDh3C4sWoXh2Fzn7Kd+YMfHywdWuRxg4dsGNH6X9RREREREREVIEwKUMVQ1ZWlp+f3+vXr3V1devUqdO0aVMlpfxPr5KSUnh4+MuXLwuPr1q1auPGjQHk5ubeuHEjKiqqVq1aTk5OamVefzcrC0VzOkRERERERESfgkkZIiIiIiIiIiI54OlLRERERERERERywKQMVUhz5swZO3ZsnPRoGyIiIiIiIqIKiNuXqOKJjIysX7++QCC4d+9es2bN5B0OERERERER0efgShmqeH766afc3NwRI0YwI0NEREREREQVF1fKUAVz586dNm3aaGhohIaG1qxZU97hEBEREREREX0mrpShikQikbi7u0skEg8PD2ZkiIiIiIiIqELjShmqSA4cODBixAgzM7PQ0NDKlSvLOxwiIiIiIiKiz8eVMlRhZGdnz5s3D8CyZcuYkSEiIiIiIqKKjkkZqjDWrl37+vVre3v74cOHyzsWIiIiIiIiotLi9iWqGOLj462srNLT0319fV1cXOQdDhEREREREVFpqco7AKJP8m7zZhM1tfa9ezMjQ0RERERERP8NXClDFUFgIJo2FRsZxd++bWppKe9oiIiIiIiIiMoAa8pQRTBrFsRi5REjmJEhIiIiIiKi/wyulCGF98cf6N0bhoYIC4OhobyjISIiIiIiIiobXClDik0ggIcHACxezIwMERERERER/ZcwKUOKbetWPH8OGxuMHy/vUIiIiIiIiIjKErcvkQJLToaVFRITce4cunWTdzREREREREREZYkrZUiBLVqExES0a8eMDBEREREREf33cKUMKarQUDRqBLEYAQFo1Eje0RARERERERGVMa6UIUW1aBEEAri6MiNDRERERERE/0mq8g6A6CM2b0bVqvD0lHccREREREREROWC25foE7x5g4gItGkDNbUPjdnZuHcP9erB1DS/RSRCYCDCwqCqChsbNGwIJaWSL5iWhocP0awZtLU/NL58iUePoKSEJk1Qq1a5PQwRERERERGRQuD2JfoER4/C2RmpqUUao6Ph7Ixz5/LfXr+OevXQrBnc3DB6NBo1goMDHj+WvVRsLBYtgrk5nJ1x6VJ+Y24uxoxB3boYMgQDBqBuXXh4gOlCIiIiIiIi+k9jUobKQmAgunRBtWoICkJKCtLS8NdfyMqCszPevPkwLCoKderg5k2MHl1k+ooVOHIEv/+OjAykp2PMGKxejTNnvvBDEBEREREREX1JTMpQWfDwgK4uzp+HrS0AKCmhbVucP4+MDCxZ8mFYjRqIi8Off8omZaZNw5Ur6NsXSkpQU8OKFQDg5/cFH4CIiIiIiIjoS2NShkotPR1Xr6JvX+jrF2m3tES7dvjjDwgEOHMGz54BgJ5eCVfQ00Pr1h/eamtDVRUpKeUZNBEREREREZGc8fQl+mQ7dqBy5Q9v373LfxEeDpEIdeuWMMXaGpcu4fVr9OwJDw+sXPlJN7p3D0IhmjUrdcREREREREREiotJGfpkBw5AReXDW4Eg/0VmJgAYGZUwRdqopIRz52Bp+Ul3EYsxfz4sLDBiRKmiJSIiIiIiIlJsTMrQJ7t1C1WqfHgbHg4rKwD5u5ZiY0uYEh0NAMbGqFPnU+8yYwZu3cK5c9DULDZG62gAAATQSURBVFW0RERERERERIqNSRkqNWtraGri6dMSup48Qd260NH5pOsIBJg6Fbt24ehRdOhQtjESERERERERKRoW+qVSq1QJ/frh1Cm8eFGk/fZt3L6NYcM+6SIJCejUCYcO4dQp9OtXHmESERERERERKRQmZagsrFgBHR20a4cTJ5CQgJgY7N2LHj1gY4NZs5CeDnt7bNkCAJGR8PfPX1bz4gX8/REVhZs3YWeHgADs2gUzM/j7w98fT57I95mIiIiIiIiIyhW3L1FZMDPD7duYPBmDBkEkAt4vn9m4EZUrIz0dampQVgaAefNw+HD+LA8PAJgxA7dvIy4OAAYN+nDN+vVL3hJFRERERERE9J+gJJFI5B0DKbzcXGRnQ08PSkofGsVipKVBSwuVKn1oTElBRARUVVG3LrS1S7hUZiby8oq0aGhAKIRQKDtSRQW6umX1BERERERERESKhkkZIiIiIiIiIiI5YE0ZIiIiIiIiIiI5YFKGiIiIiIiIiEgOmJQhIiIiIiIiIpIDJmWIiIiIiIiIiOSASRkiIiIiIiIiIjlgUoaIiIiIiIiISA6YlCEiIiIiIiIikgMmZYiIiIiIiIiI5IBJGSIiIiIiIiIiOWBShoiIiIiIiIhIDpiUISIiIiIiIiKSAyZliIiIiIiIiIjkgEkZIiIiIiIiIiI5YFKGiIjof+3YsQAAAADAIH/rYewpjAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABlIGAAAAYCBlAAAAAAZSBgAAAGAgZQAAAAAGUgYAAABgIGUAAAAABgHmNEO6bS5M6wAAAlt6VFh0cmRraXRQS0wgcmRraXQgMjAyMi4wMy4yAAB4nE3RT2jTUBwH8JeXNH39t7b507y224wONKAg8x/Ogwn+GQWlgqKTHTS3RRgiw4Mo0rrJYEJREEVwiiKKB8cY60EvaXD1MDrEwxAEnYiVMWSKG4K4gya/1tQHj3zye9/33i/kuz2xgNwRQ43BuLPTnevcWWT4nOU+WZahqMX/qpjPaR45yvriWssBnjSLPGkI0yBPTFcYU9KQmwv5ubCfixACxzAMjQYbRZbGgs0ipm2N3QxL4/7ViWCuGUy2mhD8ZdGX5EtuXZPiyb89it8kJUT1RNNxQIBm/BazRNvg6g9D2311RBBjYQabmLUwy5lcwMIcr/JBDfMEkRAKhVE4ouFwVA3FENumBpAZT6BE0sJJwRREC4uSKckWllOqpKgiRTSN0hmUyRKcaUeZDiQnzJiiYQnzKMCxmOETsiQKSWWIafw9GJ2li58dqettxXvJ1teczkMztufefN35MTeoe/6ULzvv7p8DH/ty1ymyZchol6Yddc887P1Q63OmTj+DevcDzXm67wLkK782Oj3WG8gsr5YqBwuvwbVdU/bsYo/j+ePkTf3I6m5w+PG1yo6Xo+D+IdsW6T04M/pi3DlRjRuet8+M6cPL4+CzZ77qC9sGwMXp48bJTRY4v3/E6C7I4JX+PqPr1Qj0c/T5N33r+ytgRSjpe2+LkFmaqNnVhwJ4zbyhHzh/C+4NKleNJ1tGwYOpsjGZvAz+XZgzhleq8C2x+Z/G9Tu94MXNj4xTh0Xof2ypbgys3wlO/QVZEZ0kPMpVVAAAAv56VFh0TU9MIHJka2l0IDIwMjIuMDMuMgAAeJx9VUtuGzEM3fsUuoAH/IrSIgvHTtOijV20ae7Qfe+PknJGYoCi4wwxEp4o8j2SOZR4fly+/v5T5kOXw6EU+M9f7728MQAcXkp8lMen5y/Xcn49Pe4759uv6+vP4l+MfsZ/H7Gn19vLvoPlXI62ITRDKEfcqlEXLbDBeNZZuiMdp13LETZuKEwLiTuS332yqIanTgqQgLQDJYC6VZZKPYDcO8kC8g7UAErcbSLjbkVtfSFlR9Z3l55Pk5GPNsg+dUdaIHljw4Zt+ORaa1vIuiNbuZUjRZhmFmHWbtUW0HZgD5cUl3cnJi4n940L2SbtEFB03juADSgyacqoT+iQyMOTioThv4MHkomfGiF5qO5VuYEGtLXeFBJ0ioQcWeEGjbFSsEuIBCktnDqhlOsIwbDXOr6gtswqTqlQA+tadQuFaCOprIlWnFphiOURWCVzCT0v158TcmqFNpCNO3pFsZNF0nJaUyuX8uoA5g5OvCPJBJICOMXCUEtcgZET+RFtLcc5xaIQy31W6VQj4OiQnPzUivAeZ9WxvwGK5hah1U00kAjOusbtUCG3HU2haAjllHdrXgtOaHehEk00dSK5OyXrFeKMQ53UBJ0ykd6zp+E0oJUg1x9NmTznW9Fg0tgGVAw5MUVTJ7LyvVQPFSqNBvBCrJropykURVfZ5mWEPMaJmVrNDEylfDZEAD76vF2iqhm8wmuCtnL+/PxwxPJ2+vaw5hCMKxyr0EY3+MGahHPH/zr3dL18mJH3qfl4u17W1BR/F+++pDWt2Je8RpKMn65Z5os1XLygfGlrKpG/bY0e9GVfAwbjhTRGcBhM04LCpOalwGBuUX4/JanBR0qaulgitNSqOIyllpQwq0289STMagdvMQmTyj7ooA/lHRe7SVUcAbtJxRoBB0ZSpePwo6mgMUyqsIjZTSqkYDdMS1WIY6enYhs7DKmMZAQ1/fAgWUa/fXp+mJFH4eQyifX+j9m/D38BkZ16ef4fIH8AAAJOelRYdFNNSUxFUyByZGtpdCAyMDIyLjAzLjIAAHicbZQ/bxsxDMW/SkcHOAsiJYqkDhmKLF6aZDduSItudR0E7ZYP30ddt9NiQT/xz3s8S9cfl543umKhLX4Zv71sp+vzhTvV7QHbiu1Tb9vp8frSFegJZ7Zh6T6OKO9nRDh8QSnG+typbI8RxCWKct44FkIKYphHbWpR4VI6Cfa/O+mQQRYp5EiJZlw3lGfZrq+d/wthKEGlc2eLrlg9Kl66bPTl83TWRNmUljOlpuyyrEAAbTnnVIxq4YGoVJElJ2fJOZCkVmorgYo7V6CazFV84cQ4aTtR0jpqCYk5WElFyXiw0lozMI5ialGsuTbdUTYvQxgbRRRBq+ddKxWOBqhRGzEYJ89VkYkoKRbAzE1GXrYSjmpiiuCRp+Rtd5mbhX4ksvpA7OS5BLImC/KpythSRgUYRE6uex1XGwg5xCOlOMlSoJGrBcjaWJcKgZjUsmIAxbNFBGM2a0VVco2qZYwIAa3COjLZq4wIRpehrHHGiFeJ5H2yXJVgeG3QkhvHeDAA6F41wRmNIFXRQJKyQ2iMp2SSGkGYlwxiQG3/tvhT2Bi0GD73w/L25357/bi/95xu919fsfv29v789/b950fKnY6QOh8h93KEpddJeu1ypLW3I2xdj1C7HaF1P0LHxZz0h6mZK8K1nWC8AhNnhAs78cZxjycYLSf2GC0nBhktJxap4T2YYMHLMsGKl2aCDQ/PBDseqYmSipdngtFy5rLhbZpgKJm5hJKJS4aSiUv5/Ac0t5AljP9qVwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N = int(input())\n",
    "cm = Chem.CombineMols(m,m2)\n",
    "em = Chem.EditableMol(cm)\n",
    "em.AddBond(N,len(list(m.GetAtoms()))+8,Chem.BondType.DOUBLE)\n",
    "nm = em.GetMol()\n",
    "Chem.SanitizeMol(nm)\n",
    "rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE\n",
    "Draw.MolsToGridImage([mol_with_atom_index(nm)], subImgSize=(500,500))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "aa2313ce-f08f-4c5c-a755-b14078dfed26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nm = Chem.MolFromSmiles(aa+plp)\n",
    "nm=Chem.AddHs(nm)\n",
    "# run ETKDG 10 times\n",
    "cids = AllChem.EmbedMultipleConfs(nm, numConfs=1)\n",
    "# print(len(cids))\n",
    "rmslist = []\n",
    "AllChem.AlignMolConformers(nm, RMSlist=rmslist)\n",
    "AllChem.ComputeGasteigerCharges(nm)\n",
    "pdb = Chem.MolToPDBBlock(nm)\n",
    "print(pdb, \n",
    "      file =open(file, 'w')\n",
    "     )  \n",
    "\n",
    "\n",
    "parser = PDBParser(PERMISSIVE=1)\n",
    "H, A =[], []\n",
    "structure = parser.get_structure('f', file)\n",
    "\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        for res in chain:\n",
    "            for atom in res:\n",
    "                if 'H' in atom.id:\n",
    "                    H.append(atom.id)\n",
    "                else:\n",
    "                    A.append(atom.id)\n",
    "cmd.reinitialize()\n",
    "cmd.load(file)\n",
    "cmd.select('PLP', f\"name {'+'.join(A[-15:] + H[-8:])}\")\n",
    "cmd.select(\"LIG\", 'not PLP')\n",
    "cmd.alter(\"PLP\", \"resn='PLP'\")\n",
    "cmd.alter(\"LIG\", f\"resn='{lig_res}'\")\n",
    "cmd.save(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3273a57-25b0-4b6a-88e4-6ea739ae8462",
   "metadata": {},
   "source": [
    "## Get atom numbers for reactionsNeighCAlpha"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "abf329aa-cd23-4125-b99f-20c847ce249d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[19, 3, 22, 37]"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conect.loc[CAlpha].iloc[0].dropna().astype(int).values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "a33b3770-5d29-4ada-be8e-33093cdb5305",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C14', 'C16', 'C17']"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppdb = PandasPdb().read_pdb(file)\n",
    "df = ppdb.df['HETATM']\n",
    "NAlpha = df[df['atom_name']=='N1']['atom_number'].iloc[0]\n",
    "conect = pd.read_table(StringIO(''.join([x for x in open(file).readlines() if x.startswith('CONECT')])), delim_whitespace=True, header=None).drop(0, axis=1).set_index(1)\n",
    "Neigh = conect.loc[NAlpha].dropna().astype(int).tolist()\n",
    "CAlpha = df[(df['atom_number'].isin(Neigh))&(df['residue_name']!='PLP')]['atom_number'].tolist()\n",
    "NeighCAlpha = conect.loc[CAlpha].iloc[0].dropna().astype(int).values.tolist()\n",
    "CBeta = df[(df['atom_number'].isin(NeighCAlpha))&(df['element_symbol']=='C')]['atom_number'].tolist()\n",
    "NeighCBeta = conect.loc[CBeta].iloc[0].dropna().astype(int).values.tolist()\n",
    "CBetaBonds = [k for k,v in conect.loc[CBeta].fillna('NO').to_dict('index').items() for x in v.items() if x[1]!='NO']\n",
    "CBetaAtoms = pd.DataFrame([[x, CBetaBonds.count(x)] for x in set(CBetaBonds)], columns=['atom','bonds'])\n",
    "CBetaAtoms['type'] = CBetaAtoms['bonds'].apply(lambda x: 'decarboxylase' if x==3 else 'betalyase')\n",
    "CBetaAtoms['atom'] = CBetaAtoms['atom'].apply(lambda x: df[df['atom_number']==x]['atom_name'].iloc[0])\n",
    "CBetaAtoms = dict(CBetaAtoms[['type', 'atom']].values.tolist())\n",
    "\n",
    "## PLP PLane atoms\n",
    "CPLP = df[(df['atom_number'].isin(Neigh))&(df['residue_name']=='PLP')]['atom_number'].tolist()\n",
    "NeighCPLP = conect.loc[CPLP].iloc[0].dropna().astype(int).values.tolist()\n",
    "CFulcrum = df[(df['atom_number'].isin(NeighCPLP))&(df['element_symbol']=='C')]['atom_number'].tolist()\n",
    "Plane = conect.loc[CFulcrum].iloc[0].dropna().astype(int).values.tolist()\n",
    "PlaneAtoms = df[df['atom_number'].isin(Plane)]['atom_name'].tolist()\n",
    "CAlpha = df[df['atom_number'].isin(CAlpha)]['atom_name'].tolist()[0]\n",
    "CPLP = df[df['atom_number'].isin(CPLP)]['atom_name'].tolist()[0]\n",
    "CBetaAtoms, PlaneAtoms, CAlpha\n",
    "PlaneAtoms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba65daf3-538d-42c9-af87-f2ab97367062",
   "metadata": {},
   "source": [
    "## `prepare_ligand` by ADFRsuite-1.0 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "a16b161f-0944-4eb1-a787-11c16d3eff16",
   "metadata": {},
   "outputs": [],
   "source": [
    "! {adfr_path}/prepare_ligand -l {file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52170a5a-3293-46dd-97da-c6d9331d5c0b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Receptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "52ddc2aa-bcd2-4873-889b-a87da7cccd84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd, gzip, tarfile, urllib, os, sys, glob, requests, json, random, numpy as np, re, time\n",
    "from Bio.Blast.Applications import NcbiblastpCommandline as blastp\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import SeqIO, SearchIO\n",
    "from Bio.PDB import PDBParser as PDB, PDBIO\n",
    "# from sklearn.cluster import DBSCAN\n",
    "from multiprocessing import Pool\n",
    "# from biopandas.pdb import PandasPdb\n",
    "import tempfile as tfile\n",
    "from subprocess import Popen, PIPE\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import math\n",
    "from pymol import cmd\n",
    "from io import StringIO\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c9c3bab2-4a52-41ec-99cd-35c33a7267d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_ac(list, From, To, taxon='*'):\n",
    "    '''\n",
    "    IN: list of accession number (NCBI,UniProt; ENSEMBL, orthodb, ecc)\n",
    "    original DB (see: https://www.uniprot.org/help/api_idmapping)\n",
    "    new DB (see: https://www.uniprot.org/help/api_idmapping)\n",
    "    taxid (eg. 9606), DEFAULT=all\n",
    "    OUT: dataframe with ID mapping of the 2 database selected\n",
    "    '''\n",
    "    url = 'https://www.uniprot.org/uploadlists/'\n",
    "    \n",
    "    params = {\n",
    "        'from': From,\n",
    "        'to': To,\n",
    "        'format': 'tab',\n",
    "        'query': ' '.join(list),\n",
    "        'taxon': str(taxon),\n",
    "        'columns': 'id,reviewed',\n",
    "        }\n",
    "\n",
    "    data = urllib.parse.urlencode(params)\n",
    "    data = data.encode('utf-8')\n",
    "    req = urllib.request.Request(url, data)\n",
    "    df = pd.read_csv(urllib.request.urlopen(req), sep='\\t').drop('Status', axis=1, errors='ignore')\n",
    "    df = df[df.columns[:2]]\n",
    "    if 'Entry' in df.columns:\n",
    "        df.columns = ['To', 'From']\n",
    "    return df\n",
    "\n",
    "def pdb2fasta(pdb):\n",
    "    \"\"\"convert pdb file to fasta format for each chain\"\"\"\n",
    "    fasta = []\n",
    "    \n",
    "    if pdb.startswith('http'):\n",
    "        if pdb.endswith('.gz'):\n",
    "            seq = list(SeqIO.parse(gzip.open(urllib.request.urlopen(pdb),'rt'), 'pdb-atom'))\n",
    "        else:\n",
    "            seq = list(SeqIO.parse(StringIO(urllib.request.urlopen(pdb).read().decode('UTF-8')), 'pdb-atom'))\n",
    "\n",
    "    elif pdb.endswith('.gz'):\n",
    "        seq = list(SeqIO.parse(gzip.open(pdb,'rt'), 'pdb-atom'))\n",
    "        \n",
    "    else:\n",
    "        seq = list(SeqIO.parse(open(pdb,'r'), 'pdb-atom'))\n",
    "        \n",
    "    for record in seq:\n",
    "        \n",
    "        if record.annotations['start'] < 1:\n",
    "            fasta.append(SeqRecord((record.seq[abs(record.annotations['start'])+1:]), \n",
    "                             description = pdb.strip('.pdb'), \n",
    "                             id = record.annotations['chain']))\n",
    "        else:\n",
    "            fasta.append(SeqRecord(('X'*int(record.annotations['start']-1)+record.seq[0:]),\n",
    "                             description = pdb.strip('.pdb'),\n",
    "                             id = record.annotations['chain']))\n",
    "    return fasta\n",
    "\n",
    "\n",
    "def match_fasta_position(query,subject_list,num=None):\n",
    "    \"\"\"return dataframe of position matching in subject and query fasta, or a dictionary given a list of number\"\"\"\n",
    "    \n",
    "    subject = '\\n'.join([f'>{x.id}\\n{str(x.seq)}' for x in subject_list])\n",
    "    child = Popen(' '.join(['blastp', \n",
    "                         '-query',fr'<(echo -e \">{query.id}\\n{query.seq}\")',\n",
    "                         '-subject', fr'<(echo -e \"{subject}\")',\n",
    "                         '-evalue', '10e-5',\n",
    "                         '-outfmt', '5',\n",
    "                         '-max_hsps', '1']),\n",
    "                        stdout=PIPE, stderr=PIPE, shell=True, executable='/bin/bash'\n",
    "           ) \n",
    "    \n",
    "    df = []\n",
    "    xml = list(SearchIO.parse(child.stdout, \"blast-xml\"))\n",
    "    for q in xml:\n",
    "        for s in q:\n",
    "            x=s[0]\n",
    "            hit_gap = np.array([index for index, value in enumerate(x.hit) if value == '-'])\n",
    "            q_gap = np.array([index for index, value in enumerate(x.query) if value == '-'])\n",
    "            hit_num=list(np.arange(x.hit_start+1,x.hit_end+1))\n",
    "            q_num=list(np.arange(x.query_start+1,x.query_end+1))\n",
    "            for i in hit_gap:\n",
    "                hit_num.insert(i,np.nan)\n",
    "            for i in q_gap:\n",
    "                q_num.insert(i,np.nan)\n",
    "            df.append(list(zip(len(x.query)*[query.id],len(x.query)*[s.description],x.query,q_num,x.hit,hit_num)))\n",
    "    df = pd.DataFrame([j for i in df for j in i],columns=['query','sequence','query_res','query_num','hit_res','hit_num'])\n",
    "    df.set_index(['query'],inplace=True)\n",
    "    if num==None:\n",
    "        return df\n",
    "    else:\n",
    "        return df[df.query_num.isin(list(map(int,num)))].to_dict('records')\n",
    "    \n",
    "\n",
    "def build_homoligo(subject, template, ac='structure'):\n",
    "    cmd.reinitialize()\n",
    "    try:\n",
    "        cmd.load(template, 'template')\n",
    "    except:\n",
    "        cmd.read_pdbstr(urllib.request.urlopen(template).read().decode('UTF-8'), 'template')\n",
    "    cmd.split_states('template')\n",
    "    cmd.delete('template')\n",
    "    cmd.split_chains(f'template_0001 and polymer.protein')\n",
    "    cmd.delete(f'template_0001')    \n",
    "    df = match_fasta_position(pdb2fasta(subject)[0], pdb2fasta(template))\n",
    "    objs = list(cmd.get_object_list('all'))\n",
    "    cmd.load(subject, 'subject')\n",
    "    \n",
    "    n=65  ##letter A\n",
    "    for mol in objs:\n",
    "        chain = cmd.get_chains(mol)[0]\n",
    "        if chain in set(df['sequence'].tolist()):\n",
    "            cmd.create(f'sub_{mol}', f\"resi {int(df[df['sequence']==chain]['query_num'].min())}-{int(df[df['sequence']==chain]['query_num'].max())} and subject\")\n",
    "            cmd.alter(f'(sub_{mol})', f\"chain='{chr(n)}'\")\n",
    "            cmd.align(f'sub_{mol}', mol)\n",
    "            n+=1\n",
    "    cmd.save(f'{ac}.pdb', 'sub_*')\n",
    "\n",
    "def build_homoligo_manual(subjects, template, ac='structure'):\n",
    "    cmd.reinitialize()\n",
    "    try:\n",
    "        cmd.load(template, 'template')\n",
    "    except:\n",
    "        cmd.read_pdbstr(urllib.request.urlopen(template).read().decode('UTF-8'), 'template')\n",
    "    cmd.split_states('template')\n",
    "    cmd.delete('template')\n",
    "    cmd.split_chains(f'template_0001 and polymer.protein')\n",
    "    cmd.delete(f'template_0001')    \n",
    "    objs = list(cmd.get_object_list('all'))\n",
    "    \n",
    "    n=65  ##letter A\n",
    "    for chain, sub in subjects.items():\n",
    "        df = match_fasta_position(pdb2fasta(sub)[0], [s for s in pdb2fasta('https://files.rcsb.org/view/7CQK.pdb') if s.id==chain])\n",
    "        try:\n",
    "            cmd.load(sub, 'subject')\n",
    "        except:\n",
    "            cmd.read_pdbstr(urllib.request.urlopen(sub).read().decode('UTF-8'), 'subject')\n",
    "        mainChain = cmd.get_chains('subject')[0]\n",
    "        cmd.create(f'sub_{chain}', f\"resi {int(df[df['sequence']==chain]['query_num'].min())}-{int(df[df['sequence']==chain]['query_num'].max())} and subject and chain {mainChain}\")\n",
    "        cmd.alter(f'(sub_{chain})', f\"chain='{chr(n)}'\")\n",
    "        cmd.align(f'sub_{chain}', f'template_0001_{chain}')\n",
    "        n+=1\n",
    "        cmd.delete(f'subject')\n",
    "    cmd.save(f'{ac}.pdb', 'sub_*')\n",
    "    \n",
    "def best_structure(ac):\n",
    "    try:\n",
    "        df = getfeatures(ac)\n",
    "        leng = df['sequence.length'].iloc[0]*0.7\n",
    "        pdb = pd.json_normalize([x  for x in df.dbReferences[0] if x['type']=='PDB'])\n",
    "        pdb = pdb[pdb['properties.chains'].apply(lambda x: abs(eval(x.split('=')[-1]))) > leng]\n",
    "        multi = pdb[pdb['properties.chains'].str.contains('/')]\n",
    "        pdb = multi if not multi.empty else pdb\n",
    "        pdb['properties.resolution'] = pdb['properties.resolution'].apply(lambda x: x.split(' ')[0]).astype(float)\n",
    "        return pdb.loc[pdb['properties.resolution']==pdb['properties.resolution'].min()]['id'].iloc[0]\n",
    "    except:\n",
    "        return None\n",
    "    \n",
    "def getfeatures(uniprot_id):\n",
    "    \"\"\"return the Uniprot complete dataset for a given Uniprot ID\"\"\"\n",
    "    try:\n",
    "        r = requests.get(\"https://www.ebi.ac.uk/proteins/api/proteins?size=-1&accession=\" + uniprot_id, \n",
    "                         headers={ \"Accept\" : \"application/json\"})\n",
    "        data = pd.json_normalize(r.json())\n",
    "        return data\n",
    "    except:\n",
    "        return str(uniprot_id) + ' not_found'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d1b2c195-51de-4c26-8532-c177b8dabf19",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5de1c9d2-a588-4356-a619-54baacad225b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(cwd)\n",
    "specie = 'Mus_musculus'\n",
    "taxid = 10090\n",
    "try: # existing folder\n",
    "    os.mkdir(specie)\n",
    "except:\n",
    "    pass\n",
    "os.chdir(specie)\n",
    "ncbi_acc = pd.read_table(f'http://bioinformatics.unipr.it/B6db/tmp/{specie}.tab',header=None)\n",
    "acc = convert_ac(ncbi_acc[2],'P_REFSEQ_AC','ACC', taxid)\n",
    "gene_id = convert_ac(acc['To'],'ACC', 'P_ENTREZGENEID', taxid)\n",
    "accession = convert_ac(gene_id['To'].drop_duplicates().astype(str),'P_ENTREZGENEID', 'ACC', taxid).drop_duplicates('From')['To'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fe8a2ebd-ad45-44f7-9099-f29429dc6dc8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q91W43\n",
      "P50431\n",
      "Q9CZN7\n",
      "O88986\n",
      "P08680\n",
      "Q8VC19\n",
      "O35704\n",
      "P97363_O35704\n",
      "Q8BG54_O35704\n",
      "P97363\n",
      "Q8BG54\n",
      "Q9WUB3\n",
      "Q8CI94\n",
      "Q9ET01\n",
      "P29758\n",
      "P61922\n",
      "P05202\n",
      "P05201\n",
      "Q8BGT5\n",
      "Q8QZR5\n",
      "Q9WVM8\n",
      "P24288\n",
      "O35855\n",
      "Q3UEG6\n",
      "Q71RI9\n",
      "Q8BTY1\n",
      "Q9Z1J3\n",
      "Q14CH1\n",
      "Q6P6M7\n",
      "Q9CXF0\n",
      "Q99K01\n",
      "P48318\n",
      "P48320\n",
      "P00860\n",
      "P23738\n",
      "O88533\n",
      "Q80WP8\n",
      "Q9DBE0\n",
      "Q8R0X7\n",
      "Q6XPS7\n",
      "Q91WT9\n",
      "Q8BH55\n",
      "Q80W22\n",
      "Q8R1K4\n",
      "Q8BWU8\n",
      "Q8VBT2\n",
      "Q8R238\n",
      "Q8VCN5\n",
      "Q3UX83\n",
      "A2AIG8\n",
      "Q9JLI6\n",
      "Q7TSV6\n",
      "Q9QZX7\n",
      "B7ZWH9\n",
      "O35484\n",
      "Q8BVM4\n",
      "Q8QZR1\n",
      "O35423\n",
      "Q99K85\n",
      "Q9Z2Y8\n"
     ]
    }
   ],
   "source": [
    "ALPHAFOLD = False\n",
    "\n",
    "sptlc_ac = convert_ac(['438936at2759'],'ORTHODB_ID', 'ACC', taxid)\n",
    "sptlc_ac = sptlc_ac[sptlc_ac['To'].isin(accession)]\n",
    "s_list = convert_ac(sptlc_ac['To'],'ACC', 'GENENAME',taxid).sort_values('From')['From'].tolist()\n",
    "sptlc = {s_list[0]:s_list[1:]}\n",
    "\n",
    "templates = []\n",
    "for ac in accession:\n",
    "    print(ac)\n",
    "    if ac not in s_list:\n",
    "        try:\n",
    "            j = json.load(urllib.request.urlopen(f'https://swissmodel.expasy.org/repository/uniprot/{ac}.json'))['result']['structures']\n",
    "            time.sleep(0.1)    \n",
    "            try:\n",
    "                pdb = pd.json_normalize([x for x in j if 'homo' in x['oligo-state']][0], max_level=0)\n",
    "            except:\n",
    "                pdb = pd.json_normalize([x for x in j][0], max_level=0)\n",
    "            pdb.index = [ac]\n",
    "            if ALPHAFOLD:\n",
    "                build_homoligo(f'https://alphafold.ebi.ac.uk/files/AF-{ac}-F1-model_v1.pdb', pdb['coordinates'][0], f\"{ac}_{pdb['provider'][0]}\")\n",
    "            else:\n",
    "                urllib.request.urlretrieve(pdb['coordinates'][0], f\"{ac}_{pdb['provider'][0]}.pdb\")\n",
    "            templates.append(pdb.reset_index())\n",
    "        except:\n",
    "            templates.append(pd.DataFrame([], index=[ac], columns=['provider']))\n",
    "    elif sptlc.get(ac)!=None:\n",
    "        s_temp = 'O15269'\n",
    "        # try:\n",
    "        j_temp = json.load(urllib.request.urlopen(f'https://swissmodel.expasy.org/repository/uniprot/{s_temp}.json'))['result']['structures']\n",
    "        j = json.load(urllib.request.urlopen(f'https://swissmodel.expasy.org/repository/uniprot/{ac}.json'))['result']['structures']\n",
    "\n",
    "        # time.sleep(0.1)    \n",
    "        try:\n",
    "            pdb_temp = pd.json_normalize([x for x in j_temp if 'homo' in x['oligo-state']][0], max_level=0)\n",
    "        except:\n",
    "            pdb_temp = pd.json_normalize([x for x in j_temp][0], max_level=0)\n",
    "        try:\n",
    "            pdb = pd.json_normalize([x for x in j if 'homo' in x['oligo-state']][0], max_level=0)\n",
    "        except:\n",
    "            pdb = pd.json_normalize([x for x in j][0], max_level=0)\n",
    "\n",
    "        for s in sptlc.get(ac):\n",
    "            if ALPHAFOLD:\n",
    "                subjects = {'T':f'https://alphafold.ebi.ac.uk/files/AF-{s}-F1-model_v1.pdb',\n",
    "                            'S':f'https://alphafold.ebi.ac.uk/files/AF-{ac}-F1-model_v1.pdb',\n",
    "                           }\n",
    "            else:\n",
    "                j_s = json.load(urllib.request.urlopen(f'https://swissmodel.expasy.org/repository/uniprot/{s}.json'))['result']['structures']\n",
    "                try:\n",
    "                    pdb_s = pd.json_normalize([x for x in j_s if 'homo' in x['oligo-state']][0], max_level=0)\n",
    "                except:\n",
    "                    pdb_s = pd.json_normalize([x for x in j_s][0], max_level=0)\n",
    "\n",
    "                subjects = {'T':pdb_s['coordinates'][0],\n",
    "                            'S':pdb['coordinates'][0],\n",
    "                           }\n",
    "            build_homoligo_manual(subjects,\n",
    "                                   'https://files.rcsb.org/view/7CQK.pdb', \n",
    "                                   f\"{s}_{ac}_MANUAL\"\n",
    "                          )\n",
    "\n",
    "            print(f\"{s}_{ac}\")\n",
    "            pdb.index = [f\"{s}_{ac}\"]\n",
    "            templates.append(pdb.reset_index())\n",
    "        # except:\n",
    "            # templates.append(pd.DataFrame([], index=[ac], columns=['provider']))\n",
    "            \n",
    "            \n",
    "pd.concat(templates).rename(columns={'index':'ac'}).to_csv('templates.tsv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a80eaa8b-eca3-4231-b7c9-a6c9493f4748",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A2AIG8', 'A2AIG8_SWISSMODEL.pdb']\n",
      "['O35423', 'O35423_PDB.pdb']\n",
      "['O35484', 'O35484_SWISSMODEL.pdb']\n",
      "['O35855', 'O35855_SWISSMODEL.pdb']\n",
      "['O88533', 'O88533_SWISSMODEL.pdb']\n",
      "['O88986', 'O88986_SWISSMODEL.pdb']\n",
      "['P00860', 'P00860_PDB.pdb']\n",
      "['P05201', 'P05201_SWISSMODEL.pdb']\n",
      "['P05202', 'P05202_PDB.pdb']\n",
      "['P08680', 'P08680_SWISSMODEL.pdb']\n",
      "['P23738', 'P23738_SWISSMODEL.pdb']\n",
      "['P24288', 'P24288_SWISSMODEL.pdb']\n",
      "['P29758', 'P29758_SWISSMODEL.pdb']\n",
      "['P48318', 'P48318_SWISSMODEL.pdb']\n",
      "['P48320', 'P48320_SWISSMODEL.pdb']\n",
      "['P50431', 'P50431_PDB.pdb']\n",
      "['P61922', 'P61922_SWISSMODEL.pdb']\n",
      "['P97363', 'P97363_O35704_MANUAL.pdb']\n",
      "['Q14CH1', 'Q14CH1_SWISSMODEL.pdb']\n",
      "['Q3UEG6', 'Q3UEG6_SWISSMODEL.pdb']\n",
      "['Q3UX83', 'Q3UX83_SWISSMODEL.pdb']\n",
      "['Q6P6M7', 'Q6P6M7_PDB.pdb']\n",
      "['Q6XPS7', 'Q6XPS7_SWISSMODEL.pdb']\n",
      "['Q71RI9', 'Q71RI9_PDB.pdb']\n",
      "['Q7TSV6', 'Q7TSV6_SWISSMODEL.pdb']\n",
      "['Q80W22', 'Q80W22_SWISSMODEL.pdb']\n",
      "['Q80WP8', 'Q80WP8_PDB.pdb']\n",
      "['Q8BG54', 'Q8BG54_O35704_MANUAL.pdb']\n",
      "['Q8BGT5', 'Q8BGT5_SWISSMODEL.pdb']\n",
      "['Q8BH55', 'Q8BH55_SWISSMODEL.pdb']\n",
      "['Q8BTY1', 'Q8BTY1_SWISSMODEL.pdb']\n",
      "['Q8BVM4', 'Q8BVM4_SWISSMODEL.pdb']\n",
      "['Q8BWU8', 'Q8BWU8_SWISSMODEL.pdb']\n",
      "['Q8CI94', 'Q8CI94_SWISSMODEL.pdb']\n",
      "['Q8QZR1', 'Q8QZR1_PDB.pdb']\n",
      "['Q8QZR5', 'Q8QZR5_SWISSMODEL.pdb']\n",
      "['Q8R0X7', 'Q8R0X7_SWISSMODEL.pdb']\n",
      "['Q8R1K4', 'Q8R1K4_SWISSMODEL.pdb']\n",
      "['Q8R238', 'Q8R238_SWISSMODEL.pdb']\n",
      "['Q8VBT2', 'Q8VBT2_SWISSMODEL.pdb']\n",
      "['Q8VC19', 'Q8VC19_SWISSMODEL.pdb']\n",
      "['Q8VCN5', 'Q8VCN5_SWISSMODEL.pdb']\n",
      "['Q91W43', 'Q91W43_SWISSMODEL.pdb']\n",
      "['Q91WT9', 'Q91WT9_SWISSMODEL.pdb']\n",
      "['Q99K01', 'Q99K01_SWISSMODEL.pdb']\n",
      "['Q99K85', 'Q99K85_SWISSMODEL.pdb']\n",
      "['Q9CXF0', 'Q9CXF0_SWISSMODEL.pdb']\n",
      "['Q9CZN7', 'Q9CZN7_SWISSMODEL.pdb']\n",
      "['Q9DBE0', 'Q9DBE0_PDB.pdb']\n",
      "['Q9ET01', 'Q9ET01_SWISSMODEL.pdb']\n",
      "['Q9JLI6', 'Q9JLI6_SWISSMODEL.pdb']\n",
      "['Q9QZX7', 'Q9QZX7_SWISSMODEL.pdb']\n",
      "['Q9WUB3', 'Q9WUB3_SWISSMODEL.pdb']\n",
      "['Q9WVM8', 'Q9WVM8_SWISSMODEL.pdb']\n",
      "['Q9Z1J3', 'Q9Z1J3_SWISSMODEL.pdb']\n",
      "['Q9Z2Y8', 'Q9Z2Y8_SWISSMODEL.pdb']\n"
     ]
    }
   ],
   "source": [
    "output=[]\n",
    "for pdb in glob.glob(r'*.pdb'):\n",
    "    ## LLP residue ##\n",
    "    print(re.sub(r'HETATM(.*\\s*[N|C|O|CA|CB|CG|CE|CD|NZ]\\s*)LLP',\n",
    "                 r'ATOM  \\1LYS',open(pdb).read()),\n",
    "          file=open(pdb,'w'))\n",
    "    print(re.sub(r'(?m).*LLP.*\\n?', '', open(pdb).read()),\n",
    "          file=open(pdb,'w'))\n",
    "    \n",
    "    ID = os.path.basename(pdb[0:6])\n",
    "    try:\n",
    "        structure = PDB().get_structure(pdb, pdb)[0]\n",
    "        uni =  getfeatures(ID)\n",
    "        features = pd.DataFrame(uni.features[0])\n",
    "        lys = features[(features.category == 'PTM')&(features.description.str.contains('pyridoxal'))].begin.dropna().tolist()[0]\n",
    "        for row in match_fasta_position(\n",
    "            SeqRecord(Seq(uni['sequence.sequence'].values[0]), id=ID), \n",
    "                                        pdb2fasta(pdb), \n",
    "                                        [lys]):\n",
    "            coord=tuple(structure[row['sequence']][int(row['hit_num'])]['NZ'].get_coord())\n",
    "            output.append([ID, pdb, lys, row['sequence'], row['hit_num'], *coord])\n",
    "    except:\n",
    "        output.append([ID, pdb])\n",
    "    print([ID, pdb])\n",
    "pd.DataFrame(output, columns=['uniprot_ac','pdb','lys','chain','res','x','y','z']).to_csv(f'{specie}_coord.csv',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7aa574b-1927-4b53-8cdc-edbaf33807e1",
   "metadata": {},
   "source": [
    "# Reverse docking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "bd813bb1-abe1-4cfa-a0e9-03a2d092149b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cwd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [299]\u001b[0m, in \u001b[0;36m<cell line: 8>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpymol\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m cmd\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdatetime\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m date\n\u001b[0;32m----> 8\u001b[0m os\u001b[38;5;241m.\u001b[39mchdir(\u001b[43mcwd\u001b[49m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cwd' is not defined"
     ]
    }
   ],
   "source": [
    "import re, sys, os, subprocess, glob, configparser, pandas as pd, numpy as np\n",
    "from multiprocessing import Pool\n",
    "from Bio.PDB.vectors import *\n",
    "import math\n",
    "from pymol import cmd\n",
    "from datetime import date\n",
    "\n",
    "os.chdir(cwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "02578eea-dbd2-4fab-9892-6677524f891c",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_site = False\n",
    "ligand = file.replace('.pdb', '.pdbqt')           # ligand file\n",
    "receptor_dir = specie                        # path to receptor pdb files\n",
    "coord_file = f\"{specie}/{specie}_coord.csv\"        # table of gridcenter:  receptor_name1 x y z\n",
    "processes = 5                                                  #                       receptor_name2 x y z\n",
    "outdir = f\"./{ligand.replace('.pdbqt', '')}_{specie}\"                  # directory of results\n",
    "#[GPF]\n",
    "npts=\"25 25 25\"  ## unit in ngstrm\n",
    "spacing=0.375    ## unit in ngstrm\n",
    "#[DPF]\n",
    "ga_num_evals=1000000  # number of energy evaluations\n",
    "ga_run=5   # number of independent docking runs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe3a5e07-1d88-4cb4-afc9-0bcf34bff5c8",
   "metadata": {},
   "source": [
    "## make outdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "79996b7f-d140-4be0-8891-fa2c873f6972",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'date' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [297]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m i \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m----> 2\u001b[0m today \u001b[38;5;241m=\u001b[39m \u001b[43mdate\u001b[49m\u001b[38;5;241m.\u001b[39mtoday()\n\u001b[1;32m      3\u001b[0m outdir_date \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00moutdir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtoday\u001b[38;5;241m.\u001b[39mstrftime(\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mm-\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mY\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(outdir_date\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m i):\n",
      "\u001b[0;31mNameError\u001b[0m: name 'date' is not defined"
     ]
    }
   ],
   "source": [
    "i = 1\n",
    "today = date.today()\n",
    "outdir_date = f\"{outdir}_{today.strftime('%d-%m-%Y')}\"\n",
    "\n",
    "while os.path.exists(outdir_date+'_%s' % i):\n",
    "    i += 1\n",
    "out_dir = f\"{outdir_date}_{i}/\"\n",
    "os.makedirs(outdir_date)\n",
    "if os.path.exists(outdir_date+'_%s' % i):\n",
    "    print('writing output in '+out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b5164d90-2c4e-4a11-bacb-a44088d4f18f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Mus_musculus/Mus_musculus_coord.csv ...\n",
      "Proceeding with one site docking\n"
     ]
    }
   ],
   "source": [
    "####    read coordinates    ####\n",
    "\n",
    "print(f'Reading {coord_file} ...')\n",
    "\n",
    "if all_site:\n",
    "    coordinates = (pd.read_csv(coord_file, sep='\\t', comment='#').dropna().groupby('pdb')\n",
    "                       .apply(lambda r: r[['x','y','z', 'lys', 'chain']].values.tolist())\n",
    "                       .to_dict())\n",
    "    print(f'Proceeding with all site docking')\n",
    "\n",
    "else:\n",
    "    coordinates = (pd.read_csv(coord_file, sep='\\t', comment='#').dropna().sort_values('chain').drop_duplicates('pdb').groupby('pdb')\n",
    "                     .apply(lambda r: r[['x','y','z', 'res', 'chain']].values.tolist())\n",
    "                     .to_dict())\n",
    "    print(f'Proceeding with one site docking')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f78c8732-98da-4c1e-88b1-51d6166c1dc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "####   copy the ligand   ####\n",
    "os.system(f'cp {ligand} {out_dir}')\n",
    "lig_name = os.path.splitext(os.path.basename(ligand))[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17269306-4b73-4a6a-915c-45dd7d0d32a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98394399-ebcb-461c-b35a-6c9954459dea",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from A2AIG8_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:03:10 2022\n",
      "\n",
      "loading receptor: A2AIG8_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:     1.123    59.412   -46.399\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 10 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -273.66   466    7.30     -0.59      0.97      60.25\n",
      "        2    -96.62   177    4.47     -0.55      0.85      28.72\n",
      "        3    -32.52    60    2.44     -0.54      0.81      16.21\n",
      "        4    -27.38    44    2.03     -0.62      0.85      15.63\n",
      "        5    -20.46    27    1.69     -0.76      0.97      15.18\n",
      "        6    -14.86    27    1.56     -0.55      0.86      12.64\n",
      "        7    -11.90    13    1.40     -0.92      0.99       9.08\n",
      "        8    -11.81    11    1.26     -1.07      0.99       8.58\n",
      "        9    -10.51    26    1.56     -0.40      0.63       6.72\n",
      "       10     -7.41    11    1.35     -0.67      0.86       5.99\n",
      "    merging clusters ...\n",
      "done. got 862 fill Points, in 1.29 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(1.123 59.412 -46.399) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    862 points inside the box\n",
      "\n",
      "    maps computed in 5.33 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS324:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 18.5489299297\n",
      "writing maps ... done 4.93546700478\n",
      "done adding gradient to maps 24.85 (sec)\n",
      "making target file A2AIG8_SWISSMODEL_A.trg ...done.\n",
      "    done. 32.77 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:03:45 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/A2AIG8_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    2/   5 40.0% runs exhausted their evaluations\n",
      "    1/   5 20.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -7.9     0.0    -1.0       2     1.1     0.0    003\n",
      "   2         -7.4     3.3    -1.0       1      NA      NA    002\n",
      "   3         -6.1     3.4    -1.0       1      NA      NA    001\n",
      "   4         -5.6     3.7    -1.0       1      NA      NA    005\n",
      "Writing poses to LIG_PLP_A2AIG8_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 53.21 seconds, i.e. 0 hours 00 minutes 53.210397 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpg7ywNQ/A2AIG8_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpg7ywNQ/A2AIG8_SWISSMODEL_A/A2AIG8_SWISSMODEL_A.pdbqt\" -X \"A:LYS324\" -T \"/tmp/tmpg7ywNQ/A2AIG8_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"A2AIG8_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 257380 -j 1 -o \"LIG_PLP_A2AIG8_SWISSMODEL_A/A2AIG8_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_A2AIG8_SWISSMODEL_A.droin  0.10 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from O35423_PDB\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:04:43 2022\n",
      "\n",
      "loading receptor: O35423_PDB_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    -2.928     9.111     7.598\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "ERROR: running autogrid failed in /tmp/tmpC3MbDC.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 930, in __call__\n",
      "    raise RuntimeError(\"ERROR: running autogrid failed\")\n",
      "RuntimeError: ERROR: running autogrid failed\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'O35423_PDB_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:04:43 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/O35423_PDB_A.trg\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from O35855_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:04:47 2022\n",
      "\n",
      "loading receptor: O35855_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    -3.118    39.067    17.768\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 14 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -192.85   337    4.92     -0.57      0.93      59.00\n",
      "        2    -61.50   113    3.08     -0.54      0.92      31.24\n",
      "        3    -40.25    48    2.12     -0.84      0.99      22.24\n",
      "        4    -43.14    69    2.49     -0.63      0.84      19.66\n",
      "        5    -36.31    64    2.36     -0.57      0.83      18.91\n",
      "        6    -21.36    28    1.70     -0.76      0.95      14.91\n",
      "        7    -45.88    73    3.45     -0.63      0.81      13.73\n",
      "        8    -19.85    20    1.59     -0.99      1.00      12.59\n",
      "        9    -19.15    19    1.54     -1.01      0.99      12.17\n",
      "       10    -26.42    46    2.65     -0.57      0.82      11.72\n",
      "       11    -10.06    12    1.15     -0.84      0.92       8.78\n",
      "       12     -8.65    13    1.31     -0.67      0.90       8.05\n",
      "       13    -10.79    19    1.39     -0.57      0.71       6.86\n",
      "       14     -5.33    12    1.44     -0.44      0.70       4.10\n",
      "    merging clusters ...\n",
      "done. got 873 fill Points, in 1.30 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-3.118 39.067 17.768) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    873 points inside the box\n",
      "\n",
      "    maps computed in 4.92 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS230:HZ2,HZ3,CA,CB,CG,CD,CE,NZ,HZ1,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 19.7919859886\n",
      "writing maps ... done 4.84026503563\n",
      "done adding gradient to maps 25.98 (sec)\n",
      "making target file O35855_SWISSMODEL_A.trg ...done.\n",
      "    done. 33.67 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:05:23 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/O35855_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    1/   5 20.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    4/   5 80.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1        -10.5     0.0    -1.0       1      NA      NA    003\n",
      "   2         -8.1     7.1    -1.0       3     1.1     0.3    004\n",
      "   3         -7.7     4.9    -1.0       1      NA      NA    002\n",
      "Writing poses to LIG_PLP_O35855_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 53.21 seconds, i.e. 0 hours 00 minutes 53.212991 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpI82r5x/O35855_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpI82r5x/O35855_SWISSMODEL_A/O35855_SWISSMODEL_A.pdbqt\" -X \"A:LYS230\" -T \"/tmp/tmpI82r5x/O35855_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"O35855_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 58010 -j 1 -o \"LIG_PLP_O35855_SWISSMODEL_A/O35855_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_O35855_SWISSMODEL_A.droin  0.11 (s.)\n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from O88533_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:06:21 2022\n",
      "\n",
      "loading receptor: O88533_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -36.490    92.261    58.340\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 12 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -182.26   361    5.00     -0.50      0.88      55.71\n",
      "        2    -46.11    97    2.66     -0.48      0.89      28.88\n",
      "        3    -56.11    72    2.67     -0.78      0.97      25.22\n",
      "        4    -55.42    68    2.70     -0.82      0.99      24.87\n",
      "        5    -40.30    73    2.68     -0.55      0.82      18.45\n",
      "        6    -36.03    44    2.27     -0.82      0.95      17.42\n",
      "        7    -25.83    31    1.80     -0.83      0.98      16.64\n",
      "        8    -20.72    26    1.75     -0.80      0.96      13.63\n",
      "        9    -24.33    48    1.96     -0.51      0.74      13.24\n",
      "       10    -15.78    19    1.48     -0.83      0.94      11.31\n",
      "       11     -6.89    14    1.32     -0.49      0.61       3.93\n",
      "       12     -6.25    13    1.30     -0.48      0.55       3.08\n",
      "    merging clusters ...\n",
      "done. got 866 fill Points, in 1.31 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-36.490 92.261 58.340) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    866 points inside the box\n",
      "\n",
      "    maps computed in 6.74 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS303:HZ1,HZ2,HZ3,CA,CB,CG,CD,CE,NZ,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 22.8018968105\n",
      "writing maps ... done 4.95823001862\n",
      "done adding gradient to maps 29.27 (sec)\n",
      "making target file O88533_SWISSMODEL_A.trg ...done.\n",
      "    done. 38.63 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:07:02 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/O88533_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    5/   5 100.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    0/   5  0.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -8.9     0.0    -1.0       2     1.4     0.0    004\n",
      "   2         -8.3     3.7    -1.0       1      NA      NA    003\n",
      "   3         -8.1     3.6    -1.0       1      NA      NA    001\n",
      "   4         -7.6     5.6    -1.0       1      NA      NA    005\n",
      "Writing poses to LIG_PLP_O88533_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 71.32 seconds, i.e. 0 hours 01 minutes 11.317152 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpqW7YCE/O88533_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpqW7YCE/O88533_SWISSMODEL_A/O88533_SWISSMODEL_A.pdbqt\" -X \"A:LYS303\" -T \"/tmp/tmpqW7YCE/O88533_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"O88533_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 258733 -j 1 -o \"LIG_PLP_O88533_SWISSMODEL_A/O88533_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_O88533_SWISSMODEL_A.droin  0.13 (s.)\n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from O88986_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:08:18 2022\n",
      "\n",
      "loading receptor: O88986_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    -6.171    -0.359    26.176\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 10 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1    -93.18   150    3.26     -0.62      0.97      43.57\n",
      "        2    -94.91   183    3.55     -0.52      0.81      34.11\n",
      "        3    -46.05    52    2.38     -0.89      1.00      21.81\n",
      "        4    -35.12    42    1.99     -0.84      1.00      21.15\n",
      "        5    -29.96    37    2.06     -0.81      0.95      16.34\n",
      "        6    -30.50    64    2.34     -0.48      0.77      16.30\n",
      "        7    -23.17    25    1.61     -0.93      0.97      14.52\n",
      "        8    -20.98    21    1.52     -1.00      1.00      13.81\n",
      "        9    -15.00    17    1.43     -0.88      1.00      11.86\n",
      "       10     -9.96    17    1.39     -0.59      0.67       5.45\n",
      "    merging clusters ...\n",
      "done. got 608 fill Points, in 1.62 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-6.171 -0.359 26.176) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    608 points inside the box\n",
      "\n",
      "    maps computed in 5.53 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS262:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 19.6875\n",
      "writing maps ... done 5.18958091736\n",
      "done adding gradient to maps 26.27 (sec)\n",
      "making target file O88986_SWISSMODEL_A.trg ...done.\n",
      "    done. 34.26 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:08:55 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/O88986_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    3/   5 60.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -6.6     0.0    -1.0       1      NA      NA    005\n",
      "   2         -6.4     4.6    -1.0       1      NA      NA    001\n",
      "   3         -6.0     5.5    -1.0       1      NA      NA    003\n",
      "   4         -6.0     4.4    -1.0       1      NA      NA    004\n",
      "   5         -5.9     5.5    -1.0       1      NA      NA    002\n",
      "Writing poses to LIG_PLP_O88986_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 55.22 seconds, i.e. 0 hours 00 minutes 55.216407 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmp_2P0GM/O88986_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmp_2P0GM/O88986_SWISSMODEL_A/O88986_SWISSMODEL_A.pdbqt\" -X \"A:LYS262\" -T \"/tmp/tmp_2P0GM/O88986_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"O88986_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 17303 -j 1 -o \"LIG_PLP_O88986_SWISSMODEL_A/O88986_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_O88986_SWISSMODEL_A.droin  0.11 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P00860_PDB\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:09:54 2022\n",
      "\n",
      "loading receptor: P00860_PDB_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -10.848    10.748    -8.789\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 14 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -210.01   360    5.52     -0.58      0.91      54.05\n",
      "        2    -28.27    38    1.80     -0.74      1.00      21.08\n",
      "        3    -50.36    92    2.98     -0.55      0.79      19.19\n",
      "        4    -50.90    73    3.29     -0.70      0.85      16.09\n",
      "        5    -32.95    58    2.49     -0.57      0.82      15.73\n",
      "        6    -48.51    78    2.68     -0.62      0.72      15.14\n",
      "        7    -27.50    34    1.97     -0.81      0.94      15.10\n",
      "        8    -17.32    23    1.59     -0.75      0.99      14.28\n",
      "        9    -20.73    25    1.58     -0.83      0.76       9.25\n",
      "       10    -18.45    42    2.09     -0.44      0.68       9.22\n",
      "       11     -9.64    13    1.35     -0.74      0.92       8.23\n",
      "       12    -12.21    24    1.58     -0.51      0.72       7.74\n",
      "       13    -10.22    19    1.46     -0.54      0.66       5.65\n",
      "       14    -11.86    29    1.79     -0.41      0.58       5.36\n",
      "    merging clusters ...\n",
      "done. got 908 fill Points, in 1.38 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-10.848 10.748 -8.789) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    908 points inside the box\n",
      "\n",
      "    maps computed in 6.06 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS69:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 24.9312570095\n",
      "writing maps ... done 5.05528092384\n",
      "done adding gradient to maps 31.43 (sec)\n",
      "making target file P00860_PDB_A.trg ...done.\n",
      "    done. 40.73 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:10:37 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P00860_PDB_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    4/   5 80.0% runs exhausted their evaluations\n",
      "    1/   5 20.0% runs stopped converged 1 or 2 clusters\n",
      "    0/   5  0.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -8.5     0.0    -1.0       3     1.2     0.1    001\n",
      "   2         -7.9     5.1    -1.0       1      NA      NA    003\n",
      "   3         -7.7     4.9    -1.0       1      NA      NA    004\n",
      "Writing poses to LIG_PLP_P00860_PDB_A_out.pdbqt\n",
      "\n",
      "Docking performed in 60.19 seconds, i.e. 0 hours 01 minutes 00.188619 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpnZxvwq/P00860_PDB_A\" -M rigidReceptor -R \"/tmp/tmpnZxvwq/P00860_PDB_A/P00860_PDB_A.pdbqt\" -X \"A:LYS69\" -T \"/tmp/tmpnZxvwq/P00860_PDB_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P00860_PDB_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 183497 -j 1 -o \"LIG_PLP_P00860_PDB_A/P00860_PDB_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P00860_PDB_A.droin  0.11 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P05201_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:11:42 2022\n",
      "\n",
      "loading receptor: P05201_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -31.993   -38.149    -1.910\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "ERROR: running autogrid failed in /tmp/tmp6cQij2.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 930, in __call__\n",
      "    raise RuntimeError(\"ERROR: running autogrid failed\")\n",
      "RuntimeError: ERROR: running autogrid failed\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'P05201_SWISSMODEL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:11:43 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P05201_SWISSMODEL_A.trg\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P05202_PDB\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:11:47 2022\n",
      "\n",
      "loading receptor: P05202_PDB_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    14.422    13.605    43.215\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 12 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -193.90   299    4.86     -0.65      0.94      54.36\n",
      "        2    -57.14    72    2.85     -0.79      0.98      24.22\n",
      "        3    -25.76    29    1.84     -0.89      1.00      15.75\n",
      "        4    -32.34    63    2.50     -0.51      0.74      13.92\n",
      "        5    -23.20    48    2.15     -0.48      0.75      12.65\n",
      "        6    -16.12    16    1.29     -1.01      0.99      12.21\n",
      "        7    -27.46    55    2.72     -0.50      0.71      10.22\n",
      "        8    -10.01    13    1.24     -0.77      0.94       9.36\n",
      "        9     -8.64    11    1.14     -0.79      0.96       8.96\n",
      "       10    -10.23    23    1.57     -0.44      0.64       6.00\n",
      "       11     -5.77    11    1.22     -0.52      0.73       4.83\n",
      "       12     -5.83    13    1.35     -0.45      0.58       3.27\n",
      "    merging clusters ...\n",
      "done. got 653 fill Points, in 1.15 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(14.422 13.605 43.215) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    653 points inside the box\n",
      "\n",
      "    maps computed in 5.46 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS250:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 20.2707121372\n",
      "writing maps ... done 5.08917093277\n",
      "done adding gradient to maps 26.72 (sec)\n",
      "making target file P05202_PDB_A.trg ...done.\n",
      "    done. 34.73 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:12:24 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P05202_PDB_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    3/   5 60.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1        -11.9     0.0    -1.0       2     1.1     0.0    005\n",
      "   2         -9.2     5.8    -1.0       1      NA      NA    003\n",
      "   3         -8.4     8.0    -1.0       1      NA      NA    001\n",
      "   4         -7.4     7.0    -1.0       1      NA      NA    004\n",
      "Writing poses to LIG_PLP_P05202_PDB_A_out.pdbqt\n",
      "\n",
      "Docking performed in 56.19 seconds, i.e. 0 hours 00 minutes 56.193254 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpwhnqzz/P05202_PDB_A\" -M rigidReceptor -R \"/tmp/tmpwhnqzz/P05202_PDB_A/P05202_PDB_A.pdbqt\" -X \"A:LYS250\" -T \"/tmp/tmpwhnqzz/P05202_PDB_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P05202_PDB_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 120267 -j 1 -o \"LIG_PLP_P05202_PDB_A/P05202_PDB_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P05202_PDB_A.droin  0.11 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P08680_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:13:25 2022\n",
      "\n",
      "loading receptor: P08680_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    33.723   -27.133    -1.238\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 14 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -143.04   228    4.19     -0.63      0.98      52.38\n",
      "        2    -67.89    79    2.75     -0.86      0.99      28.19\n",
      "        3    -33.38    43    1.94     -0.78      0.98      21.35\n",
      "        4    -39.20    48    2.31     -0.82      0.98      20.11\n",
      "        5    -32.82    53    2.10     -0.62      0.89      19.97\n",
      "        6    -30.08    40    2.31     -0.75      0.99      16.86\n",
      "        7    -20.74    22    1.61     -0.94      1.00      13.67\n",
      "        8    -22.44    23    1.64     -0.98      0.95      12.76\n",
      "        9    -10.77    13    1.26     -0.83      1.00      10.32\n",
      "       10    -13.37    14    1.38     -0.96      0.99       9.98\n",
      "       11    -15.64    16    1.32     -0.98      0.90       9.92\n",
      "       12    -11.05    10    1.02     -1.11      0.96       8.94\n",
      "       13    -12.74    14    1.37     -0.91      0.93       8.85\n",
      "       14     -5.03    10    1.24     -0.50      0.91       6.74\n",
      "    merging clusters ...\n",
      "done. got 613 fill Points, in 1.33 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(33.723 -27.133 -1.238) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    613 points inside the box\n",
      "\n",
      "    maps computed in 7.08 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS391:CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,CA,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 658.44551897\n",
      "writing maps ... done 6.99160814285\n",
      "done adding gradient to maps 667.08 (sec)\n",
      "making target file P08680_SWISSMODEL_A.trg ...done.\n",
      "    done. 677.99 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:24:46 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P08680_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    0/   5  0.0% runs exhausted their evaluations\n",
      "    3/   5 60.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -7.9     0.0    -1.0       1      NA      NA    001\n",
      "   2         -7.6     3.3    -1.0       2     1.1     0.0    005\n",
      "   3         -3.4     4.9    -1.0       1      NA      NA    003\n",
      "   4         -3.0     4.6    -1.0       1      NA      NA    002\n",
      "Writing poses to LIG_PLP_P08680_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 38.26 seconds, i.e. 0 hours 00 minutes 38.260579 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpXYYvUg/P08680_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpXYYvUg/P08680_SWISSMODEL_A/P08680_SWISSMODEL_A.pdbqt\" -X \"A:LYS391\" -T \"/tmp/tmpXYYvUg/P08680_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P08680_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 451382 -j 1 -o \"LIG_PLP_P08680_SWISSMODEL_A/P08680_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P08680_SWISSMODEL_A.droin  0.16 (s.)\n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P23738_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:25:33 2022\n",
      "\n",
      "loading receptor: P23738_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -27.299    18.311   -26.551\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 7 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -187.88   314    4.82     -0.60      0.93      56.23\n",
      "        2    -15.60    17    1.35     -0.92      0.99      12.21\n",
      "        3    -12.24    15    1.33     -0.82      0.98      10.97\n",
      "        4    -13.50    19    1.46     -0.71      0.90      10.57\n",
      "        5    -18.00    37    2.03     -0.49      0.68       8.34\n",
      "        6    -13.58    24    1.69     -0.57      0.70       7.02\n",
      "        7    -10.94    25    1.54     -0.44      0.65       6.83\n",
      "    merging clusters ...\n",
      "done. got 451 fill Points, in 1.40 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-27.299 18.311 -26.551) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    451 points inside the box\n",
      "\n",
      "    maps computed in 7.16 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS312:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 42.0826249123\n",
      "writing maps ... done 5.94831490517\n",
      "done adding gradient to maps 52.01 (sec)\n",
      "making target file P23738_SWISSMODEL_A.trg ...done.\n",
      "    done. 61.78 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:26:36 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P23738_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    4/   5 80.0% runs exhausted their evaluations\n",
      "    1/   5 20.0% runs stopped converged 1 or 2 clusters\n",
      "    0/   5  0.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -9.2     0.0    -1.0       1      NA      NA    004\n",
      "   2         -8.2     3.8    -1.0       1      NA      NA    001\n",
      "   3         -7.5     5.4    -1.0       1      NA      NA    003\n",
      "   4         -7.3     6.6    -1.0       1      NA      NA    002\n",
      "   5         -5.4     6.9    -1.0       1      NA      NA    005\n",
      "Writing poses to LIG_PLP_P23738_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 59.22 seconds, i.e. 0 hours 00 minutes 59.218422 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpBw3Yp0/P23738_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpBw3Yp0/P23738_SWISSMODEL_A/P23738_SWISSMODEL_A.pdbqt\" -X \"A:LYS312\" -T \"/tmp/tmpBw3Yp0/P23738_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P23738_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 535932 -j 1 -o \"LIG_PLP_P23738_SWISSMODEL_A/P23738_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P23738_SWISSMODEL_A.droin  0.12 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P24288_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:27:40 2022\n",
      "\n",
      "loading receptor: P24288_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -33.230    -4.624   -17.849\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "ERROR: running autogrid failed in /tmp/tmpNKXQ3E.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 930, in __call__\n",
      "    raise RuntimeError(\"ERROR: running autogrid failed\")\n",
      "RuntimeError: ERROR: running autogrid failed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:27:41 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P24288_SWISSMODEL_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'P24288_SWISSMODEL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P29758_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:27:45 2022\n",
      "\n",
      "loading receptor: P29758_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    58.839    48.915    16.812\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 8 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -293.23   480    6.64     -0.61      0.94      63.91\n",
      "        2    -40.04    42    2.20     -0.95      1.00      18.91\n",
      "        3    -30.58    36    2.12     -0.85      0.99      16.54\n",
      "        4    -39.16    72    2.54     -0.54      0.73      14.99\n",
      "        5    -21.41    22    1.54     -0.97      0.99      13.92\n",
      "        6    -16.86    17    1.29     -0.99      0.96      12.07\n",
      "        7    -18.37    32    1.77     -0.57      0.71       9.16\n",
      "        8    -11.42    11    1.23     -1.04      1.00       8.92\n",
      "    merging clusters ...\n",
      "done. got 712 fill Points, in 1.20 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(58.839 48.915 16.812) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    712 points inside the box\n",
      "\n",
      "    maps computed in 5.19 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS292:CG,CD,CE,NZ,HZ1,HZ2,HZ3,CA,CB,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 20.2701749802\n",
      "writing maps ... done 4.86972999573\n",
      "done adding gradient to maps 26.52 (sec)\n",
      "making target file P29758_SWISSMODEL_A.trg ...done.\n",
      "    done. 34.35 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:28:21 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P29758_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    1/   5 20.0% runs exhausted their evaluations\n",
      "    1/   5 20.0% runs stopped converged 1 or 2 clusters\n",
      "    3/   5 60.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -7.7     0.0    -1.0       1      NA      NA    005\n",
      "   2         -7.5     7.8    -1.0       1      NA      NA    002\n",
      "   3         -7.0     8.5    -1.0       2     1.3     0.0    001\n",
      "   4         -6.8     7.2    -1.0       1      NA      NA    003\n",
      "Writing poses to LIG_PLP_P29758_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 60.27 seconds, i.e. 0 hours 01 minutes 00.274792 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpE8_yIb/P29758_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpE8_yIb/P29758_SWISSMODEL_A/P29758_SWISSMODEL_A.pdbqt\" -X \"A:LYS292\" -T \"/tmp/tmpE8_yIb/P29758_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P29758_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 231118 -j 1 -o \"LIG_PLP_P29758_SWISSMODEL_A/P29758_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P29758_SWISSMODEL_A.droin  0.12 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P48318_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:29:28 2022\n",
      "\n",
      "loading receptor: P48318_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -10.647     0.114    29.002\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 11 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -184.87   293    4.80     -0.63      0.92      51.78\n",
      "        2    -45.72    56    2.30     -0.82      0.98      23.30\n",
      "        3    -43.51    55    2.52     -0.79      0.97      20.43\n",
      "        4    -39.15    64    2.83     -0.61      0.86      16.84\n",
      "        5    -18.95    23    1.48     -0.82      0.97      14.56\n",
      "        6    -17.80    25    1.66     -0.71      0.97      14.13\n",
      "        7    -26.02    47    2.17     -0.55      0.74      11.97\n",
      "        8    -31.26    57    2.59     -0.55      0.70      10.78\n",
      "        9    -13.42    13    1.24     -1.03      1.00      10.52\n",
      "       10    -17.14    23    1.76     -0.75      0.84       9.14\n",
      "       11     -8.03    10    1.30     -0.80      0.98       7.38\n",
      "    merging clusters ...\n",
      "done. got 666 fill Points, in 1.40 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-10.647 0.114 29.002) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    666 points inside the box\n",
      "\n",
      "    maps computed in 7.23 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS404:CD,CE,NZ,HZ1,HZ2,HZ3,CA,CB,CG,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 23.1573340893\n",
      "writing maps ... done 5.98670983315\n",
      "done adding gradient to maps 30.73 (sec)\n",
      "making target file P48318_SWISSMODEL_A.trg ...done.\n",
      "    done. 41.24 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:30:11 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P48318_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    2/   5 40.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    3/   5 60.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -7.7     0.0    -1.0       1      NA      NA    003\n",
      "   2         -7.3     4.2    -1.0       2     1.5     0.0    005\n",
      "   3         -7.2     5.8    -1.0       1      NA      NA    002\n",
      "   4         -6.2     6.9    -1.0       1      NA      NA    004\n",
      "Writing poses to LIG_PLP_P48318_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 60.26 seconds, i.e. 0 hours 01 minutes 00.264659 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpiHR04J/P48318_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpiHR04J/P48318_SWISSMODEL_A/P48318_SWISSMODEL_A.pdbqt\" -X \"A:LYS404\" -T \"/tmp/tmpiHR04J/P48318_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P48318_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 424038 -j 1 -o \"LIG_PLP_P48318_SWISSMODEL_A/P48318_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P48318_SWISSMODEL_A.droin  0.11 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P48320_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:31:18 2022\n",
      "\n",
      "loading receptor: P48320_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   -10.702     0.108    29.077\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 12 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -193.26   314    5.52     -0.62      0.92      47.90\n",
      "        2    -47.09    56    2.27     -0.84      0.98      23.68\n",
      "        3    -61.82   118    3.63     -0.52      0.76      18.86\n",
      "        4    -21.39    30    1.77     -0.71      0.97      15.84\n",
      "        5    -19.37    27    1.69     -0.72      0.98      15.38\n",
      "        6    -24.86    29    1.81     -0.86      0.96      14.66\n",
      "        7    -20.78    27    1.72     -0.77      0.97      14.66\n",
      "        8    -20.02    23    1.51     -0.87      0.97      14.26\n",
      "        9    -29.54    42    2.29     -0.70      0.83      12.64\n",
      "       10    -16.57    16    1.36     -1.04      1.00      11.75\n",
      "       11    -14.86    18    1.55     -0.83      0.99      11.29\n",
      "       12     -4.56    11    1.28     -0.41      0.67       3.87\n",
      "    merging clusters ...\n",
      "done. got 711 fill Points, in 1.25 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(-10.702 0.108 29.077) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    711 points inside the box\n",
      "\n",
      "    maps computed in 5.86 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS396:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 19.3186180592\n",
      "writing maps ... done 5.04338479042\n",
      "done adding gradient to maps 26.19 (sec)\n",
      "making target file P48320_SWISSMODEL_A.trg ...done.\n",
      "    done. 34.61 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:31:54 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P48320_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    3/   5 60.0% runs exhausted their evaluations\n",
      "    0/   5  0.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -9.5     0.0    -1.0       1      NA      NA    004\n",
      "   2         -7.7     5.4    -1.0       2     1.0     0.0    001\n",
      "   3         -6.7     4.1    -1.0       2     1.1     0.0    002\n",
      "Writing poses to LIG_PLP_P48320_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 53.25 seconds, i.e. 0 hours 00 minutes 53.249607 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpjnASbg/P48320_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpjnASbg/P48320_SWISSMODEL_A/P48320_SWISSMODEL_A.pdbqt\" -X \"A:LYS396\" -T \"/tmp/tmpjnASbg/P48320_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P48320_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 613420 -j 1 -o \"LIG_PLP_P48320_SWISSMODEL_A/P48320_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P48320_SWISSMODEL_A.droin  0.11 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE15:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE88:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE133:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE148:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE164:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE220:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE223:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE260:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE314:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE355:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:A:MSE433:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE15:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE88:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE133:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE148:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE164:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE220:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE223:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE260:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE314:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE355:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:B:MSE433:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE15:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE88:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE133:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE148:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE164:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE220:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE223:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE260:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE314:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE355:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:C:MSE433:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE15:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE88:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE133:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE148:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE164:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE220:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE223:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE260:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE314:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE355:SE\n",
      "Sorry, there are no Gasteiger parameters available for atom P50431_PDB:D:MSE433:SE\n",
      "'Deleting non-standard residues:AMSE15_AMSE88_AMSE133_AMSE148_AMSE164_AMSE220_AMSE223_AMSE260_AMSE314_AMSE355_AMSE433_BMSE15_BMSE88_BMSE133_BMSE148_BMSE164_BMSE220_BMSE223_BMSE260_BMSE314_BMSE355_BMSE433_CMSE15_CMSE88_CMSE133_CMSE148_CMSE164_CMSE220_CMSE223_CMSE260_CMSE314_CMSE355_CMSE433_DMSE15_DMSE88_DMSE133_DMSE148_DMSE164_DMSE220_DMSE223_DMSE260_DMSE314_DMSE355_DMSE433_ from P50431_PDB\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:32:58 2022\n",
      "\n",
      "loading receptor: P50431_PDB_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   104.112    82.645    89.949\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 13 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -264.73   453    6.14     -0.58      0.97      69.51\n",
      "        2   -123.79   169    3.87     -0.73      0.98      41.57\n",
      "        3    -55.18    66    2.40     -0.84      1.00      27.30\n",
      "        4    -41.62    57    2.03     -0.73      0.98      27.20\n",
      "        5    -44.13    66    2.83     -0.67      0.97      21.93\n",
      "        6    -29.05    51    2.19     -0.57      0.97      21.67\n",
      "        7    -57.10    79    3.29     -0.72      0.89      18.91\n",
      "        8    -23.19    30    1.69     -0.77      0.99      17.32\n",
      "        9    -32.51    46    2.43     -0.71      0.94      16.76\n",
      "       10    -15.88    17    1.39     -0.93      1.00      12.19\n",
      "       11     -9.54    10    1.14     -0.95      1.00       8.77\n",
      "       12     -9.06    13    1.26     -0.70      0.91       8.57\n",
      "       13     -8.71    10    1.17     -0.87      0.96       7.88\n",
      "    merging clusters ...\n",
      "done. got 1067 fill Points, in 1.74 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(104.112 82.645 89.949) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    1067 points inside the box\n",
      "\n",
      "    maps computed in 8.36 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS251:HZ3,CA,CB,CG,CD,CE,NZ,HZ1,HZ2,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 20.1480500698\n",
      "writing maps ... done 4.92418813705\n",
      "done adding gradient to maps 26.48 (sec)\n",
      "making target file P50431_PDB_A.trg ...done.\n",
      "    done. 37.57 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:33:38 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P50431_PDB_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    1/   5 20.0% runs exhausted their evaluations\n",
      "    2/   5 40.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -8.2     0.0    -1.0       2     1.3     0.0    003\n",
      "   2         -7.7     5.6    -1.0       1      NA      NA    004\n",
      "   3         -7.4     5.5    -1.0       1      NA      NA    002\n",
      "   4         -5.0     4.7    -1.0       1      NA      NA    005\n",
      "Writing poses to LIG_PLP_P50431_PDB_A_out.pdbqt\n",
      "\n",
      "Docking performed in 52.19 seconds, i.e. 0 hours 00 minutes 52.194653 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmp4goKCx/P50431_PDB_A\" -M rigidReceptor -R \"/tmp/tmp4goKCx/P50431_PDB_A/P50431_PDB_A.pdbqt\" -X \"A:LYS251\" -T \"/tmp/tmp4goKCx/P50431_PDB_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"P50431_PDB_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 759999 -j 1 -o \"LIG_PLP_P50431_PDB_A/P50431_PDB_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_P50431_PDB_A.droin  0.12 (s.)\n",
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P61922_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:34:36 2022\n",
      "\n",
      "loading receptor: P61922_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    33.454     4.136    52.380\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 10 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -311.65   507    6.97     -0.61      0.97      67.81\n",
      "        2    -47.49    71    2.32     -0.67      0.96      28.06\n",
      "        3    -36.81    51    2.22     -0.72      0.95      20.82\n",
      "        4    -35.46    53    2.36     -0.67      0.96      20.82\n",
      "        5    -22.67    29    1.74     -0.78      0.95      14.95\n",
      "        6    -17.74    20    1.46     -0.89      1.00      13.70\n",
      "        7    -15.62    18    1.42     -0.87      1.00      12.67\n",
      "        8     -9.77    11    1.11     -0.89      1.00       9.95\n",
      "        9     -9.99    11    1.16     -0.91      0.95       8.59\n",
      "       10     -5.04    11    1.34     -0.46      0.63       3.32\n",
      "    merging clusters ...\n",
      "done. got 782 fill Points, in 1.31 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(33.454 4.136 52.380) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    782 points inside the box\n",
      "\n",
      "    maps computed in 6.04 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS357:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 991, in __call__\n",
      "    addGradients=self.receptorGradient)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 594, in computeGrids\n",
      "    addGradients=addGradients)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 649, in generateTrgFile\n",
      "    addGradientToMaps(mapFiles, mtypes, self.data['spacing'], self.cutOffValue, errorCut=0.01, logFileName=logFileName)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/addGradients.py\", line 43, in addGradientToMaps\n",
      "    thread.join()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/threading.py\", line 670, in join\n",
      "    self.__block.wait()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/threading.py\", line 243, in wait\n",
      "    waiter.acquire()\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:35:06 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P61922_SWISSMODEL_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'P61922_SWISSMODEL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from P97363_O35704_MANUAL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:35:10 2022\n",
      "\n",
      "loading receptor: P97363_O35704_MANUAL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:   125.530   177.959   113.238\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 15 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -177.25   315    5.00     -0.56      0.95      56.40\n",
      "        2    -79.44   123    3.24     -0.65      0.97      35.78\n",
      "        3    -79.59   153    3.27     -0.52      0.80      29.80\n",
      "        4    -37.10    50    2.07     -0.74      0.97      22.92\n",
      "        5    -28.17    34    2.11     -0.83      0.95      14.60\n",
      "        6    -22.72    36    2.09     -0.63      0.91      14.23\n",
      "        7    -26.22    56    2.31     -0.47      0.74      13.43\n",
      "        8    -20.41    25    1.80     -0.82      0.97      13.03\n",
      "        9    -18.58    19    1.52     -0.98      0.99      12.14\n",
      "       10    -14.05    20    1.49     -0.70      0.94      11.93\n",
      "       11    -14.91    19    1.43     -0.78      0.94      11.78\n",
      "       12    -19.11    16    1.37     -1.19      0.98      11.17\n",
      "       13    -10.30    14    1.30     -0.74      1.00      10.81\n",
      "       14     -7.01    10    1.19     -0.70      0.95       7.60\n",
      "       15     -4.62    10    1.21     -0.46      0.46       1.78\n",
      "    merging clusters ...\n",
      "done. got 900 fill Points, in 1.47 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(125.530 177.959 113.238) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    900 points inside the box\n",
      "\n",
      "ERROR: running autogrid failed in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P97363_O35704_MANUAL_A.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 994, in __call__\n",
      "    raise RuntimeError('ERROR: AutoGrid failed to run in %s'%gc.folder)\n",
      "RuntimeError: ERROR: AutoGrid failed to run in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P97363_O35704_MANUAL_A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:35:17 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/P97363_O35704_MANUAL_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'P97363_O35704_MANUAL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from Q14CH1_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:35:21 2022\n",
      "\n",
      "loading receptor: Q14CH1_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    28.977    48.787     6.696\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 9 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -157.97   294    5.95     -0.54      0.97      46.61\n",
      "        2   -101.89   177    3.45     -0.58      0.93      44.83\n",
      "        3    -96.88   159    4.74     -0.61      0.82      22.66\n",
      "        4    -48.56    60    2.62     -0.81      0.99      22.37\n",
      "        5    -18.46    22    1.57     -0.84      0.95      12.73\n",
      "        6    -14.90    19    1.43     -0.78      0.96      12.30\n",
      "        7    -12.90    15    1.34     -0.86      0.98      10.71\n",
      "        8    -13.58    14    1.32     -0.97      1.00      10.65\n",
      "        9    -11.85    19    1.76     -0.62      0.82       7.26\n",
      "    merging clusters ...\n",
      "done. got 779 fill Points, in 1.25 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(28.977 48.787 6.696) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    779 points inside the box\n",
      "\n",
      "    maps computed in 5.67 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS264:CA,CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,\n",
      "Adding gradient to maps ...\n",
      "processing maps ... done 20.1463971138\n",
      "writing maps ... done 4.89348316193\n",
      "done adding gradient to maps 26.44 (sec)\n",
      "making target file Q14CH1_SWISSMODEL_A.trg ...done.\n",
      "    done. 34.64 (sec)\n",
      "\n",
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:35:58 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q14CH1_SWISSMODEL_A.trg\n",
      "Performing search (5 GA evolutions with 1000000 maxEvals each) ...\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      "Termination status\n",
      "    0/   5  0.0% runs failed\n",
      "    0/   5  0.0% runs exhausted their evaluations\n",
      "    3/   5 60.0% runs stopped converged 1 or 2 clusters\n",
      "    2/   5 40.0% runs stopped after no improvement in clusters\n",
      "    0/   5  0.0% runs stopped because GA ran out of choices\n",
      "    0/   5  0.0% runs stopped because GA population converged\n",
      "\n",
      "Refining results ...\n",
      "done.\n",
      "\n",
      "mode |  affinity  | clust. | ref. | clust. | rmsd | energy | best |\n",
      "     | (kcal/mol) | rmsd   | rmsd |  size  | stdv |  stdv  | run  |\n",
      "-----+------------+--------+------+--------+------+--------+------+\n",
      "   1         -4.3     0.0    -1.0       1      NA      NA    003\n",
      "   2         -4.3     5.1    -1.0       3     0.8     0.6    005\n",
      "   3         -2.4     4.1    -1.0       1      NA      NA    004\n",
      "Writing poses to LIG_PLP_Q14CH1_SWISSMODEL_A_out.pdbqt\n",
      "\n",
      "Docking performed in 23.23 seconds, i.e. 0 hours 00 minutes 23.233286 seconds \n",
      "\n",
      "*************** first GA command ***************************\n",
      "\"/home/marco/ADFRsuite-1.0/bin/pythonsh\" \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runOneGA.py\" -F \"/tmp/tmpiiDm9h/Q14CH1_SWISSMODEL_A\" -M rigidReceptor -R \"/tmp/tmpiiDm9h/Q14CH1_SWISSMODEL_A/Q14CH1_SWISSMODEL_A.pdbqt\" -X \"A:LYS264\" -T \"/tmp/tmpiiDm9h/Q14CH1_SWISSMODEL_A/translationPoints.npy\" \"-l\" \"LIG_PLP.pdbqt\" \"-J\" \"Q14CH1_SWISSMODEL_A\" \"-e\" \"1000000\" \"-D\" \"3\" -S 223424 -j 1 -o \"LIG_PLP_Q14CH1_SWISSMODEL_A/Q14CH1_SWISSMODEL_A0001.dlg\"\n",
      "\n",
      "packaging docking results in to LIG_PLP_Q14CH1_SWISSMODEL_A.droin  0.11 (s.)\n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from Q3UEG6_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:36:31 2022\n",
      "\n",
      "loading receptor: Q3UEG6_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    44.318   133.252     9.279\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 11 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -234.10   426    6.39     -0.55      0.92      56.20\n",
      "        2    -35.30    44    1.91     -0.80      0.99      22.55\n",
      "        3    -32.77    52    2.24     -0.63      0.91      19.33\n",
      "        4    -26.89    31    1.78     -0.87      0.99      17.27\n",
      "        5    -21.36    26    1.76     -0.82      0.99      14.38\n",
      "        6    -26.54    32    2.04     -0.83      0.94      14.02\n",
      "        7    -15.84    19    1.36     -0.83      1.00      13.99\n",
      "        8    -22.31    30    2.03     -0.74      0.96      13.70\n",
      "        9    -15.45    20    1.45     -0.77      0.93      11.93\n",
      "       10     -8.46    13    1.23     -0.65      0.76       6.03\n",
      "       11     -4.45    10    1.21     -0.45      0.69       3.97\n",
      "    merging clusters ...\n",
      "done. got 703 fill Points, in 1.39 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(44.318 133.252 9.279) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    703 points inside the box\n",
      "\n",
      "    maps computed in 8.32 (sec)\n",
      "the following 9 flexible receptor atoms did not contribute to the grid calculation:\n",
      "  A:LYS349:CB,CG,CD,CE,NZ,HZ1,HZ2,HZ3,CA,\n",
      "Adding gradient to maps ...\n",
      "processing maps ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 991, in __call__\n",
      "    addGradients=self.receptorGradient)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 594, in computeGrids\n",
      "    addGradients=addGradients)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 649, in generateTrgFile\n",
      "    addGradientToMaps(mapFiles, mtypes, self.data['spacing'], self.cutOffValue, errorCut=0.01, logFileName=logFileName)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/addGradients.py\", line 43, in addGradientToMaps\n",
      "    thread.join()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/threading.py\", line 670, in join\n",
      "    self.__block.wait()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/threading.py\", line 243, in wait\n",
      "    waiter.acquire()\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:04 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q3UEG6_SWISSMODEL_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'Q3UEG6_SWISSMODEL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from Q3UX83_SWISSMODEL\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:08 2022\n",
      "\n",
      "loading receptor: Q3UX83_SWISSMODEL_A.pdbqt\n",
      "set box using user\n",
      "    Box center:     1.506    59.488   -46.322\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 11 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -189.60   362    5.43     -0.52      0.90      54.02\n",
      "        2    -68.25    93    3.16     -0.73      0.99      28.79\n",
      "        3    -31.39    45    2.33     -0.70      0.89      15.21\n",
      "        4    -29.41    40    2.15     -0.74      0.89      14.87\n",
      "        5    -17.29    20    1.49     -0.86      0.99      13.18\n",
      "        6    -15.80    17    1.30     -0.93      0.93      11.37\n",
      "        7    -15.67    41    1.96     -0.38      0.70      10.30\n",
      "        8     -9.14    13    1.25     -0.70      0.93       9.01\n",
      "        9     -7.05    12    1.21     -0.59      0.84       6.97\n",
      "       10     -7.34    10    1.30     -0.73      0.86       5.63\n",
      "       11    -11.18    24    1.75     -0.47      0.61       5.15\n",
      "    merging clusters ...\n",
      "done. got 677 fill Points, in 1.20 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(1.506 59.488 -46.322) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    677 points inside the box\n",
      "\n",
      "ERROR: running autogrid failed in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q3UX83_SWISSMODEL_A.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 994, in __call__\n",
      "    raise RuntimeError('ERROR: AutoGrid failed to run in %s'%gc.folder)\n",
      "RuntimeError: ERROR: AutoGrid failed to run in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q3UX83_SWISSMODEL_A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:14 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q3UX83_SWISSMODEL_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'Q3UX83_SWISSMODEL_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'_ ' apparently composed of not std residues. Deleting \n",
      "adding gasteiger charges to peptide\n",
      "'Deleting non-standard residues: from Q6P6M7_PDB\n",
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:24 2022\n",
      "\n",
      "loading receptor: Q6P6M7_PDB_A.pdbqt\n",
      "set box using user\n",
      "    Box center:    23.471    43.741    14.516\n",
      "    Box length:    25.000    25.000    25.000\n",
      "    Box size  :        68        68        68\n",
      "    padding   :     0.000\n",
      "    spacing   :     0.375\n",
      "\n",
      "identifying pockets using AutoSite ....\n",
      "    found 11 pocket(s)\n",
      "\n",
      "    pocket|  energy | # of |Rad. of | energy |   bns    | score  \n",
      "    number|         |points|gyration|per vol.|buriedness|v*b^2/rg\n",
      "    ------+---------+------+--------+--------+----------+---------\n",
      "        1   -104.65   165    4.03     -0.63      0.97      38.79\n",
      "        2    -68.61    88    2.80     -0.78      1.00      31.41\n",
      "        3    -63.89    85    3.10     -0.75      0.97      25.74\n",
      "        4    -24.49    33    1.75     -0.74      0.96      17.46\n",
      "        5    -33.37    66    2.38     -0.51      0.72      14.51\n",
      "        6    -15.24    19    1.45     -0.80      0.99      12.93\n",
      "        7    -27.14    47    2.06     -0.58      0.70      11.17\n",
      "        8    -28.90    39    2.32     -0.74      0.71       8.54\n",
      "        9    -17.50    27    1.83     -0.65      0.72       7.68\n",
      "       10     -8.01    15    1.42     -0.53      0.77       6.21\n",
      "       11     -9.36    17    1.52     -0.55      0.57       3.66\n",
      "    merging clusters ...\n",
      "done. got 601 fill Points, in 1.27 (sec)\n",
      "\n",
      "setting map types using: all to ['HS', 'Mg', 'HD', 'NA', 'Fe', 'Br', 'NS', 'A', 'C', 'Mn', 'G', 'F', 'I', 'H', 'J', 'N', 'Q', 'P', 'S', 'GA', 'Z', 'Zn', 'Cl', 'Ca', 'OA', 'SA', 'OS']\n",
      "\n",
      "computing maps for center=(23.471 43.741 14.516) size=(25.000 25.000 25.000) dims=(68 68 68) ...\n",
      "    601 points inside the box\n",
      "\n",
      "ERROR: running autogrid failed in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q6P6M7_PDB_A.\n",
      "  \"/home/marco/ADFRsuite-1.0/bin/autogrid4\" -p rigidReceptor.gpf -l rigidReceptor.glg\n",
      "Segmentation fault (core dumped)\n",
      ". \n",
      " \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 994, in __call__\n",
      "    raise RuntimeError('ERROR: AutoGrid failed to run in %s'%gc.folder)\n",
      "RuntimeError: ERROR: AutoGrid failed to run in /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q6P6M7_PDB_A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:33 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q6P6M7_PDB_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'Q6P6M7_PDB_A.trg'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/site.py\", line 563, in <module>\n",
      "    main()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/site.py\", line 545, in main\n",
      "    known_paths = addusersitepackages(known_paths)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/site.py\", line 278, in addusersitepackages\n",
      "    user_site = getusersitepackages()\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/site.py\", line 253, in getusersitepackages\n",
      "    user_base = getuserbase() # this will also set USER_BASE\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/site.py\", line 243, in getuserbase\n",
      "    USER_BASE = get_config_var('userbase')\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/sysconfig.py\", line 521, in get_config_var\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used ADFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "#                                                               #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "#                                                               #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Docking on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:34 2022\n",
      "reading ligand /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/LIG_PLP.pdbqt\n",
      "Detected 12 cores, request 5 cores, using 5 cores\n",
      "Unpacking maps /mnt/c/Users/marco/Desktop/jupyter_nb/LIG_PLP_Mus_musculus_1/Q71RI9_PDB_A.trg\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runADFR.py\", line 42, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runADFR.py\", line 497, in __call__\n",
      "    mf.unzipMaps()\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/maps.py\", line 121, in unzipMaps\n",
      "    zf = ZipFile(self.filename)\n",
      "  File \"/home/marco/ADFRsuite-1.0/lib/python2.7/zipfile.py\", line 701, in __init__\n",
      "    self.fp = open(file, modeDict[mode])\n",
      "IOError: [Errno 2] No such file or directory: 'Q71RI9_PDB_A.trg'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################################################################\n",
      "# If you used AGFR in your work, please cite:                   #\n",
      "#                                                               #\n",
      "# P.A. Ravindranath S. Forli, D.S. Goodsell, A.J. Olson and     #\n",
      "# M.F. Sanner                                                   #\n",
      "# AutoDockFR: Advances in Protein-Ligand Docking with           #\n",
      "# Explicitly Specified Binding Site Flexibility                 #\n",
      "# PLoS Comput Biol 11(12): e1004586                             #\n",
      "# DOI:10.1371/journal.pcbi.1004586                              #\n",
      "#                                                               #\n",
      "# P. Ananad Ravindranath and M.F. Sanner                        #\n",
      "# AutoSite: an automated approach for pseudoligands prediction  #\n",
      "# - From ligand binding sites identification to predicting key  #\n",
      "# ligand atoms                                                  #\n",
      "# Bioinformatics (2016)                                         #\n",
      "# DOI:10.1093/bioinformatics/btw367                             #\n",
      "#                                                               #\n",
      "# Please see http://adfr.scripps.edu for more information.      #\n",
      "#################################################################\n",
      "\n",
      "Computing grids on DESKTOP-E7E97C7 a Linux-4.4.0-22000-Microsoft-x86_64-with-debian-bullseye-sid computer\n",
      "Date Wed May 18 14:37:35 2022\n",
      "\n",
      "loading receptor: Q7TSV6_SWISSMODEL_A.pdbqt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/bin/runAGFR.py\", line 59, in <module>\n",
      "    runner(**kw)\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 824, in __call__\n",
      "    self.loadReceptor(kw['receptorFile'])  # reads receptor molecule from file and assigns Molecule instance to self.receptor \n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/utils/runAGFR.py\", line 147, in loadReceptor\n",
      "    if checkLigandFile(filename):\n",
      "  File \"/home/marco/ADFRsuite-1.0/CCSBpckgs/ADFR/__init__.py\", line 160, in checkLigandFile\n",
      "    f = open(filename)\n",
      "IOError: [Errno 2] No such file or directory: 'Q7TSV6_SWISSMODEL_A.pdbqt'\n"
     ]
    }
   ],
   "source": [
    "####    reverse docking    ####\n",
    "\n",
    "for pdb in coordinates:\n",
    "    try:\n",
    "        xyz = coordinates.get(pdb)\n",
    "        pdb_name = os.path.splitext(pdb)[0]\n",
    "        for c in xyz:\n",
    "            i = c[4]\n",
    "            if 'rec_files_dir' in locals():\n",
    "                os.system(f'cp {rec_files_dir}/{pdb_name}_{i}.trg {out_dir}')\n",
    "                os.system(f'cp {rec_files_dir}/{pdb_name}_{i}.pdbqt {out_dir}')\n",
    "            else:\n",
    "\n",
    "                ####    prepare receptor    ####\n",
    "                print(re.sub('HETATM', '#HETATM', open(f'{receptor_dir}/{pdb}', 'r').read()), \n",
    "                      file=open(out_dir+pdb, 'w'))                   # comment HETATM in pdb file\n",
    "                os.system(f\"{adfr_path}/prepare_receptor -A 'checkhydrogens' -e -r {receptor_dir}/{pdb} -o {out_dir}/{pdb_name}_{i}.pdbqt\") \n",
    "\n",
    "                ####    AGFR    ####\n",
    "                os.chdir(out_dir)\n",
    "                os.system(f'{adfr_path}/agfr -r {pdb_name}_{i}.pdbqt -b user {\" \".join(map(lambda x: str(round(x,3)),c[0:3]))} {npts} -s {spacing} -f {c[4]}:LYS{str(int(c[3]))}')\n",
    "                os.chdir(cwd)\n",
    "\n",
    "            ####    ADFR    ####\n",
    "            os.chdir(out_dir)\n",
    "            os.system(f'{adfr_path}/adfr -t {pdb_name}_{i}.trg -l {os.path.basename(ligand)} -J {pdb_name}_{i} -e {ga_num_evals} -n {ga_run} -f -c {processes} -D 3')\n",
    "            os.chdir(cwd)\n",
    "        \n",
    "    except:\n",
    "        print(f'{pdb} not found')\n",
    "        os.chdir(cwd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6937856d-2ddc-4c1c-ae9c-4ba054b77e89",
   "metadata": {},
   "source": [
    "# Results analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "577c0f80-9d22-438f-b64a-e6c61c8ea821",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import urllib, itertools, sys\n",
    "# from multiprocessing import Pool\n",
    "# from pymol import cmd, stored\n",
    "# from Bio.PDB.PDBParser import PDBParser\n",
    "# import pandas as pd, math\n",
    "# from revdock.revdocking import *\n",
    "# from math import sqrt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "b38b78a9-bac6-46bf-8f93-7df02129efa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'betalyase': 'C8', 'decarboxylase': 'C10'}"
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CBetaAtoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "b96fb257-fe54-4688-af29-f8c3f4e9016d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import sqrt \n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "def planeEq(p1,p2,p3):\n",
    "    p1 = np.array(p1)\n",
    "    p2 = np.array(p2)\n",
    "    p3 = np.array(p3)\n",
    "\n",
    "    v1 = p3 - p1\n",
    "    v2 = p2 - p1\n",
    "\n",
    "    cp = np.cross(v2, v1)\n",
    "    a, b, c = cp\n",
    "\n",
    "    d = np.dot(cp, p3)\n",
    "    return [a,b,c,d]\n",
    "\n",
    "def lineEq(p1,p2):\n",
    "    p1 = np.array(p1)\n",
    "    p2 = np.array(p2)\n",
    "\n",
    "    a, b, c = p2-p1\n",
    "\n",
    "    return [a,b,c]\n",
    "\n",
    "def angle_plane_line(a1,b1,c1,a2,b2,c2):\n",
    "    return abs((a1*a2 + b1*b2 + c1*c2)/(sqrt(a1**2 + b1**2 + c1**2)*sqrt(a2**2 + b2**2 + c2**2)))\n",
    "\n",
    "def calc_run(run_file, dihedral):\n",
    "    pdbqt = os.path.basename(run_file[:-10])\n",
    "    pdbqt = f\"{os.path.dirname(os.path.dirname(run_file))}/{os.path.basename(run_file[:-10])}\"\n",
    "\n",
    "    run = os.path.basename(run_file)\n",
    "    DIR = os.path.basename(os.path.dirname(run_file))\n",
    "    ## PYMOL\n",
    "    try:\n",
    "        cmd.reinitialize()\n",
    "        cmd.load(run_file)\n",
    "        # cmd.load(f\"{pdbqt}.pdbqt\")\n",
    "        cmd.h_add(f'name {str(dihedral[2])} and not resn lys')\n",
    "        n = cmd.count_atoms('not resn lys')\n",
    "\n",
    "        angles = {}\n",
    "        # aminotransferase\n",
    "        plane = cmd.get_coords(f\"name {'+'.join(dihedral[1])} and not resn lys\")\n",
    "        a2,b2,c2,d2 = planeEq(*plane)\n",
    "        line = cmd.get_coords(f'name {str(dihedral[2])} or id {str(n)} and not resn lys')\n",
    "        a1,b1,c1 = lineEq(*line)\n",
    "        angle = angle_plane_line(a1,b1,c1,a2,b2,c2)\n",
    "        angles.update({'aminotransferase':angle})\n",
    "\n",
    "        # betalyase, decarboxylase\n",
    "        for k,v in dihedral[0].items():\n",
    "            line = cmd.get_coords(f'name {str(dihedral[2])}+{v} and not resn lys')\n",
    "            a1,b1,c1 = lineEq(*line)\n",
    "            angle = angle_plane_line(a1,b1,c1,a2,b2,c2)\n",
    "            angles.update({k:angle})\n",
    "\n",
    "        d = cmd.get_distance(f'name {str(dihedral[3])} and not resn lys', 'name NZ and resn lys')\n",
    "        results = pd.DataFrame([[DIR, os.path.basename(run_file), d, *list(angles.values())]],\n",
    "                               columns = ['DIR', 'run_file', 'd', \n",
    "                                          f'{chr(948)} (A)', f'{chr(948)} (B)', f'{chr(948)} (D)']\n",
    "                              )\n",
    "        for column in results.columns[results.columns.str.contains(chr(948))]:\n",
    "            results[column.replace(f\"{chr(948)} \", '')] = results.apply(lambda x: 'yes' if all([x[column]>=0.9, x['d']<=5]) else 'no', axis=1) \n",
    "        return results\n",
    "    except:\n",
    "        return pd.DataFrame([[DIR, os.path.basename(run_file),None, None, None, None, None, None, None]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "3391ddb7-f781-4166-9c7e-0885d7c84904",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = out_dir\n",
    "os.chdir(directory)\n",
    "\n",
    "\n",
    "if glob.glob('*totali.csv'):\n",
    "    ade = pd.concat([pd.read_csv(x,names = ['trg_file', 'run_file', 'energy']) for x in glob.glob('*energie_totali.csv')])\n",
    "    ade = ade[ade['energy']<=0]\n",
    "else:\n",
    "    ade = pd.DataFrame({'energy': [], 'run_file':[]})\n",
    "\n",
    "dihedral = [CBetaAtoms, PlaneAtoms, CAlpha, CPLP]\n",
    "\n",
    "\n",
    "df = pd.concat(Pool(200).starmap(calc_run, itertools.product(glob.glob(f'*/*[0-9].pdbqt'), [dihedral])),\n",
    "                  columns = ['DIR', 'run_file', 'd', f'{chr(948)} (A)', f'{chr(948)} (B)', f'{chr(948)} (D)']\n",
    "                 )\n",
    "\n",
    "\n",
    "df.merge(ade, on='run_file', how='outer').to_csv(f'{os.path.basename(os.path.dirname(directory))}_runs.csv', index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfec9093-1718-4baf-9434-802db103ef40",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
